
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/taxa_genomes_stats_2025-11-13/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Taxa & Genomes Sheet Statistics - 2025-11-13 - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#taxa-genomes-sheet-statistics-2025-11-13" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Taxa &amp; Genomes Sheet Statistics - 2025-11-13
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#summary-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Summary Statistics
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-quality-scores" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Quality Scores
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Quality Scores">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#completeness-936100" class="md-nav__link">
    <span class="md-ellipsis">
      
        Completeness: 93.6/100
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#actionability-910100" class="md-nav__link">
    <span class="md-ellipsis">
      
        Actionability: 91.0/100
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#knowledge-integration-853100" class="md-nav__link">
    <span class="md-ellipsis">
      
        Knowledge Integration: 85.3/100
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#critical-issues" class="md-nav__link">
    <span class="md-ellipsis">
      
        Critical Issues
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Critical Issues">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#issue-1-source-field-pollution-high-priority" class="md-nav__link">
    <span class="md-ellipsis">
      
        ❌ Issue 1: Source Field Pollution (HIGH PRIORITY)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#issue-2-duplicate-taxon-ids-medium-priority" class="md-nav__link">
    <span class="md-ellipsis">
      
        ❌ Issue 2: Duplicate Taxon IDs (MEDIUM PRIORITY)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#issue-3-inconsistent-sp-formatting-low-priority" class="md-nav__link">
    <span class="md-ellipsis">
      
        ⚠ Issue 3: Inconsistent "sp." Formatting (LOW PRIORITY)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-provenance-maintenance" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Provenance &amp; Maintenance
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Provenance &amp; Maintenance">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-the-sheet-is-populated-5-methods" class="md-nav__link">
    <span class="md-ellipsis">
      
        How the Sheet is Populated (5 Methods)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#automated-maintenance-workflows" class="md-nav__link">
    <span class="md-ellipsis">
      
        Automated Maintenance Workflows
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#deduplication-strategy" class="md-nav__link">
    <span class="md-ellipsis">
      
        Deduplication Strategy
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kg_node_ids-column-structure" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg_node_ids Column Structure
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="kg_node_ids Column Structure">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#format-specification" class="md-nav__link">
    <span class="md-ellipsis">
      
        Format Specification
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#coverage-pattern" class="md-nav__link">
    <span class="md-ellipsis">
      
        Coverage Pattern
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-validation-results" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Validation Results
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Validation Results">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#no-mismatches-found" class="md-nav__link">
    <span class="md-ellipsis">
      
        ✓ No Mismatches Found
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#incomplete-records-19-rows" class="md-nav__link">
    <span class="md-ellipsis">
      
        Incomplete Records (19 rows)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#recommendations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Recommendations
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Recommendations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#quick-wins-can-implement-immediately" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quick Wins (Can implement immediately)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#requires-investigation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Requires Investigation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#long-term-improvements" class="md-nav__link">
    <span class="md-ellipsis">
      
        Long-term Improvements
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#action-items" class="md-nav__link">
    <span class="md-ellipsis">
      
        Action Items
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Action Items">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#high-priority" class="md-nav__link">
    <span class="md-ellipsis">
      
        High Priority
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#medium-priority" class="md-nav__link">
    <span class="md-ellipsis">
      
        Medium Priority
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#low-priority" class="md-nav__link">
    <span class="md-ellipsis">
      
        Low Priority
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#files-generated" class="md-nav__link">
    <span class="md-ellipsis">
      
        Files Generated
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="taxa-genomes-sheet-statistics-2025-11-13">Taxa &amp; Genomes Sheet Statistics - 2025-11-13</h1>
<p><strong>Sheet:</strong> <code>BER CMM Data for AI - for editing</code> → <code>taxa_and_genomes</code><br />
<strong>Date:</strong> 2025-11-13<br />
<strong>Total Rows:</strong> 211 (up from 197 in git)<br />
<strong>Overall Grade:</strong> B (90/100) - Good with minor normalization issues</p>
<hr />
<h2 id="summary-statistics">Summary Statistics</h2>
<table>
<thead>
<tr>
<th>Column Name</th>
<th>Blank</th>
<th>Populated</th>
<th>Unique</th>
<th>Duplicated</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td>Scientific name</td>
<td>0</td>
<td>211</td>
<td>211</td>
<td>0</td>
<td>Perfect primary key; mostly consistent 'sp.' usage</td>
</tr>
<tr>
<td>NCBITaxon id</td>
<td>1</td>
<td>210</td>
<td>208</td>
<td>2</td>
<td>⚠ 2 duplicates: 270351, 157278</td>
</tr>
<tr>
<td>Genome identifier (GenBank, IMG etc)</td>
<td>19</td>
<td>192</td>
<td>190</td>
<td>2</td>
<td>GCF:158 RefSeq, GCA:34 GenBank; 2 dups</td>
</tr>
<tr>
<td>source</td>
<td>12</td>
<td>199</td>
<td>65</td>
<td>2</td>
<td>⚠ INCONSISTENT: 63 URLs mixed with 136 labels</td>
</tr>
<tr>
<td>kg_node_ids</td>
<td>31</td>
<td>180</td>
<td>178</td>
<td>2</td>
<td>180 linked to both KGs; 2 dups</td>
</tr>
<tr>
<td>Annotation download URL</td>
<td>19</td>
<td>192</td>
<td>190</td>
<td>2</td>
<td>All FTP; 2 dups (match genome dups)</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="data-quality-scores">Data Quality Scores</h2>
<h3 id="completeness-936100">Completeness: 93.6/100</h3>
<ul>
<li><strong>100%</strong> Scientific names (211/211)</li>
<li><strong>99.5%</strong> NCBITaxon IDs (210/211)</li>
<li><strong>91.0%</strong> Genome identifiers (192/211)</li>
<li><strong>91.0%</strong> Annotation download URLs (192/211)</li>
<li><strong>94.3%</strong> Source labels (199/211, but 63 are incorrect)</li>
<li><strong>85.3%</strong> KG linkage (180/211)</li>
</ul>
<h3 id="actionability-910100">Actionability: 91.0/100</h3>
<ul>
<li><strong>192/211 (91%)</strong> have complete data (taxon + genome + URL)</li>
<li>Can immediately download 192 genome annotations</li>
<li>Can query phenotypes for 180 organisms via KG</li>
<li><strong>19 records</strong> need data completion</li>
</ul>
<h3 id="knowledge-integration-853100">Knowledge Integration: 85.3/100</h3>
<ul>
<li><strong>180/211 (85%)</strong> linked to kg-microbe (phenotypes) AND kg-microbe-function (proteins)</li>
<li><strong>132/211 (62.6%)</strong> found via lanthanide protein searches (xoxF, lanM, etc.)</li>
<li><strong>31/211 (14.7%)</strong> too new for KG (no experimental data yet)</li>
</ul>
<hr />
<h2 id="critical-issues">Critical Issues</h2>
<h3 id="issue-1-source-field-pollution-high-priority">❌ Issue 1: Source Field Pollution (HIGH PRIORITY)</h3>
<p><strong>Problem:</strong> 63 records (30%) have FTP URLs in the <code>source</code> field</p>
<p><strong>Expected:</strong>
<div class="highlight"><pre><span></span><code>source: &quot;UniProt API search: lanM, xoxF, mxaF...&quot;
source: &quot;extend1&quot;
source: &quot;fix_validation&quot;
</code></pre></div></p>
<p><strong>Actual (wrong):</strong>
<div class="highlight"><pre><span></span><code>source: &quot;ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/333/655/...&quot;
</code></pre></div></p>
<p><strong>Root Cause:</strong>
- Early implementation used <code>source</code> for FTP URLs (no separate column existed)
- <code>Annotation download URL</code> column was added later via <code>add_annotation_urls.py</code>
- Old URLs in <code>source</code> field were never cleaned up
- <strong>This is a migration artifact</strong></p>
<p><strong>Impact:</strong>
- Source provenance is lost/unclear
- Can't track which pipeline generated the data
- URLs duplicate the <code>Annotation download URL</code> column</p>
<p><strong>Distribution:</strong>
- 132 proper labels: "UniProt API search..."
- 63 legacy URLs: <code>ftp://...</code> ❌
- 12 blank: No tracking
- 4 proper labels: "fix_validation"</p>
<hr />
<h3 id="issue-2-duplicate-taxon-ids-medium-priority">❌ Issue 2: Duplicate Taxon IDs (MEDIUM PRIORITY)</h3>
<p><strong>Problem:</strong> 2 taxon IDs appear multiple times with different organism names</p>
<p><strong>Case 1: NCBITaxon:270351</strong>
- Row A: "Methylobacterium aquaticum CCM : 7218 , CECT : 5998 , CIP : 108333 , DSM : 16371 , personal : GR16" (no genome)
- Row B: "Methylobacterium aquaticum" (has genome GCF_050408745.1)
- <strong>Analysis:</strong> Same species, different strain designations
- <strong>Decision needed:</strong> Merge or keep separate?</p>
<p><strong>Case 2: NCBITaxon:157278</strong>
- Row A: "uncultured Methylobacterium sp." (with period)
- Row B: "uncultured Methylobacterium sp" (no period)
- <strong>Same genome:</strong> GCA_965531505.1
- <strong>Analysis:</strong> TRUE DUPLICATE (typo in name)
- <strong>Action:</strong> Merge rows</p>
<p><strong>Impact:</strong>
- Duplicate entry inflates row count
- Inconsistent organism naming
- May cause join/merge issues in downstream analysis</p>
<hr />
<h3 id="issue-3-inconsistent-sp-formatting-low-priority">⚠ Issue 3: Inconsistent "sp." Formatting (LOW PRIORITY)</h3>
<p><strong>Problem:</strong> Mixed usage of "sp." (59) vs "sp" (2)</p>
<p><strong>Examples:</strong>
- "Methylobacterium sp. MB200" ✓ (with period - correct scientific notation)
- "Paracoccus sp NSM" ✗ (without period)</p>
<p><strong>Impact:</strong> Minor - won't break analysis but reduces polish</p>
<p><strong>Fix:</strong> Standardize to "sp." (with period, per scientific convention)</p>
<hr />
<h2 id="data-provenance-maintenance">Data Provenance &amp; Maintenance</h2>
<h3 id="how-the-sheet-is-populated-5-methods">How the Sheet is Populated (5 Methods)</h3>
<table>
<thead>
<tr>
<th>Method</th>
<th>Count</th>
<th>Script</th>
<th>Names From</th>
<th>Source Label</th>
</tr>
</thead>
<tbody>
<tr>
<td>1. Manual Seed</td>
<td>2</td>
<td>Hand-entered</td>
<td>Researchers</td>
<td>(blank)</td>
</tr>
<tr>
<td>2. NCBI Assembly Search</td>
<td>~60</td>
<td><code>ncbi_search.py</code> → <code>extend_lanthanide_data.py</code></td>
<td>NCBI Assembly metadata</td>
<td>FTP URLs ❌ (should be "extend1")</td>
</tr>
<tr>
<td>3. UniProt API Protein Search</td>
<td>132</td>
<td><code>extend_from_lanm.py</code></td>
<td>UniProt organism field</td>
<td>"UniProt API search: lanM, xoxF..." ⭐</td>
</tr>
<tr>
<td>4. Validation Fixes</td>
<td>4</td>
<td><code>add_missing_organisms.py</code></td>
<td>genes/proteins table → NCBI Taxonomy</td>
<td>"fix_validation"</td>
</tr>
<tr>
<td>5. Recent Manual Additions</td>
<td>12</td>
<td>Direct Google Sheet edits</td>
<td>Researchers</td>
<td>(blank)</td>
</tr>
</tbody>
</table>
<p><strong>Key Insight:</strong> The majority (132/211 = 62.6%) of organisms were discovered via <strong>UniProt protein searches</strong>, not genome searches. These organisms have lanthanide-relevant proteins (xoxF, lanM, mxaF, exaF) but may not have sequenced genomes.</p>
<h3 id="automated-maintenance-workflows">Automated Maintenance Workflows</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Core extension scripts (run periodically)</span>
make<span class="w"> </span>update-genomes<span class="w">                 </span><span class="c1"># NCBI Assembly search</span>
src/extend_from_lanm.py<span class="w">             </span><span class="c1"># UniProt protein search (majority method)</span>
src/add_missing_organisms.py<span class="w">        </span><span class="c1"># Validation fixes</span>
src/annotate_kg_identifiers.py<span class="w">     </span><span class="c1"># Adds KG linkage</span>
src/add_annotation_urls.py<span class="w">         </span><span class="c1"># Generates download URLs</span>
</code></pre></div>
<h3 id="deduplication-strategy">Deduplication Strategy</h3>
<p>Current approach:
<div class="highlight"><pre><span></span><code><span class="n">df</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Scientific name&quot;</span><span class="p">],</span> <span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)</span>
</code></pre></div></p>
<p><strong>Gap:</strong> Doesn't catch duplicate taxon IDs with different names (NCBITaxon:270351, 157278)</p>
<p><strong>Proposed:</strong>
<div class="highlight"><pre><span></span><code><span class="c1"># Check for duplicate taxon IDs</span>
<span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;NCBITaxon id&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
</code></pre></div></p>
<hr />
<h2 id="kg_node_ids-column-structure">kg_node_ids Column Structure</h2>
<h3 id="format-specification">Format Specification</h3>
<div class="highlight"><pre><span></span><code>NCBITaxon:270351|kg-microbe; NCBITaxon:270351|kg-microbe-function
</code></pre></div>
<p><strong>Components:</strong>
- <code>NCBITaxon:XXXXX</code> - Organism's NCBI Taxonomy ID
- <code>|</code> - Pipe delimiter separates ID from KG name
- <code>kg-microbe</code> - Phenotypic knowledge graph (BacDive-derived: growth conditions, traits)
- <code>;</code> - Semicolon-space delimiter between multiple KGs
- <code>kg-microbe-function</code> - Functional knowledge graph (UniProt-derived: proteins, pathways)</p>
<h3 id="coverage-pattern">Coverage Pattern</h3>
<ul>
<li><strong>All 180</strong> organisms with KG data are in <strong>BOTH</strong> graphs (100%)</li>
<li><strong>0</strong> organisms in only kg-microbe</li>
<li><strong>0</strong> organisms in only kg-microbe-function</li>
<li>Script: <code>src/annotate_kg_identifiers.py</code> queries both KGs simultaneously</li>
</ul>
<p><strong>What this means:</strong>
- If an organism has <code>kg_node_ids</code>, you can query BOTH phenotypes AND protein functions
- Missing <code>kg_node_ids</code> (31 organisms) = too new for KG databases (no experimental data yet)</p>
<hr />
<h2 id="data-validation-results">Data Validation Results</h2>
<h3 id="no-mismatches-found">✓ No Mismatches Found</h3>
<ol>
<li><strong>Genome ID ↔ Download URL:</strong> All genome IDs match their download URLs ✓</li>
<li><strong>Taxon ID ↔ KG Node IDs:</strong> All taxon IDs match their KG node entries ✓</li>
<li><strong>Data Structure:</strong> Internally consistent ✓</li>
</ol>
<h3 id="incomplete-records-19-rows">Incomplete Records (19 rows)</h3>
<p><strong>Pattern:</strong> Have taxon ID but no genome ID</p>
<p><strong>Organisms affected:</strong>
- Methylobacterium aquaticum Strain 22A (no taxon OR genome)
- Methylobacterium aquaticum CCM:7218... (no genome)
- Bradyrhizobium sp. SM-2017B (no genome)
- Bradyrhizobium sp. MAFF 211645 (no genome)
- Rhizobium sp. SM-2017B (no genome)
- ... (14 more)</p>
<p><strong>Likely reasons:</strong>
- Organisms known to NCBI but no sequenced genome yet
- Culture collection strains not yet deposited in GenBank
- Pending genome assembly submissions</p>
<hr />
<h2 id="recommendations">Recommendations</h2>
<h3 id="quick-wins-can-implement-immediately">Quick Wins (Can implement immediately)</h3>
<ol>
<li><strong>Clean source field pollution</strong></li>
<li>Remove 63 FTP URLs from <code>source</code> column</li>
<li>Add proper provenance labels (e.g., "extend1")</li>
<li>
<p>Script: Simple find/replace based on pattern matching</p>
</li>
<li>
<p><strong>Merge NCBITaxon:157278 duplicate</strong></p>
</li>
<li>Same genome (GCA_965531505.1)</li>
<li>Typo in organism name (sp. vs sp)</li>
<li>
<p>Keep row with period: "uncultured Methylobacterium sp."</p>
</li>
<li>
<p><strong>Standardize "sp." formatting</strong></p>
</li>
<li>Global find/replace: " sp " → " sp. "</li>
<li>
<p>Ensure space before and period after</p>
</li>
<li>
<p><strong>Add source labels to 12 blank rows</strong></p>
</li>
<li>Likely manual additions</li>
<li>Label as "manual_addition_2025-11"</li>
</ol>
<h3 id="requires-investigation">Requires Investigation</h3>
<ol>
<li><strong>NCBITaxon:270351 duplication</strong></li>
<li>Are these different strains of the same species?</li>
<li>Should we keep both or merge with strain info in parentheses?</li>
<li>
<p>Decision needed from domain expert</p>
</li>
<li>
<p><strong>19 incomplete records</strong></p>
</li>
<li>Search NCBI for newly available genomes</li>
<li>Contact culture collections for deposition status</li>
<li>
<p>Consider removing if permanently unavailable</p>
</li>
<li>
<p><strong>Validate FTP URLs</strong></p>
</li>
<li>Check if all 192 download links are still live</li>
<li>NCBI periodically reorganizes FTP structure</li>
<li>Script: <code>src/test_annotation_urls.py</code> exists but needs run</li>
</ol>
<h3 id="long-term-improvements">Long-term Improvements</h3>
<ol>
<li><strong>Automate Google Sheet sync</strong></li>
<li>Current: Manual export/import</li>
<li>Proposed: Scheduled sync or webhook-based updates</li>
<li>
<p>Prevents drift between Google Sheet and git</p>
</li>
<li>
<p><strong>Enhanced deduplication</strong></p>
</li>
<li>Check taxon IDs, not just names</li>
<li>Flag potential duplicates for manual review</li>
<li>
<p>Add to <code>validate_consistency.py</code></p>
</li>
<li>
<p><strong>Source field schema enforcement</strong></p>
</li>
<li>Validate that <code>source</code> contains only provenance labels</li>
<li>Reject FTP URLs at write time</li>
<li>
<p>Add to <code>validate_consistency.py</code></p>
</li>
<li>
<p><strong>Complete KG coverage</strong></p>
</li>
<li>31 organisms lack KG linkage</li>
<li>Monitor kg-microbe updates for new taxa</li>
<li>Re-run <code>annotate_kg_identifiers.py</code> periodically</li>
</ol>
<hr />
<h2 id="action-items">Action Items</h2>
<h3 id="high-priority">High Priority</h3>
<ul>
<li>[ ] Clean 63 URLs from source field</li>
<li>[ ] Merge NCBITaxon:157278 duplicate</li>
<li>[ ] Decide on NCBITaxon:270351 (strain vs species)</li>
</ul>
<h3 id="medium-priority">Medium Priority</h3>
<ul>
<li>[ ] Add source labels to 12 blank rows</li>
<li>[ ] Standardize "sp." formatting</li>
<li>[ ] Search for genomes for 19 incomplete records</li>
</ul>
<h3 id="low-priority">Low Priority</h3>
<ul>
<li>[ ] Validate all 192 FTP URLs</li>
<li>[ ] Document 5 data sources in schema/README.md</li>
<li>[ ] Enhance deduplication checks</li>
<li>[ ] Automate Google Sheet sync</li>
</ul>
<hr />
<h2 id="files-generated">Files Generated</h2>
<p>Analysis scripts used:
- <code>/tmp/summary_stats.py</code> - Column statistics
- <code>/tmp/check_mismatches.py</code> - Data validation
- <code>/tmp/investigate_data.py</code> - Provenance analysis
- <code>/tmp/enhanced_table.py</code> - Final table generation</p>
<p>Repository scripts relevant to this sheet:
- <code>src/annotate_kg_identifiers.py</code> - Populates kg_node_ids
- <code>src/add_annotation_urls.py</code> - Populates Annotation download URL
- <code>src/add_missing_organisms.py</code> - Adds organisms from genes table
- <code>src/extend_from_lanm.py</code> - UniProt protein search (majority method)
- <code>src/ncbi_search.py</code> - NCBI Assembly search
- <code>src/validate_consistency.py</code> - Cross-sheet validation</p>
<hr />
<p><strong>Generated:</strong> 2025-11-13<br />
<strong>Analyst:</strong> Data quality assessment via Claude Code<br />
<strong>Next Review:</strong> After implementing Quick Wins</p>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>