
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/CMM-AI_Analysis/08_appendices/">
      
      
      
      
        
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Chapter 8: Appendices - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#chapter-8-appendices" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Chapter 8: Appendices
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#appendix-a-file-inventory" class="md-nav__link">
    <span class="md-ellipsis">
      
        Appendix A: File Inventory
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Appendix A: File Inventory">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#original-excel-file" class="md-nav__link">
    <span class="md-ellipsis">
      
        Original Excel File
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#converted-tsv-files-from-google-sheet" class="md-nav__link">
    <span class="md-ellipsis">
      
        Converted TSV Files (from Google Sheet)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extended-tsv-files-from-python-scripts" class="md-nav__link">
    <span class="md-ellipsis">
      
        Extended TSV Files (from Python scripts)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#appendix-b-key-file-locations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Appendix B: Key File Locations
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Appendix B: Key File Locations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#python-scripts-for-media" class="md-nav__link">
    <span class="md-ellipsis">
      
        Python Scripts for Media
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#conversion-scripts" class="md-nav__link">
    <span class="md-ellipsis">
      
        Conversion Scripts
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#configuration" class="md-nav__link">
    <span class="md-ellipsis">
      
        Configuration
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      
        Schema
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#appendix-c-data-creation-methods" class="md-nav__link">
    <span class="md-ellipsis">
      
        Appendix C: Data Creation Methods
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Appendix C: Data Creation Methods">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-manual-entry-in-google-sheets" class="md-nav__link">
    <span class="md-ellipsis">
      
        Method 1: Manual Entry in Google Sheets
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-xlsx-download-and-conversion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Method 2: XLSX Download and Conversion
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-python-extend-scripts-deterministic" class="md-nav__link">
    <span class="md-ellipsis">
      
        Method 3: Python Extend Scripts (Deterministic)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-hardcoded-data-in-python" class="md-nav__link">
    <span class="md-ellipsis">
      
        Method 4: Hardcoded Data in Python
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-5-kg-microbe-queries" class="md-nav__link">
    <span class="md-ellipsis">
      
        Method 5: KG-Microbe Queries
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-6-llm-generated-content-suspected" class="md-nav__link">
    <span class="md-ellipsis">
      
        Method 6: LLM-Generated Content (Suspected)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#appendix-d-proposed-prefix-registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        Appendix D: Proposed Prefix Registry
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Appendix D: Proposed Prefix Registry">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#recommendation-create-a-prefix-registry-sheet" class="md-nav__link">
    <span class="md-ellipsis">
      
        Recommendation: Create a Prefix Registry Sheet
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#proposed-schema" class="md-nav__link">
    <span class="md-ellipsis">
      
        Proposed Schema
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#proposed-content" class="md-nav__link">
    <span class="md-ellipsis">
      
        Proposed Content
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#linkml-schema-addition" class="md-nav__link">
    <span class="md-ellipsis">
      
        LinkML Schema Addition
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#benefits-of-a-prefix-registry" class="md-nav__link">
    <span class="md-ellipsis">
      
        Benefits of a Prefix Registry
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#appendix-e-chebi-id-overlap-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Appendix E: CHEBI ID Overlap Analysis
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#appendix-f-schema-validation-features" class="md-nav__link">
    <span class="md-ellipsis">
      
        Appendix F: Schema Validation Features
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Appendix F: Schema Validation Features">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#whats-missing-from-the-schema" class="md-nav__link">
    <span class="md-ellipsis">
      
        What's Missing from the Schema
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="chapter-8-appendices">Chapter 8: Appendices</h1>
<p><a href="../00_index/">&lt;- Back to Index</a> | <a href="../07_recommendations/">Previous: Recommendations</a></p>
<hr />
<h2 id="appendix-a-file-inventory">Appendix A: File Inventory</h2>
<h3 id="original-excel-file">Original Excel File</h3>
<div class="highlight"><pre><span></span><code>data/sheet/BER CMM Data for AI.xlsx (254 KB, Dec 4 2025)
</code></pre></div>
<h3 id="converted-tsv-files-from-google-sheet">Converted TSV Files (from Google Sheet)</h3>
<div class="highlight"><pre><span></span><code>data/txt/sheet/
+-- BER_CMM_Data_for_AI_assays.tsv              (5 KB)
+-- BER_CMM_Data_for_AI_bioprocesses.tsv        (2 KB)
+-- BER_CMM_Data_for_AI_biosamples.tsv          (12 KB)
+-- BER_CMM_Data_for_AI_chemicals.tsv           (15 KB)
+-- BER_CMM_Data_for_AI_datasets.tsv            (4 KB)
+-- BER_CMM_Data_for_AI_genes_and_proteins.tsv  (73 KB)
+-- BER_CMM_Data_for_AI_growth_media.tsv        (3 KB)
+-- BER_CMM_Data_for_AI_macromolecular_structures.tsv (3 KB)
+-- BER_CMM_Data_for_AI_media_ingredients.tsv   (12 KB)
+-- BER_CMM_Data_for_AI_pathways.tsv            (15 KB)
+-- BER_CMM_Data_for_AI_protocols.tsv           (7 KB)
+-- BER_CMM_Data_for_AI_publications.tsv        (26 KB)
+-- BER_CMM_Data_for_AI_screening_results.tsv   (1 KB)
+-- BER_CMM_Data_for_AI_strains.tsv             (13 KB)
+-- BER_CMM_Data_for_AI_taxa_and_genomes.tsv    (60 KB)
+-- BER_CMM_Data_for_AI_transcriptomics.tsv     (10 KB)
</code></pre></div>
<h3 id="extended-tsv-files-from-python-scripts">Extended TSV Files (from Python scripts)</h3>
<div class="highlight"><pre><span></span><code>data/txt/sheet/extended/
+-- BER_CMM_Data_for_AI_assays_extended.tsv              (7 KB)
+-- BER_CMM_Data_for_AI_bioprocesses_extended.tsv        (4 KB)
+-- BER_CMM_Data_for_AI_biosamples_extended.tsv          (8 KB)
+-- BER_CMM_Data_for_AI_chemicals_extended.tsv           (15 KB)
+-- BER_CMM_Data_for_AI_datasets_extended.tsv            (3 KB)
+-- BER_CMM_Data_for_AI_genes_and_proteins_extended.tsv  (72 KB)
+-- BER_CMM_Data_for_AI_growth_media_extended.tsv        (2 KB)
+-- BER_CMM_Data_for_AI_macromolecular_structures_extended.tsv (2 KB)
+-- BER_CMM_Data_for_AI_media_ingredients_extended.tsv   (9 KB)
+-- BER_CMM_Data_for_AI_pathways_extended.tsv            (13 KB)
+-- BER_CMM_Data_for_AI_protocols_extended.tsv           (2 KB)
+-- BER_CMM_Data_for_AI_publications_extended.tsv        (21 KB)
+-- BER_CMM_Data_for_AI_screening_results_extended.tsv   (1 KB)
+-- BER_CMM_Data_for_AI_strains_extended.tsv             (13 KB)
+-- BER_CMM_Data_for_AI_taxa_and_genomes_extended.tsv    (61 KB)
+-- BER_CMM_Data_for_AI_transcriptomics_extended.tsv     (10 KB)
</code></pre></div>
<hr />
<h2 id="appendix-b-key-file-locations">Appendix B: Key File Locations</h2>
<h3 id="python-scripts-for-media">Python Scripts for Media</h3>
<ul>
<li><code>src/media_search.py</code> - Contains <code>CURATED_MEDIA</code> dict and KG queries</li>
<li><code>src/extend_media.py</code> - Wrapper script that calls media_search</li>
</ul>
<h3 id="conversion-scripts">Conversion Scripts</h3>
<ul>
<li><code>src/convert_sheets.py</code> - XLSX -&gt; TSV conversion</li>
<li><code>src/parsers.py</code> - General file parsing utilities</li>
</ul>
<h3 id="configuration">Configuration</h3>
<ul>
<li><code>Makefile</code> - Pipeline automation (974 lines)</li>
<li><code>CLAUDE.md</code> - AI assistant guidance (29 KB)</li>
<li><code>.mcp.json</code> - Google Drive MCP server config</li>
</ul>
<h3 id="schema">Schema</h3>
<ul>
<li><code>schema/lanthanide_bioprocessing.yaml</code> - LinkML schema</li>
<li><code>src/linkml_models.py</code> - Generated Python models (auto-generated)</li>
</ul>
<hr />
<h2 id="appendix-c-data-creation-methods">Appendix C: Data Creation Methods</h2>
<h3 id="method-1-manual-entry-in-google-sheets">Method 1: Manual Entry in Google Sheets</h3>
<p><strong>Who:</strong> Experimental collaborators
<strong>What:</strong> Typing or pasting data directly
<strong>When:</strong> Unknown (no changelog)
<strong>Tracking:</strong> None</p>
<h3 id="method-2-xlsx-download-and-conversion">Method 2: XLSX Download and Conversion</h3>
<p><strong>Who:</strong> Marcin
<strong>What:</strong> Download Google Sheet as XLSX, commit to repo
<strong>Command:</strong> <code>make convert-excel</code>
<strong>Script:</strong> <code>src/convert_sheets.py</code>
<strong>Output:</strong> <code>data/txt/sheet/*.tsv</code> (overwrites previous)</p>
<h3 id="method-3-python-extend-scripts-deterministic">Method 3: Python Extend Scripts (Deterministic)</h3>
<p><strong>Commands:</strong>
<div class="highlight"><pre><span></span><code>make<span class="w"> </span>update-media<span class="w">      </span><span class="c1"># -&gt; extend_media.py -&gt; media_search.py</span>
make<span class="w"> </span>update-genomes<span class="w">    </span><span class="c1"># -&gt; extend_lanthanide_data.py -&gt; ncbi_search.py</span>
make<span class="w"> </span>update-pathways<span class="w">   </span><span class="c1"># -&gt; extend_pathways.py -&gt; pathway_search.py</span>
make<span class="w"> </span>update-genes<span class="w">      </span><span class="c1"># -&gt; extend_genes.py -&gt; gene_search.py</span>
make<span class="w"> </span>update-structures<span class="w"> </span><span class="c1"># -&gt; extend_structures.py -&gt; structure_search.py</span>
make<span class="w"> </span>update-publications<span class="w"> </span><span class="c1"># -&gt; extend_publications.py -&gt; publication_search.py</span>
make<span class="w"> </span>update-datasets<span class="w">   </span><span class="c1"># -&gt; extend_datasets.py -&gt; dataset_search.py</span>
</code></pre></div></p>
<p><strong>Output:</strong> <code>data/txt/sheet/extended/*_extended.tsv</code></p>
<h3 id="method-4-hardcoded-data-in-python">Method 4: Hardcoded Data in Python</h3>
<p><strong>Example:</strong> <code>src/media_search.py</code> lines 135-274</p>
<div class="highlight"><pre><span></span><code><span class="n">CURATED_MEDIA</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;ATCC:1306&quot;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;media_name&quot;</span><span class="p">:</span> <span class="s2">&quot;ATCC Medium 1306 (Methanol mineral salts)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ingredients&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;Methanol&quot;</span><span class="p">,</span> <span class="s2">&quot;formula&quot;</span><span class="p">:</span> <span class="s2">&quot;CH3OH&quot;</span><span class="p">,</span> <span class="s2">&quot;chebi&quot;</span><span class="p">:</span> <span class="s2">&quot;CHEBI:17790&quot;</span><span class="p">},</span>
            <span class="o">...</span>
        <span class="p">]</span>
    <span class="p">},</span>
    <span class="c1"># 7 media hardcoded here</span>
<span class="p">}</span>
</code></pre></div>
<p>This is <strong>not data entry</strong> - it's code. Changes require editing Python files.</p>
<h3 id="method-5-kg-microbe-queries">Method 5: KG-Microbe Queries</h3>
<p><strong>Script:</strong> <code>src/media_search.py</code> functions:
- <code>query_kg_microbe_for_ingredient()</code> - lines 24-85
- <code>query_kg_microbe_for_medium()</code> - lines 88-131</p>
<p><strong>Database:</strong> <code>data/kgm/kg-microbe.duckdb</code>
<strong>Note:</strong> Contains SQL injection vulnerability (f-string interpolation)</p>
<h3 id="method-6-llm-generated-content-suspected">Method 6: LLM-Generated Content (Suspected)</h3>
<p><strong>Evidence:</strong> Some descriptions appear LLM-generated (verbose, structured)
<strong>Example:</strong> MP medium description in Google Sheet
<strong>Tracking:</strong> None - no way to distinguish from manual entry</p>
<hr />
<h2 id="appendix-d-proposed-prefix-registry">Appendix D: Proposed Prefix Registry</h2>
<h3 id="recommendation-create-a-prefix-registry-sheet">Recommendation: Create a Prefix Registry Sheet</h3>
<p>A new TSV file (<code>BER_CMM_Data_for_AI_prefixes.tsv</code>) should define all prefixes used in the data:</p>
<h3 id="proposed-schema">Proposed Schema</h3>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>prefix_id</code></td>
<td>Short prefix string (e.g., <code>CHEBI</code>, <code>ingredient</code>)</td>
</tr>
<tr>
<td><code>prefix_name</code></td>
<td>Human-readable name</td>
</tr>
<tr>
<td><code>uri_pattern</code></td>
<td>URI expansion pattern (if resolvable)</td>
</tr>
<tr>
<td><code>example_id</code></td>
<td>Example local ID</td>
</tr>
<tr>
<td><code>example_uri</code></td>
<td>Resolved example URI</td>
</tr>
<tr>
<td><code>resolver</code></td>
<td>Where to resolve (bioregistry.io, direct URL, internal DB)</td>
</tr>
<tr>
<td><code>source</code></td>
<td>Authority (OBO, DSMZ, LBL internal, etc.)</td>
</tr>
<tr>
<td><code>notes</code></td>
<td>Usage notes</td>
</tr>
</tbody>
</table>
<h3 id="proposed-content">Proposed Content</h3>
<table>
<thead>
<tr>
<th>prefix_id</th>
<th>prefix_name</th>
<th>uri_pattern</th>
<th>example_id</th>
<th>resolver</th>
<th>source</th>
</tr>
</thead>
<tbody>
<tr>
<td>CHEBI</td>
<td>ChEBI</td>
<td>http://purl.obolibrary.org/obo/CHEBI_</td>
<td>17790</td>
<td>bioregistry.io</td>
<td>OBO</td>
</tr>
<tr>
<td>GO</td>
<td>Gene Ontology</td>
<td>http://purl.obolibrary.org/obo/GO_</td>
<td>0008150</td>
<td>bioregistry.io</td>
<td>OBO</td>
</tr>
<tr>
<td>EC</td>
<td>Enzyme Commission</td>
<td>https://www.enzyme-database.org/query.php?ec=</td>
<td>1.1.2.7</td>
<td>bioregistry.io</td>
<td>ExPASy</td>
</tr>
<tr>
<td>NCBITaxon</td>
<td>NCBI Taxonomy</td>
<td>http://purl.obolibrary.org/obo/NCBITaxon_</td>
<td>270351</td>
<td>bioregistry.io</td>
<td>NCBI</td>
</tr>
<tr>
<td>UniProtKB</td>
<td>UniProt</td>
<td>http://purl.uniprot.org/uniprot/</td>
<td>P12345</td>
<td>bioregistry.io</td>
<td>UniProt</td>
</tr>
<tr>
<td>PDB</td>
<td>Protein Data Bank</td>
<td>https://www.rcsb.org/structure/</td>
<td>1ABC</td>
<td>bioregistry.io</td>
<td>RCSB</td>
</tr>
<tr>
<td>KEGG</td>
<td>KEGG</td>
<td>https://www.kegg.jp/entry/</td>
<td>K00001</td>
<td>bioregistry.io</td>
<td>KEGG</td>
</tr>
<tr>
<td>PMID</td>
<td>PubMed</td>
<td>http://www.ncbi.nlm.nih.gov/pubmed/</td>
<td>36719530</td>
<td>bioregistry.io</td>
<td>NCBI</td>
</tr>
<tr>
<td>DOI</td>
<td>Digital Object Identifier</td>
<td>https://doi.org/</td>
<td>10.1038/s41586-024-07070-8</td>
<td>doi.org</td>
<td>DOI Foundation</td>
</tr>
<tr>
<td>CAS-RN</td>
<td>CAS Registry Number</td>
<td>(not directly resolvable)</td>
<td>5625-37-6</td>
<td>commonchemistry.cas.org</td>
<td>CAS</td>
</tr>
<tr>
<td>PubChem</td>
<td>PubChem Compound</td>
<td>https://pubchem.ncbi.nlm.nih.gov/compound/</td>
<td>12345</td>
<td>bioregistry.io</td>
<td>NCBI</td>
</tr>
<tr>
<td>mediadive.medium</td>
<td>MediaDive Medium</td>
<td>https://mediadive.dsmz.de/medium/</td>
<td>77</td>
<td>bioregistry.io</td>
<td>DSMZ</td>
</tr>
<tr>
<td>mediadive.solution</td>
<td>MediaDive Solution</td>
<td>https://mediadive.dsmz.de/solution/</td>
<td>S6</td>
<td>bioregistry.io</td>
<td>DSMZ</td>
</tr>
<tr>
<td>ingredient</td>
<td>KG-Microbe Ingredient</td>
<td>(internal)</td>
<td>1633</td>
<td>kg-microbe.duckdb</td>
<td>LBL</td>
</tr>
<tr>
<td>solution</td>
<td>KG-Microbe Solution</td>
<td>(internal)</td>
<td>4592</td>
<td>kg-microbe.duckdb</td>
<td>LBL</td>
</tr>
<tr>
<td>medium</td>
<td>KG-Microbe Medium</td>
<td>(internal)</td>
<td>J562</td>
<td>kg-microbe.duckdb</td>
<td>LBL</td>
</tr>
<tr>
<td>ATCC</td>
<td>ATCC Culture Collection</td>
<td>https://www.atcc.org/products/</td>
<td>BAA-1231</td>
<td>direct</td>
<td>ATCC</td>
</tr>
<tr>
<td>DSMZ</td>
<td>DSMZ Culture Collection</td>
<td>https://www.dsmz.de/collection/catalogue/</td>
<td>88</td>
<td>direct</td>
<td>DSMZ</td>
</tr>
</tbody>
</table>
<h3 id="linkml-schema-addition">LinkML Schema Addition</h3>
<p>The schema should also be updated to include these prefixes:</p>
<div class="highlight"><pre><span></span><code><span class="nt">prefixes</span><span class="p">:</span>
<span class="w">  </span><span class="c1"># === External Ontologies (OBO Foundry) ===</span>
<span class="w">  </span><span class="nt">CHEBI</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://purl.obolibrary.org/obo/CHEBI_</span>
<span class="w">  </span><span class="nt">GO</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://purl.obolibrary.org/obo/GO_</span>
<span class="w">  </span><span class="nt">NCBITaxon</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://purl.obolibrary.org/obo/NCBITaxon_</span>
<span class="w">  </span><span class="nt">OBI</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://purl.obolibrary.org/obo/OBI_</span>
<span class="w">  </span><span class="nt">ENVO</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://purl.obolibrary.org/obo/ENVO_</span>

<span class="w">  </span><span class="c1"># === External Databases ===</span>
<span class="w">  </span><span class="nt">UniProtKB</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://purl.uniprot.org/uniprot/</span>
<span class="w">  </span><span class="nt">PDB</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.rcsb.org/structure/</span>
<span class="w">  </span><span class="nt">KEGG</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.kegg.jp/entry/</span>
<span class="w">  </span><span class="nt">EC</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.enzyme-database.org/query.php?ec=</span>
<span class="w">  </span><span class="nt">RHEA</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.rhea-db.org/rhea/</span>
<span class="w">  </span><span class="nt">PubChem</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://pubchem.ncbi.nlm.nih.gov/compound/</span>
<span class="w">  </span><span class="nt">ChEMBL</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.ebi.ac.uk/chembl/compound_report_card/</span>

<span class="w">  </span><span class="c1"># === Literature ===</span>
<span class="w">  </span><span class="nt">DOI</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://doi.org/</span>
<span class="w">  </span><span class="nt">PMID</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">http://www.ncbi.nlm.nih.gov/pubmed/</span>

<span class="w">  </span><span class="c1"># === MediaDive (DSMZ) ===</span>
<span class="w">  </span><span class="nt">mediadive.medium</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://mediadive.dsmz.de/medium/</span>
<span class="w">  </span><span class="nt">mediadive.solution</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://mediadive.dsmz.de/solution/</span>

<span class="w">  </span><span class="c1"># === KG-Microbe Internal (LBL) ===</span>
<span class="w">  </span><span class="c1"># Note: These are internal IDs, not resolvable URIs</span>
<span class="w">  </span><span class="c1"># Query kg-microbe.duckdb for actual data</span>
<span class="w">  </span><span class="nt">kgm.ingredient</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tag:kg-microbe.lbl.gov,2024:ingredient/</span>
<span class="w">  </span><span class="nt">kgm.solution</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tag:kg-microbe.lbl.gov,2024:solution/</span>
<span class="w">  </span><span class="nt">kgm.medium</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">tag:kg-microbe.lbl.gov,2024:medium/</span>

<span class="w">  </span><span class="c1"># === Culture Collections ===</span>
<span class="w">  </span><span class="nt">ATCC</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.atcc.org/products/</span>
<span class="w">  </span><span class="nt">DSMZ</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">https://www.dsmz.de/collection/catalogue/details/culture/DSM-</span>
</code></pre></div>
<h3 id="benefits-of-a-prefix-registry">Benefits of a Prefix Registry</h3>
<ol>
<li><strong>Self-documenting</strong>: Users can understand what <code>ingredient:1633</code> means</li>
<li><strong>Validation</strong>: Can check that IDs use registered prefixes</li>
<li><strong>Resolution</strong>: Know where to look up each ID type</li>
<li><strong>Interoperability</strong>: Standard prefix -&gt; URI mappings for RDF export</li>
<li><strong>KG Integration</strong>: Clear mapping between CMM-AI IDs and KG-Microbe/MediaDive</li>
</ol>
<hr />
<h2 id="appendix-e-chebi-id-overlap-analysis">Appendix E: CHEBI ID Overlap Analysis</h2>
<div class="highlight"><pre><span></span><code>media_ingredients.ontology_id values with CHEBI: ~48 unique IDs
chemicals.chebi_id values: ~30 unique IDs

Overlap examples:
- CHEBI:17790 (Methanol) - in both tables
- CHEBI:17234 (Glucose) - in both tables
- CHEBI:32599 (MgSO4*7H2O) - only in media_ingredients
</code></pre></div>
<p>The <strong>chemicals</strong> table focuses on <strong>REE-related compounds</strong> (lanthanides, lanthanophores), while <strong>media_ingredients</strong> has <strong>common lab chemicals</strong> (salts, buffers). They serve different purposes but share the CHEBI namespace.</p>
<hr />
<h2 id="appendix-f-schema-validation-features">Appendix F: Schema Validation Features</h2>
<p>The LinkML schema enforces:</p>
<ol>
<li><strong>Required fields</strong>: <code>scientific_name</code>, <code>sample_id</code>, <code>ingredient_id</code>, etc.</li>
<li><strong>Identifier patterns</strong>: <code>ingredient_id</code> must match <code>^[A-Z0-9_:]+_ING_\\d+$</code></li>
<li><strong>Enumerated values</strong>: <code>role</code> must be from <code>ingredient_role_enum</code></li>
<li><strong>Data types</strong>: <code>concentration</code> is <code>float</code>, <code>ncbi_taxon_id</code> is <code>integer</code></li>
<li><strong>Foreign keys</strong>: <code>media_id</code> described as "Foreign key to GrowthMediaRecord" (not enforced programmatically)</li>
</ol>
<h3 id="whats-missing-from-the-schema">What's Missing from the Schema</h3>
<p>The schema's <code>prefixes:</code> block (lines 16-41) only covers common ontology prefixes.</p>
<p><strong>NOT included</strong> (but appear in data):
- <code>ingredient:</code> - KG-Microbe internal
- <code>solution:</code> - KG-Microbe internal
- <code>medium:</code> - KG-Microbe internal
- <code>mediadive.medium:</code> - DSMZ MediaDive
- <code>mediadive.solution:</code> - DSMZ MediaDive
- <code>CAS-RN:</code> - Chemical Abstracts Service</p>
<hr />
<p><em>Report generated: 2025-12-05</em>
<em>Split into chapters from original 2,323-line file to improve navigation and reduce redundancy.</em></p>
<hr />
<p><a href="../00_index/">&lt;- Back to Index</a> | <a href="../07_recommendations/">Previous: Recommendations</a></p>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>