
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/elements/">
      
      
        <link rel="prev" href="../best-practices/">
      
      
        <link rel="next" href="../architecture/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Schema - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cmm-ai-automation-schema" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Schema
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Classes
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#slots" class="md-nav__link">
    <span class="md-ellipsis">
      
        Slots
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#enumerations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Enumerations
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#types" class="md-nav__link">
    <span class="md-ellipsis">
      
        Types
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsets" class="md-nav__link">
    <span class="md-ellipsis">
      
        Subsets
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Classes
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#slots" class="md-nav__link">
    <span class="md-ellipsis">
      
        Slots
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#enumerations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Enumerations
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#types" class="md-nav__link">
    <span class="md-ellipsis">
      
        Types
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsets" class="md-nav__link">
    <span class="md-ellipsis">
      
        Subsets
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="cmm-ai-automation-schema">CMM AI Automation Schema</h1>
<p>Schema for Critical Mineral Metabolism (CMM) data curation.</p>
<p>Models microbial strains, growth media, ingredients, and their relationships
for knowledge graph generation.</p>
<p>URI: https://w3id.org/turbomam/cmm-ai-automation</p>
<p>Name: cmm-ai-automation</p>
<h2 id="classes">Classes</h2>
<table>
<thead>
<tr>
<th>Class</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="CMMDatabase/">CMMDatabase</a></td>
<td>Container for all CMM entities</td>
</tr>
<tr>
<td><a href="CrossReference/">CrossReference</a></td>
<td>A cross-reference to an external database or identifier</td>
</tr>
<tr>
<td><a href="DataConflict/">DataConflict</a></td>
<td>Records conflicts between data sources for reconciliation review</td>
</tr>
<tr>
<td><a href="EnrichedIngredient/">EnrichedIngredient</a></td>
<td>A chemical entity enriched with data from multiple sources</td>
</tr>
<tr>
<td><a href="Genome/">Genome</a></td>
<td>A genome assembly from NCBI GenBank/RefSeq or other sources</td>
</tr>
<tr>
<td><a href="GrowthPreference/">GrowthPreference</a></td>
<td>Reified relationship: a strain's growth in a specific medium</td>
</tr>
<tr>
<td><a href="IngredientComponent/">IngredientComponent</a></td>
<td>Reified relationship: an ingredient as used in a mixture, with concentration ...</td>
</tr>
<tr>
<td><a href="NamedThing/">NamedThing</a></td>
<td>A generic grouping for any identifiable entity</td>
</tr>
<tr>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="Ingredient/">Ingredient</a></td>
<td>A chemical entity that can be a component of solutions or media</td>
</tr>
<tr>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="Mixture/">Mixture</a></td>
<td>Abstract base class for things composed of ingredients</td>
</tr>
<tr>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="GrowthMedium/">GrowthMedium</a></td>
<td>A complete formulation for cultivating microorganisms</td>
</tr>
<tr>
<td>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<a href="Solution/">Solution</a></td>
<td>A pre-made concentrated mixture of ingredients, typically diluted into media</td>
</tr>
<tr>
<td><a href="SolutionComponent/">SolutionComponent</a></td>
<td>Reified relationship: a solution as used in a medium, with volume/dilution</td>
</tr>
<tr>
<td><a href="SourceRecord/">SourceRecord</a></td>
<td>Tracks the provenance of data from a specific source</td>
</tr>
<tr>
<td><a href="Strain/">Strain</a></td>
<td>A microbial strain enriched with data from multiple sources</td>
</tr>
<tr>
<td><a href="Taxon/">Taxon</a></td>
<td>A taxonomic entity from NCBI Taxonomy</td>
</tr>
</tbody>
</table>
<h2 id="slots">Slots</h2>
<table>
<thead>
<tr>
<th>Slot</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="annotation_url/">annotation_url</a></td>
<td>URL to download genome annotation file (GFF format)</td>
</tr>
<tr>
<td><a href="application_roles/">application_roles</a></td>
<td>Application roles from ChEBI</td>
</tr>
<tr>
<td><a href="assembly_level/">assembly_level</a></td>
<td>Level of assembly completeness</td>
</tr>
<tr>
<td><a href="assembly_name/">assembly_name</a></td>
<td>Human-readable assembly name assigned by submitter</td>
</tr>
<tr>
<td><a href="atcc_id/">atcc_id</a></td>
<td>ATCC catalog number (ATCC is authoritative)</td>
</tr>
<tr>
<td><a href="bacdive_id/">bacdive_id</a></td>
<td>BacDive strain ID (BacDive is authoritative for strain metadata)</td>
</tr>
<tr>
<td><a href="belongs_to_taxon/">belongs_to_taxon</a></td>
<td>Links a strain to its taxonomic classification</td>
</tr>
<tr>
<td><a href="biological_roles/">biological_roles</a></td>
<td>Biological roles from ChEBI (ChEBI is authoritative)</td>
</tr>
<tr>
<td><a href="biosafety_level/">biosafety_level</a></td>
<td>Biosafety level (1-4)</td>
</tr>
<tr>
<td><a href="cas_rn/">cas_rn</a></td>
<td>CAS Registry Number</td>
</tr>
<tr>
<td><a href="cell_shape/">cell_shape</a></td>
<td>Cell morphology (e</td>
</tr>
<tr>
<td><a href="charge/">charge</a></td>
<td>Formal charge of the molecule</td>
</tr>
<tr>
<td><a href="chebi_id/">chebi_id</a></td>
<td>ChEBI identifier</td>
</tr>
<tr>
<td><a href="chemical_formula/">chemical_formula</a></td>
<td>Chemical formula (e</td>
</tr>
<tr>
<td><a href="chemical_roles/">chemical_roles</a></td>
<td>Chemical roles from ChEBI (ChEBI is authoritative)</td>
</tr>
<tr>
<td><a href="cip_id/">cip_id</a></td>
<td>CIP (Collection de l'Institut Pasteur) number</td>
</tr>
<tr>
<td><a href="concentration_unit/">concentration_unit</a></td>
<td>Unit of concentration</td>
</tr>
<tr>
<td><a href="concentration_value/">concentration_value</a></td>
<td>Numeric concentration value</td>
</tr>
<tr>
<td><a href="conflicting_source/">conflicting_source</a></td>
<td>The source with a different value</td>
</tr>
<tr>
<td><a href="conflicting_value/">conflicting_value</a></td>
<td>Value that conflicts with the authoritative source</td>
</tr>
<tr>
<td><a href="conflicts/">conflicts</a></td>
<td>Recorded conflicts between data sources</td>
</tr>
<tr>
<td><a href="derived_from/">derived_from</a></td>
<td>Parent medium this formulation is based on</td>
</tr>
<tr>
<td><a href="description/">description</a></td>
<td>A human-readable description for a thing</td>
</tr>
<tr>
<td><a href="does_not_grow_in_medium/">does_not_grow_in_medium</a></td>
<td>Media in which this strain cannot grow (negative results)</td>
</tr>
<tr>
<td><a href="doubling_time/">doubling_time</a></td>
<td>Quantitative growth rate measured as doubling time in hours</td>
</tr>
<tr>
<td><a href="drugbank_id/">drugbank_id</a></td>
<td>DrugBank identifier</td>
</tr>
<tr>
<td><a href="dsm_id/">dsm_id</a></td>
<td>DSMZ strain number (DSMZ is authoritative)</td>
</tr>
<tr>
<td><a href="field_name/">field_name</a></td>
<td>Name of the field with conflicting values</td>
</tr>
<tr>
<td><a href="ftp_path/">ftp_path</a></td>
<td>Base FTP path for genome files at NCBI</td>
</tr>
<tr>
<td><a href="genbank_accession/">genbank_accession</a></td>
<td>GenBank assembly accession (GCA_NNNNNNNN</td>
</tr>
<tr>
<td><a href="genome_accessions/">genome_accessions</a></td>
<td>Genome assembly accessions (e</td>
</tr>
<tr>
<td><a href="genomes/">genomes</a></td>
<td>Collection of all genome assemblies</td>
</tr>
<tr>
<td><a href="gram_stain/">gram_stain</a></td>
<td>Gram staining result</td>
</tr>
<tr>
<td><a href="grows/">grows</a></td>
<td>True if strain grows in medium, false if it does not grow</td>
</tr>
<tr>
<td><a href="grows_in_medium/">grows_in_medium</a></td>
<td>Media in which this strain can grow</td>
</tr>
<tr>
<td><a href="growth_preferences/">growth_preferences</a></td>
<td>Collection of strain-medium growth relationships with properties</td>
</tr>
<tr>
<td><a href="growth_rate/">growth_rate</a></td>
<td>Qualitative growth rate in this medium</td>
</tr>
<tr>
<td><a href="has_genome/">has_genome</a></td>
<td>Links a strain or taxon to its genome assembly(ies)</td>
</tr>
<tr>
<td><a href="has_ingredient_component/">has_ingredient_component</a></td>
<td>Ingredients contained in this mixture with their concentrations</td>
</tr>
<tr>
<td><a href="has_lanmodulin/">has_lanmodulin</a></td>
<td>Whether strain has lanmodulin (lanM) gene for lanthanide binding</td>
</tr>
<tr>
<td><a href="has_solution_component/">has_solution_component</a></td>
<td>Solutions contained in this medium</td>
</tr>
<tr>
<td><a href="has_taxonomic_rank/">has_taxonomic_rank</a></td>
<td>Taxonomic rank (species, strain, subspecies, etc</td>
</tr>
<tr>
<td><a href="has_xox_genes/">has_xox_genes</a></td>
<td>Whether strain has XoxF, ExaF, or other lanthanide-dependent alcohol dehydrog...</td>
</tr>
<tr>
<td><a href="id/">id</a></td>
<td>A unique identifier for a thing</td>
</tr>
<tr>
<td><a href="inchi/">inchi</a></td>
<td>InChI string for structural identification</td>
</tr>
<tr>
<td><a href="inchikey/">inchikey</a></td>
<td>InChIKey for structural identification</td>
</tr>
<tr>
<td><a href="incubation_time/">incubation_time</a></td>
<td>Duration of incubation (e</td>
</tr>
<tr>
<td><a href="ingredient/">ingredient</a></td>
<td>Reference to the ingredient entity</td>
</tr>
<tr>
<td><a href="ingredients/">ingredients</a></td>
<td>Collection of all chemical ingredients in the database</td>
</tr>
<tr>
<td><a href="isolation_country/">isolation_country</a></td>
<td>Country where the strain was isolated</td>
</tr>
<tr>
<td><a href="isolation_date/">isolation_date</a></td>
<td>Date or year of isolation</td>
</tr>
<tr>
<td><a href="isolation_source/">isolation_source</a></td>
<td>Sample type or environment from which strain was isolated</td>
</tr>
<tr>
<td><a href="jcm_id/">jcm_id</a></td>
<td>JCM (Japan Collection of Microorganisms) number</td>
</tr>
<tr>
<td><a href="kegg_id/">kegg_id</a></td>
<td>KEGG Compound ID (e</td>
</tr>
<tr>
<td><a href="kg_node_ids/">kg_node_ids</a></td>
<td>Node IDs in external knowledge graphs (e</td>
</tr>
<tr>
<td><a href="last_enriched/">last_enriched</a></td>
<td>Timestamp of last enrichment run</td>
</tr>
<tr>
<td><a href="lmg_id/">lmg_id</a></td>
<td>LMG (Belgian Coordinated Collections of Microorganisms) number</td>
</tr>
<tr>
<td><a href="media/">media</a></td>
<td>Collection of all growth medium recipes</td>
</tr>
<tr>
<td><a href="mediadive_id/">mediadive_id</a></td>
<td>MediaDive ingredient ID (MediaDive is authoritative)</td>
</tr>
<tr>
<td><a href="medium_type/">medium_type</a></td>
<td>Classification of the medium</td>
</tr>
<tr>
<td><a href="mesh_id/">mesh_id</a></td>
<td>MeSH identifier</td>
</tr>
<tr>
<td><a href="modifications/">modifications</a></td>
<td>List of modifications from the parent medium (derived_from)</td>
</tr>
<tr>
<td><a href="molecular_mass/">molecular_mass</a></td>
<td>Average molecular mass in Daltons</td>
</tr>
<tr>
<td><a href="monoisotopic_mass/">monoisotopic_mass</a></td>
<td>Monoisotopic mass in Daltons</td>
</tr>
<tr>
<td><a href="motility/">motility</a></td>
<td>Whether the organism is motile</td>
</tr>
<tr>
<td><a href="name/">name</a></td>
<td>A human-readable name for a thing</td>
</tr>
<tr>
<td><a href="nbrc_id/">nbrc_id</a></td>
<td>NBRC (NITE Biological Resource Center, Japan) number</td>
</tr>
<tr>
<td><a href="ncbi_taxon_id/">ncbi_taxon_id</a></td>
<td>NCBITaxon ID at strain level when available</td>
</tr>
<tr>
<td><a href="ncimb_id/">ncimb_id</a></td>
<td>NCIMB (National Collection of Industrial, Food and Marine Bacteria) number</td>
</tr>
<tr>
<td><a href="notes/">notes</a></td>
<td>Free-text notes about this component</td>
</tr>
<tr>
<td><a href="object_medium/">object_medium</a></td>
<td>The medium that is the object of the growth relationship</td>
</tr>
<tr>
<td><a href="oxygen_tolerance/">oxygen_tolerance</a></td>
<td>Oxygen requirement/tolerance category</td>
</tr>
<tr>
<td><a href="parent_taxon_id/">parent_taxon_id</a></td>
<td>Parent taxon in the NCBI Taxonomy hierarchy</td>
</tr>
<tr>
<td><a href="ph/">ph</a></td>
<td>Target pH of the medium</td>
</tr>
<tr>
<td><a href="ph_range/">ph_range</a></td>
<td>Growth pH range or optimum (e</td>
</tr>
<tr>
<td><a href="primary_source/">primary_source</a></td>
<td>The authoritative source for this field</td>
</tr>
<tr>
<td><a href="primary_value/">primary_value</a></td>
<td>Value from the authoritative source</td>
</tr>
<tr>
<td><a href="pubchem_cid/">pubchem_cid</a></td>
<td>PubChem Compound ID</td>
</tr>
<tr>
<td><a href="refseq_accession/">refseq_accession</a></td>
<td>RefSeq assembly accession (GCF_NNNNNNNN</td>
</tr>
<tr>
<td><a href="resolution/">resolution</a></td>
<td>How the conflict was resolved</td>
</tr>
<tr>
<td><a href="resolution_notes/">resolution_notes</a></td>
<td>Notes about the conflict resolution</td>
</tr>
<tr>
<td><a href="roles/">roles</a></td>
<td>Functional roles of this ingredient in this specific mixture</td>
</tr>
<tr>
<td><a href="scientific_name/">scientific_name</a></td>
<td>Binomial scientific name (genus species)</td>
</tr>
<tr>
<td><a href="smiles/">smiles</a></td>
<td>SMILES string for structural representation</td>
</tr>
<tr>
<td><a href="solution/">solution</a></td>
<td>Reference to the solution entity</td>
</tr>
<tr>
<td><a href="solution_type/">solution_type</a></td>
<td>Classification of the solution</td>
</tr>
<tr>
<td><a href="solutions/">solutions</a></td>
<td>Collection of all solution recipes (pre-made mixtures)</td>
</tr>
<tr>
<td><a href="source_fields/">source_fields</a></td>
<td>List of field names populated by this source</td>
</tr>
<tr>
<td><a href="source_id/">source_id</a></td>
<td>Identifier returned by this source</td>
</tr>
<tr>
<td><a href="source_name/">source_name</a></td>
<td>Name of the data source (e</td>
</tr>
<tr>
<td><a href="source_query/">source_query</a></td>
<td>The query used to retrieve data from this source</td>
</tr>
<tr>
<td><a href="source_records/">source_records</a></td>
<td>Records tracking which source provided which data</td>
</tr>
<tr>
<td><a href="source_reference/">source_reference</a></td>
<td>Authoritative reference for this medium formulation</td>
</tr>
<tr>
<td><a href="source_timestamp/">source_timestamp</a></td>
<td>When this data was retrieved from the source</td>
</tr>
<tr>
<td><a href="species_taxon_id/">species_taxon_id</a></td>
<td>NCBITaxon ID at species level</td>
</tr>
<tr>
<td><a href="sterilization_method/">sterilization_method</a></td>
<td>How the medium is sterilized (e</td>
</tr>
<tr>
<td><a href="strain_designation/">strain_designation</a></td>
<td>Strain name or identifier (e</td>
</tr>
<tr>
<td><a href="strains/">strains</a></td>
<td>Collection of all microbial strains</td>
</tr>
<tr>
<td><a href="subject_strain/">subject_strain</a></td>
<td>The strain that is the subject of the growth relationship</td>
</tr>
<tr>
<td><a href="synonyms/">synonyms</a></td>
<td>Alternative names for this entity</td>
</tr>
<tr>
<td><a href="target_organisms/">target_organisms</a></td>
<td>Taxa or organism types this medium is designed for</td>
</tr>
<tr>
<td><a href="taxa/">taxa</a></td>
<td>Collection of all taxonomic entities</td>
</tr>
<tr>
<td><a href="temperature/">temperature</a></td>
<td>Growth temperature in degrees Celsius for this observation</td>
</tr>
<tr>
<td><a href="temperature_range/">temperature_range</a></td>
<td>Growth temperature range or optimum (e</td>
</tr>
<tr>
<td><a href="type_strain/">type_strain</a></td>
<td>Whether this is the type strain for its species</td>
</tr>
<tr>
<td><a href="volume_per_liter/">volume_per_liter</a></td>
<td>Volume of solution added per liter of medium</td>
</tr>
<tr>
<td><a href="volume_unit/">volume_unit</a></td>
<td>Unit of volume (typically mL)</td>
</tr>
<tr>
<td><a href="xref_id/">xref_id</a></td>
<td>The identifier value</td>
</tr>
<tr>
<td><a href="xref_label/">xref_label</a></td>
<td>Human-readable label from the external source</td>
</tr>
<tr>
<td><a href="xref_type/">xref_type</a></td>
<td>Type of cross-reference (e</td>
</tr>
<tr>
<td><a href="xrefs/">xrefs</a></td>
<td>Cross-references to external databases</td>
</tr>
</tbody>
</table>
<h2 id="enumerations">Enumerations</h2>
<table>
<thead>
<tr>
<th>Enumeration</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="AssemblyLevel/">AssemblyLevel</a></td>
<td>Level of genome assembly completeness</td>
</tr>
<tr>
<td><a href="ConcentrationUnit/">ConcentrationUnit</a></td>
<td>Units for concentration</td>
</tr>
<tr>
<td><a href="ConflictResolution/">ConflictResolution</a></td>
<td>How a data conflict between sources was resolved</td>
</tr>
<tr>
<td><a href="DataSource/">DataSource</a></td>
<td>Names of data sources used for enrichment</td>
</tr>
<tr>
<td><a href="GramStain/">GramStain</a></td>
<td>Gram staining classification</td>
</tr>
<tr>
<td><a href="GrowthRate/">GrowthRate</a></td>
<td>Qualitative growth rate categories</td>
</tr>
<tr>
<td><a href="IngredientRole/">IngredientRole</a></td>
<td>Functional role of an ingredient in a growth medium</td>
</tr>
<tr>
<td><a href="MediumType/">MediumType</a></td>
<td>Classification of growth media by composition</td>
</tr>
<tr>
<td><a href="OxygenTolerance/">OxygenTolerance</a></td>
<td>Oxygen requirement categories for microorganisms</td>
</tr>
<tr>
<td><a href="SolutionType/">SolutionType</a></td>
<td>Classification of stock solutions</td>
</tr>
<tr>
<td><a href="TaxonomicRank/">TaxonomicRank</a></td>
<td>NCBI Taxonomy ranks relevant to microbial classification</td>
</tr>
<tr>
<td><a href="VolumeUnit/">VolumeUnit</a></td>
<td>Units for volume</td>
</tr>
</tbody>
</table>
<h2 id="types">Types</h2>
<table>
<thead>
<tr>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="Boolean/">Boolean</a></td>
<td>A binary (true or false) value</td>
</tr>
<tr>
<td><a href="Curie/">Curie</a></td>
<td>a compact URI</td>
</tr>
<tr>
<td><a href="Date/">Date</a></td>
<td>a date (year, month and day) in an idealized calendar</td>
</tr>
<tr>
<td><a href="DateOrDatetime/">DateOrDatetime</a></td>
<td>Either a date or a datetime</td>
</tr>
<tr>
<td><a href="Datetime/">Datetime</a></td>
<td>The combination of a date and time</td>
</tr>
<tr>
<td><a href="Decimal/">Decimal</a></td>
<td>A real number with arbitrary precision that conforms to the xsd:decimal speci...</td>
</tr>
<tr>
<td><a href="Double/">Double</a></td>
<td>A real number that conforms to the xsd:double specification</td>
</tr>
<tr>
<td><a href="Float/">Float</a></td>
<td>A real number that conforms to the xsd:float specification</td>
</tr>
<tr>
<td><a href="Integer/">Integer</a></td>
<td>An integer</td>
</tr>
<tr>
<td><a href="Jsonpath/">Jsonpath</a></td>
<td>A string encoding a JSON Path</td>
</tr>
<tr>
<td><a href="Jsonpointer/">Jsonpointer</a></td>
<td>A string encoding a JSON Pointer</td>
</tr>
<tr>
<td><a href="Ncname/">Ncname</a></td>
<td>Prefix part of CURIE</td>
</tr>
<tr>
<td><a href="Nodeidentifier/">Nodeidentifier</a></td>
<td>A URI, CURIE or BNODE that represents a node in a model</td>
</tr>
<tr>
<td><a href="Objectidentifier/">Objectidentifier</a></td>
<td>A URI or CURIE that represents an object in the model</td>
</tr>
<tr>
<td><a href="Sparqlpath/">Sparqlpath</a></td>
<td>A string encoding a SPARQL Property Path</td>
</tr>
<tr>
<td><a href="String/">String</a></td>
<td>A character string</td>
</tr>
<tr>
<td><a href="Time/">Time</a></td>
<td>A time object represents a (local) time of day, independent of any particular...</td>
</tr>
<tr>
<td><a href="Uri/">Uri</a></td>
<td>a complete URI</td>
</tr>
<tr>
<td><a href="Uriorcurie/">Uriorcurie</a></td>
<td>a URI or a CURIE</td>
</tr>
</tbody>
</table>
<h2 id="subsets">Subsets</h2>
<table>
<thead>
<tr>
<th>Subset</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td></td>
</tr>
</tbody>
</table>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
    <nav
      class="md-footer__inner md-grid"
      aria-label="footer.title"
    >
      
        <a
          href="../best-practices/"
          class="md-footer__link md-footer__link--prev"
          rel="prev"
        >
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
          </div>
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Previous
              </span>
              Best Practices
            </div>
          </div>
        </a>
      
      
        <a
          href="../architecture/"
          class="md-footer__link md-footer__link--next"
          rel="next"
        >
          <div class="md-footer__title">
            <div class="md-ellipsis">
              <span class="md-footer__direction">
                Next
              </span>
              Architecture
            </div>
          </div>
          <div class="md-footer__button md-icon">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
          </div>
        </a>
      
    </nav>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>