
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/bacdive_mongodb_reconciliation_queries/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>BacDive MongoDB Reconciliation Queries - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#bacdive-mongodb-reconciliation-queries" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              BacDive MongoDB Reconciliation Queries
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#important-fields-for-reconciliation-deep-dive" class="md-nav__link">
    <span class="md-ellipsis">
      
        Important Fields for Reconciliation - Deep Dive
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Important Fields for Reconciliation - Deep Dive">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-dsm-number-generaldsm-number" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. DSM Number (General.DSM-Number)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-bacdive-id-generalbacdive-id-or-_id" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. BacDive ID (General.BacDive-ID or _id)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-ncbi-taxonomy-id-generalncbi-tax-idncbi-tax-id" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. NCBI Taxonomy ID (General.NCBI tax id.NCBI tax id)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-culture-collection-ids-external-linksculture-collection-no" class="md-nav__link">
    <span class="md-ellipsis">
      
        4. Culture Collection IDs (External links.culture collection no.)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-strain-designation-name-and-taxonomic-classificationstrain-designation" class="md-nav__link">
    <span class="md-ellipsis">
      
        5. Strain Designation (Name and taxonomic classification.strain designation)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-species-name-name-and-taxonomic-classificationspecies" class="md-nav__link">
    <span class="md-ellipsis">
      
        6. Species Name (Name and taxonomic classification.species)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-synonyms-name-and-taxonomic-classificationlpsnsynonyms" class="md-nav__link">
    <span class="md-ellipsis">
      
        7. Synonyms (Name and taxonomic classification.LPSN.synonyms)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#8-genome-accession-sequence-informationgenome-sequencesaccession" class="md-nav__link">
    <span class="md-ellipsis">
      
        8. Genome Accession (Sequence information.Genome sequences.accession)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#9-16s-rrna-sequences-sequence-information16s-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      
        9. 16S rRNA Sequences (Sequence information.16S sequences)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#10-full-scientific-name-name-and-taxonomic-classificationfull-scientific-name" class="md-nav__link">
    <span class="md-ellipsis">
      
        10. Full Scientific Name (Name and taxonomic classification.full scientific name)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#11-type-strain-status-name-and-taxonomic-classificationtype-strain" class="md-nav__link">
    <span class="md-ellipsis">
      
        11. Type Strain Status (Name and taxonomic classification.type strain)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-strain-history-generalstrain-history" class="md-nav__link">
    <span class="md-ellipsis">
      
        12. Strain History (General.strain history)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-straininfo-links-external-linksstraininfo-linkstraininfo" class="md-nav__link">
    <span class="md-ellipsis">
      
        13. StrainInfo Links (External links.straininfo link.straininfo)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#multi-field-reconciliation-query" class="md-nav__link">
    <span class="md-ellipsis">
      
        Multi-field Reconciliation Query
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#important-caveats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Important Caveats
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-comprehensive-search-for-dsm1338" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example: Comprehensive Search for DSM:1338
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#python-wrapper-function-template" class="md-nav__link">
    <span class="md-ellipsis">
      
        Python Wrapper Function Template
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="bacdive-mongodb-reconciliation-queries">BacDive MongoDB Reconciliation Queries</h1>
<p>Reference guide for querying the local BacDive MongoDB database for strain reconciliation.</p>
<h2 id="important-fields-for-reconciliation-deep-dive">Important Fields for Reconciliation - Deep Dive</h2>
<h3 id="1-dsm-number-generaldsm-number">1. <strong>DSM Number</strong> (<code>General.DSM-Number</code>)</h3>
<p><strong>Why critical</strong>: Primary identifier for DSMZ strains, most reliable lookup
<strong>Data type</strong>: Integer (NOT string)</p>
<p><strong>MongoDB queries</strong>:
<div class="highlight"><pre><span></span><code><span class="c1">// Exact match (fastest, indexed)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">1338</span><span class="p">})</span>

<span class="c1">// IMPORTANT: This will FAIL - DSM-Number is an integer, not a string</span>
<span class="c1">// db.strains.findOne({&quot;General.DSM-Number&quot;: &quot;1338&quot;})  // ‚ùå WRONG</span>

<span class="c1">// Find all DSM strains in a range</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span><span class="nx">$gte</span><span class="o">:</span><span class="w"> </span><span class="mf">1330</span><span class="p">,</span><span class="w"> </span><span class="nx">$lte</span><span class="o">:</span><span class="w"> </span><span class="mf">1340</span><span class="p">}})</span>

<span class="c1">// Convert string input to integer for search</span>
<span class="c1">// In JavaScript</span>
<span class="kd">let</span><span class="w"> </span><span class="nx">dsmInput</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1338&quot;</span><span class="p">;</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="nb">parseInt</span><span class="p">(</span><span class="nx">dsmInput</span><span class="p">)})</span>

<span class="c1">// In Python</span>
<span class="nx">dsm_input</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1338&quot;</span>
<span class="nx">collection</span><span class="p">.</span><span class="nx">find_one</span><span class="p">({</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="kr">int</span><span class="p">(</span><span class="nx">dsm_input</span><span class="p">)})</span>
</code></pre></div></p>
<h3 id="2-bacdive-id-generalbacdive-id-or-_id">2. <strong>BacDive ID</strong> (<code>General.BacDive-ID</code> or <code>_id</code>)</h3>
<p><strong>Why critical</strong>: Unique identifier across entire BacDive database</p>
<div class="highlight"><pre><span></span><code><span class="c1">// By BacDive ID</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span><span class="s2">&quot;General.BacDive-ID&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">7143</span><span class="p">})</span>

<span class="c1">// By MongoDB _id (usually same as BacDive-ID)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span><span class="s2">&quot;_id&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">7143</span><span class="p">})</span>

<span class="c1">// Find multiple BacDive IDs at once</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;General.BacDive-ID&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span><span class="nx">$in</span><span class="o">:</span><span class="w"> </span><span class="p">[</span><span class="mf">7143</span><span class="p">,</span><span class="w"> </span><span class="mf">7144</span><span class="p">,</span><span class="w"> </span><span class="mf">7145</span><span class="p">]}})</span>
</code></pre></div>
<h3 id="3-ncbi-taxonomy-id-generalncbi-tax-idncbi-tax-id">3. <strong>NCBI Taxonomy ID</strong> (<code>General.NCBI tax id.NCBI tax id</code>)</h3>
<p><strong>Why critical</strong>: Links to NCBI Taxonomy, but <strong>NOT unique</strong> - many strains can share same species-level taxon</p>
<div class="highlight"><pre><span></span><code><span class="c1">// By NCBI taxon ID</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;General.NCBI tax id.NCBI tax id&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">408</span><span class="p">})</span>

<span class="c1">// With matching level filter (species vs strain)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;General.NCBI tax id.NCBI tax id&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">408</span><span class="p">,</span>
<span class="w">  </span><span class="s2">&quot;General.NCBI tax id.Matching level&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;species&quot;</span>
<span class="p">})</span>

<span class="c1">// Find strain-level taxon matches (more specific)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;General.NCBI tax id.Matching level&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;strain&quot;</span>
<span class="p">})</span>

<span class="c1">// Alternative path if structure differs</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;NCBI tax id&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">408</span><span class="p">})</span>
</code></pre></div>
<h3 id="4-culture-collection-ids-external-linksculture-collection-no">4. <strong>Culture Collection IDs</strong> (<code>External links.culture collection no.</code>)</h3>
<p><strong>Why critical</strong>: Cross-references to other culture collections - <strong>STRING FIELD</strong> with comma-separated list</p>
<div class="highlight"><pre><span></span><code><span class="c1">// Exact substring match for ATCC 14718</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/ATCC 14718/</span>
<span class="p">})</span>

<span class="c1">// Word boundary regex (prevents &quot;DSM 1&quot; from matching &quot;DSM 11&quot;)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/(^|,\s*)ATCC 14718(\s*,|$)/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Case-insensitive search for any NCIMB entry</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/NCIMB/i</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Find strains in multiple collections</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/DSM 1338|JCM 2805/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// All strains with ATCC numbers</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/\bATCC\s+\d+/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="5-strain-designation-name-and-taxonomic-classificationstrain-designation">5. <strong>Strain Designation</strong> (<code>Name and taxonomic classification.strain designation</code>)</h3>
<p><strong>Why critical</strong>: Human-readable strain identifier (e.g., "AM1", "KT2440"), but can have <strong>multiple comma-separated values</strong></p>
<div class="highlight"><pre><span></span><code><span class="c1">// Exact match for single designation</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;AM1&quot;</span>
<span class="p">})</span>

<span class="c1">// Partial match (handles &quot;0355, D355, AM1&quot;)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/AM1/</span>
<span class="p">})</span>

<span class="c1">// Word boundary match (prevents &quot;AM1&quot; from matching &quot;AM10&quot;)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/\bAM1\b/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Case-insensitive</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/am-1/i</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Alternative field path</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/KT2440/</span><span class="p">})</span>
</code></pre></div>
<h3 id="6-species-name-name-and-taxonomic-classificationspecies">6. <strong>Species Name</strong> (<code>Name and taxonomic classification.species</code>)</h3>
<p><strong>Why critical</strong>: Binomial name, but <strong>NOT unique</strong> - returns ALL strains of that species</p>
<div class="highlight"><pre><span></span><code><span class="c1">// Exact species match</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.species&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylorubrum extorquens&quot;</span>
<span class="p">})</span>

<span class="c1">// Case-insensitive</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.species&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/^Methylorubrum extorquens$/i</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Genus-level search</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.genus&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylorubrum&quot;</span>
<span class="p">})</span>

<span class="c1">// Partial species name</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.species&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/Methylo.*extorquens/i</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Alternative paths (older records)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;species&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylorubrum extorquens&quot;</span><span class="p">})</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span><span class="s2">&quot;Name and taxonomic classification.LPSN.species&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylorubrum extorquens&quot;</span><span class="p">})</span>
</code></pre></div>
<h3 id="7-synonyms-name-and-taxonomic-classificationlpsnsynonyms">7. <strong>Synonyms</strong> (<code>Name and taxonomic classification.LPSN.synonyms</code>)</h3>
<p><strong>Why critical</strong>: Finds strains by old/alternative scientific names, <strong>ARRAY field</strong></p>
<div class="highlight"><pre><span></span><code><span class="c1">// Find by synonym (exact match in array)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.LPSN.synonyms.synonym&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylobacterium extorquens&quot;</span>
<span class="p">})</span>

<span class="c1">// Case-insensitive synonym search</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.LPSN.synonyms.synonym&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/Protomonas extorquens/i</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Find any strain with &quot;Methylobacterium&quot; synonym</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.LPSN.synonyms.synonym&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/Methylobacterium/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Check if synonyms array exists and is not empty</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.LPSN.synonyms&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$exists</span><span class="o">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span>
<span class="w">    </span><span class="nx">$ne</span><span class="o">:</span><span class="w"> </span><span class="p">[]</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="8-genome-accession-sequence-informationgenome-sequencesaccession">8. <strong>Genome Accession</strong> (<code>Sequence information.Genome sequences.accession</code>)</h3>
<p><strong>Why critical</strong>: Links to NCBI GenBank genome assembly, <strong>unique identifier</strong></p>
<div class="highlight"><pre><span></span><code><span class="c1">// Exact genome accession</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.Genome sequences.accession&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;GCA_000022685&quot;</span>
<span class="p">})</span>

<span class="c1">// Partial accession (GCA_ or GCF_)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.Genome sequences.accession&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/^GCA_000022/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// All strains with genomes</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.Genome sequences&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$exists</span><span class="o">:</span><span class="w"> </span><span class="kc">true</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Alternative path for arrays</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.Genome sequences.0.accession&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;GCA_000022685&quot;</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="9-16s-rrna-sequences-sequence-information16s-sequences">9. <strong>16S rRNA Sequences</strong> (<code>Sequence information.16S sequences</code>)</h3>
<p><strong>Why critical</strong>: Links to NCBI nucleotide sequences, <strong>ARRAY field</strong></p>
<div class="highlight"><pre><span></span><code><span class="c1">// By 16S accession</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.16S sequences.accession&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;AF293375&quot;</span>
<span class="p">})</span>

<span class="c1">// Find strains with 16S sequences</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.16S sequences&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$exists</span><span class="o">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span>
<span class="w">    </span><span class="nx">$ne</span><span class="o">:</span><span class="w"> </span><span class="p">[]</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// By 16S sequence NCBI tax ID</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Sequence information.16S sequences.NCBI tax ID&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">408</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="10-full-scientific-name-name-and-taxonomic-classificationfull-scientific-name">10. <strong>Full Scientific Name</strong> (<code>Name and taxonomic classification.full scientific name</code>)</h3>
<p><strong>Why critical</strong>: Complete name with author/year, good for disambiguation</p>
<div class="highlight"><pre><span></span><code><span class="c1">// Exact match</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.full scientific name&quot;</span><span class="o">:</span>
<span class="w">    </span><span class="s2">&quot;Methylorubrum extorquens (Urakami and Komagata 1984) Green and Ardley 2018&quot;</span>
<span class="p">})</span>

<span class="c1">// Partial match (ignore author/year)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.full scientific name&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/^Methylorubrum extorquens/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Also check LPSN full name (may differ)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.LPSN.full scientific name&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/Methylorubrum extorquens/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="11-type-strain-status-name-and-taxonomic-classificationtype-strain">11. <strong>Type Strain Status</strong> (<code>Name and taxonomic classification.type strain</code>)</h3>
<p><strong>Why critical</strong>: Identifies authoritative reference strains</p>
<div class="highlight"><pre><span></span><code><span class="c1">// Find type strains only</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.type strain&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;yes&quot;</span>
<span class="p">})</span>

<span class="c1">// Find non-type strains</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.type strain&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;no&quot;</span>
<span class="p">})</span>

<span class="c1">// Type strains of specific species</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.species&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylorubrum extorquens&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="s2">&quot;Name and taxonomic classification.type strain&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;yes&quot;</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="12-strain-history-generalstrain-history">12. <strong>Strain History</strong> (<code>General.strain history</code>)</h3>
<p><strong>Why critical</strong>: Tracks strain provenance, may mention alternative designations</p>
<div class="highlight"><pre><span></span><code><span class="c1">// Search in strain history text</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;General.strain history.history&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/NCIB 9133/i</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>

<span class="c1">// Find strains with provenance from specific source</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;General.strain history.history&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$regex</span><span class="o">:</span><span class="w"> </span><span class="sr">/Quayle/</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>
</code></pre></div>
<h3 id="13-straininfo-links-external-linksstraininfo-linkstraininfo">13. <strong>StrainInfo Links</strong> (<code>External links.straininfo link.straininfo</code>)</h3>
<p><strong>Why critical</strong>: Cross-reference to StrainInfo database</p>
<div class="highlight"><pre><span></span><code><span class="c1">// By StrainInfo ID</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">findOne</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.straininfo link.straininfo&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">45565</span>
<span class="p">})</span>

<span class="c1">// All strains with StrainInfo entries</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="s2">&quot;External links.straininfo link&quot;</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$exists</span><span class="o">:</span><span class="w"> </span><span class="kc">true</span>
<span class="w">  </span><span class="p">}</span>
<span class="p">})</span>
</code></pre></div>
<hr />
<h2 id="multi-field-reconciliation-query"><strong>Multi-field Reconciliation Query</strong></h2>
<p>Combine multiple fields for comprehensive search:</p>
<div class="highlight"><pre><span></span><code><span class="c1">// Find strain by ANY identifier</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="nx">$or</span><span class="o">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">1338</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;General.BacDive-ID&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">7143</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;General.NCBI tax id.NCBI tax id&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">408</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/ATCC 14718/</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/AM1/</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;Sequence information.Genome sequences.accession&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;GCA_000022685&quot;</span><span class="p">}</span>
<span class="w">  </span><span class="p">]</span>
<span class="p">})</span>

<span class="c1">// Score matches by number of fields matched (pseudo-confidence)</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">aggregate</span><span class="p">([</span>
<span class="w">  </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$match</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="nx">$or</span><span class="o">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">        </span><span class="p">{</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">1338</span><span class="p">},</span>
<span class="w">        </span><span class="p">{</span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/JCM 2805/</span><span class="p">},</span>
<span class="w">        </span><span class="p">{</span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/AM1/</span><span class="p">}</span>
<span class="w">      </span><span class="p">]</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">  </span><span class="p">},</span>
<span class="w">  </span><span class="p">{</span>
<span class="w">    </span><span class="nx">$addFields</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">      </span><span class="nx">matchScore</span><span class="o">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="nx">$add</span><span class="o">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">          </span><span class="p">{</span><span class="nx">$cond</span><span class="o">:</span><span class="w"> </span><span class="p">[{</span><span class="nx">$eq</span><span class="o">:</span><span class="w"> </span><span class="p">[</span><span class="s2">&quot;$General.DSM-Number&quot;</span><span class="p">,</span><span class="w"> </span><span class="mf">1338</span><span class="p">]},</span><span class="w"> </span><span class="mf">1</span><span class="p">,</span><span class="w"> </span><span class="mf">0</span><span class="p">]},</span>
<span class="w">          </span><span class="p">{</span><span class="nx">$cond</span><span class="o">:</span><span class="w"> </span><span class="p">[{</span><span class="nx">$regexMatch</span><span class="o">:</span><span class="w"> </span><span class="p">{</span><span class="nx">input</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;$External links.culture collection no.&quot;</span><span class="p">,</span><span class="w"> </span><span class="nx">regex</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;JCM 2805&quot;</span><span class="p">}},</span><span class="w"> </span><span class="mf">1</span><span class="p">,</span><span class="w"> </span><span class="mf">0</span><span class="p">]},</span>
<span class="w">          </span><span class="p">{</span><span class="nx">$cond</span><span class="o">:</span><span class="w"> </span><span class="p">[{</span><span class="nx">$regexMatch</span><span class="o">:</span><span class="w"> </span><span class="p">{</span><span class="nx">input</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;$Name and taxonomic classification.strain designation&quot;</span><span class="p">,</span><span class="w"> </span><span class="nx">regex</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;AM1&quot;</span><span class="p">}},</span><span class="w"> </span><span class="mf">1</span><span class="p">,</span><span class="w"> </span><span class="mf">0</span><span class="p">]}</span>
<span class="w">        </span><span class="p">]</span>
<span class="w">      </span><span class="p">}</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">  </span><span class="p">},</span>
<span class="w">  </span><span class="p">{</span><span class="nx">$sort</span><span class="o">:</span><span class="w"> </span><span class="p">{</span><span class="nx">matchScore</span><span class="o">:</span><span class="w"> </span><span class="o">-</span><span class="mf">1</span><span class="p">}}</span>
<span class="p">])</span>
</code></pre></div>
<hr />
<h2 id="important-caveats"><strong>Important Caveats</strong></h2>
<ol>
<li><strong>Culture collection string is comma-separated</strong> - always use regex with word boundaries</li>
<li><strong>NCBI taxon ID is NOT unique</strong> - one species can have 100s of strains</li>
<li><strong>Strain designation can have multiple values</strong> - "0355, D355, AM1"</li>
<li><strong>Synonyms are in nested arrays</strong> - need to query the <code>synonym</code> field within array</li>
<li><strong>Some records may use different field paths</strong> - always check alternate locations</li>
</ol>
<hr />
<h2 id="example-comprehensive-search-for-dsm1338">Example: Comprehensive Search for DSM:1338</h2>
<div class="highlight"><pre><span></span><code><span class="c1">// Search by all available identifiers</span>
<span class="nx">db</span><span class="p">.</span><span class="nx">strains</span><span class="p">.</span><span class="nx">find</span><span class="p">({</span>
<span class="w">  </span><span class="nx">$or</span><span class="o">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">    </span><span class="c1">// Primary DSM number</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">1338</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// BacDive ID</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;General.BacDive-ID&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">7143</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// NCBI taxonomy</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;General.NCBI tax id.NCBI tax id&quot;</span><span class="o">:</span><span class="w"> </span><span class="mf">408</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// Culture collection IDs</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/(^|,\s*)DSM 1338(\s*,|$)/</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/(^|,\s*)ATCC 14718(\s*,|$)/</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/(^|,\s*)NCIB 9133(\s*,|$)/</span><span class="p">},</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;External links.culture collection no.&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/(^|,\s*)JCM 2805(\s*,|$)/</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// Strain designation</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="o">:</span><span class="w"> </span><span class="sr">/\b(0355|D355|AM1)\b/i</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// Species name</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;Name and taxonomic classification.species&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylorubrum extorquens&quot;</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// Synonyms</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;Name and taxonomic classification.LPSN.synonyms.synonym&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Methylobacterium extorquens&quot;</span><span class="p">},</span>

<span class="w">    </span><span class="c1">// Genome accession</span>
<span class="w">    </span><span class="p">{</span><span class="s2">&quot;Sequence information.Genome sequences.accession&quot;</span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;GCA_000022685&quot;</span><span class="p">}</span>
<span class="w">  </span><span class="p">]</span>
<span class="p">})</span>
</code></pre></div>
<hr />
<h2 id="python-wrapper-function-template">Python Wrapper Function Template</h2>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">search_bacdive_comprehensive</span><span class="p">(</span>
    <span class="n">collection</span><span class="p">,</span>
    <span class="n">dsm_number</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">bacdive_id</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">ncbi_taxon</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">culture_collection_ids</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">strain_designation</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">species_name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">genome_accession</span><span class="o">=</span><span class="kc">None</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Comprehensive BacDive search by any identifier.&quot;&quot;&quot;</span>

    <span class="n">query_parts</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="n">dsm_number</span><span class="p">:</span>
        <span class="c1"># DSM-Number is stored as integer, not string</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;General.DSM-Number&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">dsm_number</span><span class="p">)})</span>

    <span class="k">if</span> <span class="n">bacdive_id</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;General.BacDive-ID&quot;</span><span class="p">:</span> <span class="n">bacdive_id</span><span class="p">})</span>

    <span class="k">if</span> <span class="n">ncbi_taxon</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span><span class="s2">&quot;General.NCBI tax id.NCBI tax id&quot;</span><span class="p">:</span> <span class="n">ncbi_taxon</span><span class="p">})</span>

    <span class="k">if</span> <span class="n">culture_collection_ids</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">cc_id</span> <span class="ow">in</span> <span class="n">culture_collection_ids</span><span class="p">:</span>
            <span class="n">pattern</span> <span class="o">=</span> <span class="sa">rf</span><span class="s2">&quot;(^|,\s*)</span><span class="si">{</span><span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">(</span><span class="n">cc_id</span><span class="p">)</span><span class="si">}</span><span class="s2">(\s*,|$)&quot;</span>
            <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
                <span class="s2">&quot;External links.culture collection no.&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;$regex&quot;</span><span class="p">:</span> <span class="n">pattern</span><span class="p">}</span>
            <span class="p">})</span>

    <span class="k">if</span> <span class="n">strain_designation</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;Name and taxonomic classification.strain designation&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;$regex&quot;</span><span class="p">:</span> <span class="sa">rf</span><span class="s2">&quot;\b</span><span class="si">{</span><span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">(</span><span class="n">strain_designation</span><span class="p">)</span><span class="si">}</span><span class="s2">\b&quot;</span><span class="p">,</span>
                <span class="s2">&quot;$options&quot;</span><span class="p">:</span> <span class="s2">&quot;i&quot;</span>
            <span class="p">}</span>
        <span class="p">})</span>

    <span class="k">if</span> <span class="n">species_name</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;Name and taxonomic classification.species&quot;</span><span class="p">:</span> <span class="n">species_name</span>
        <span class="p">})</span>

    <span class="k">if</span> <span class="n">genome_accession</span><span class="p">:</span>
        <span class="n">query_parts</span><span class="o">.</span><span class="n">append</span><span class="p">({</span>
            <span class="s2">&quot;Sequence information.Genome sequences.accession&quot;</span><span class="p">:</span> <span class="n">genome_accession</span>
        <span class="p">})</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">query_parts</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>

    <span class="n">query</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;$or&quot;</span><span class="p">:</span> <span class="n">query_parts</span><span class="p">}</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">query_parts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="k">else</span> <span class="n">query_parts</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">collection</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">query</span><span class="p">))</span>
</code></pre></div>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>