
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/CMM_KG_ML_Pipeline_Overview/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>CMM Knowledge Graph & Machine Learning Pipeline - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cmm-knowledge-graph-machine-learning-pipeline" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CMM Knowledge Graph &amp; Machine Learning Pipeline
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    <span class="md-ellipsis">
      
        Table of Contents
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pipeline-overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Pipeline Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-sources" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Sources
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Sources">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#bacdive-mediadive" class="md-nav__link">
    <span class="md-ellipsis">
      
        BacDive / MediaDive
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dsmz-media-database" class="md-nav__link">
    <span class="md-ellipsis">
      
        DSMZ Media Database
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#chemical-ontologies" class="md-nav__link">
    <span class="md-ellipsis">
      
        Chemical Ontologies
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#component-details" class="md-nav__link">
    <span class="md-ellipsis">
      
        Component Details
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Component Details">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-micromediaparam-public" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. MicroMediaParam (PUBLIC)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-kg-ai-prep-private" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. kg-ai-prep (PRIVATE)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-microgrowlink-private" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. MicroGrowLink (PRIVATE)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-microbe-rules-public" class="md-nav__link">
    <span class="md-ellipsis">
      
        4. microbe-rules (PUBLIC)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#5-microgrowlinkservice-public" class="md-nav__link">
    <span class="md-ellipsis">
      
        5. MicroGrowLinkService (PUBLIC)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#6-cmm-ai-private" class="md-nav__link">
    <span class="md-ellipsis">
      
        6. CMM-AI (PRIVATE)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#7-mate-llm-private" class="md-nav__link">
    <span class="md-ellipsis">
      
        7. MATE-LLM (PRIVATE)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#kgx-data-format" class="md-nav__link">
    <span class="md-ellipsis">
      
        KGX Data Format
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="KGX Data Format">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#nodes-file-nodestsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        Nodes File (nodes.tsv)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#edges-file-edgestsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        Edges File (edges.tsv)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#key-predicates" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Predicates
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#machine-learning-models" class="md-nav__link">
    <span class="md-ellipsis">
      
        Machine Learning Models
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Machine Learning Models">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#graph-embedding-models-microgrowlink" class="md-nav__link">
    <span class="md-ellipsis">
      
        Graph Embedding Models (MicroGrowLink)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#interpretable-models-microbe-rules" class="md-nav__link">
    <span class="md-ellipsis">
      
        Interpretable Models (microbe-rules)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#export-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Export Formats
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#your-role-knowledge-graph-construction" class="md-nav__link">
    <span class="md-ellipsis">
      
        Your Role: Knowledge Graph Construction
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Your Role: Knowledge Graph Construction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#what-you-need-to-build" class="md-nav__link">
    <span class="md-ellipsis">
      
        What You Need to Build
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#input-sources" class="md-nav__link">
    <span class="md-ellipsis">
      
        Input Sources
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      
        Output Requirements
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tools-available" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tools Available
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#workflow" class="md-nav__link">
    <span class="md-ellipsis">
      
        Workflow
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#anticipated-payoffs" class="md-nav__link">
    <span class="md-ellipsis">
      
        Anticipated Payoffs
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Anticipated Payoffs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#scientific-goals" class="md-nav__link">
    <span class="md-ellipsis">
      
        Scientific Goals
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#technical-benefits" class="md-nav__link">
    <span class="md-ellipsis">
      
        Technical Benefits
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#broader-impact" class="md-nav__link">
    <span class="md-ellipsis">
      
        Broader Impact
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#repository-access" class="md-nav__link">
    <span class="md-ellipsis">
      
        Repository Access
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Repository Access">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#public-repositories-direct-access" class="md-nav__link">
    <span class="md-ellipsis">
      
        Public Repositories (Direct Access)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#private-repositories-gh-cli-access" class="md-nav__link">
    <span class="md-ellipsis">
      
        Private Repositories (gh CLI Access)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#quick-reference" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quick Reference
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quick Reference">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#key-target-relation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Key Target Relation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#kgx-node-categories" class="md-nav__link">
    <span class="md-ellipsis">
      
        KGX Node Categories
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filtering-commands-kg-ai-prep" class="md-nav__link">
    <span class="md-ellipsis">
      
        Filtering Commands (kg-ai-prep)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#related-slack-workspaces" class="md-nav__link">
    <span class="md-ellipsis">
      
        Related Slack Workspaces
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#related-github-repos" class="md-nav__link">
    <span class="md-ellipsis">
      
        Related GitHub Repos
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="cmm-knowledge-graph-machine-learning-pipeline">CMM Knowledge Graph &amp; Machine Learning Pipeline</h1>
<blockquote>
<p>Comprehensive overview of the CultureBotAI/CMM project's data pipeline, knowledge graph construction, and machine learning components.</p>
</blockquote>
<p><strong>Last updated:</strong> 2025-12-18</p>
<hr />
<h2 id="table-of-contents">Table of Contents</h2>
<ol>
<li><a href="#pipeline-overview">Pipeline Overview</a></li>
<li><a href="#data-sources">Data Sources</a></li>
<li><a href="#component-details">Component Details</a></li>
<li><a href="#1-micromediaparam-public">MicroMediaParam</a></li>
<li><a href="#2-kg-ai-prep-private">kg-ai-prep</a></li>
<li><a href="#3-microgrowlink-private">MicroGrowLink</a></li>
<li><a href="#4-microbe-rules-public">microbe-rules</a></li>
<li><a href="#5-microgrowlinkservice-public">MicroGrowLinkService</a></li>
<li><a href="#6-cmm-ai-private">CMM-AI</a></li>
<li><a href="#7-mate-llm-private">MATE-LLM</a></li>
<li><a href="#kgx-data-format">KGX Data Format</a></li>
<li><a href="#machine-learning-models">Machine Learning Models</a></li>
<li><a href="#your-role-knowledge-graph-construction">Your Role: Knowledge Graph Construction</a></li>
<li><a href="#anticipated-payoffs">Anticipated Payoffs</a></li>
<li><a href="#repository-access">Repository Access</a></li>
</ol>
<hr />
<h2 id="pipeline-overview">Pipeline Overview</h2>
<div class="highlight"><pre><span></span><code>┌─────────────────────────────────────────────────────────────────────────────┐
│  1. DATA SOURCES                                                            │
│  ├── BacDive/MediaDive (strain-media relationships)                         │
│  ├── DSMZ (media compositions, chemical ingredients)                        │
│  └── ChEBI/PubChem (chemical ontologies)                                    │
└─────────────────────────────────────────────────────────────────────────────┘
                                    ↓
┌─────────────────────────────────────────────────────────────────────────────┐
│  2. MicroMediaParam                                                         │
│  Chemical compound → Knowledge Graph entity mapping                         │
│  Output: composition_kg_mapping_final.tsv                                   │
└─────────────────────────────────────────────────────────────────────────────┘
                                    ↓
┌─────────────────────────────────────────────────────────────────────────────┐
│  3. Knowledge Graph Construction (YOUR WORK)                                │
│  Create strain→medium edges in KGX format                                   │
│  Output: nodes.tsv, edges.tsv                                               │
└─────────────────────────────────────────────────────────────────────────────┘
                                    ↓
┌─────────────────────────────────────────────────────────────────────────────┐
│  4. kg-ai-prep                                                              │
│  Graph preprocessing &amp; filtering for ML                                     │
│  Output: Model-ready train/val/test splits                                  │
└─────────────────────────────────────────────────────────────────────────────┘
                                    ↓
┌─────────────────────────────────────────────────────────────────────────────┐
│  5. MicroGrowLink                                                           │
│  Train graph transformer models (RGT, HGT, NBFNet)                          │
│  Task: Link prediction for &quot;microbe_grows_in_medium&quot;                        │
└─────────────────────────────────────────────────────────────────────────────┘
                                    ↓
┌─────────────────────────────────────────────────────────────────────────────┐
│  6. microbe-rules                                                           │
│  CatBoost models for feature importance &amp; interpretability                  │
│  Output: Which chemicals/features predict growth                            │
└─────────────────────────────────────────────────────────────────────────────┘
                                    ↓
┌─────────────────────────────────────────────────────────────────────────────┐
│  7. MicroGrowLinkService                                                    │
│  Gradio web application serving predictions                                 │
└─────────────────────────────────────────────────────────────────────────────┘
</code></pre></div>
<hr />
<h2 id="data-sources">Data Sources</h2>
<h3 id="bacdive-mediadive">BacDive / MediaDive</h3>
<ul>
<li><strong>URL:</strong> https://bacdive.dsmz.de / https://mediadive.dsmz.de</li>
<li><strong>Content:</strong> Strain metadata, growth conditions, media compositions</li>
<li><strong>Format:</strong> JSON API</li>
</ul>
<h3 id="dsmz-media-database">DSMZ Media Database</h3>
<ul>
<li><strong>Content:</strong> 1,807+ microbial growth media formulations</li>
<li><strong>Details:</strong> Chemical compositions, concentrations, preparation instructions</li>
<li><strong>Solutions:</strong> Expandable solution references (e.g., "solution:241")</li>
</ul>
<h3 id="chemical-ontologies">Chemical Ontologies</h3>
<ul>
<li><strong>ChEBI:</strong> Chemical Entities of Biological Interest</li>
<li><strong>PubChem:</strong> Chemical compound database</li>
<li><strong>CAS-RN:</strong> Chemical Abstracts Service Registry Numbers</li>
</ul>
<hr />
<h2 id="component-details">Component Details</h2>
<h3 id="1-micromediaparam-public">1. MicroMediaParam (PUBLIC)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/MicroMediaParam</p>
<p><strong>Purpose:</strong> Map chemical compounds from growth media to knowledge graph entities.</p>
<p><strong>Key Statistics:</strong>
- 23,181 chemical entries processed
- 1,807 microbial growth media
- 78% ChEBI coverage (18,088 compounds mapped)
- 99.99% mapping accuracy</p>
<p><strong>Features:</strong>
- Hydrate parsing (e.g., "MgCl2 6-hydrate" → base + water molecules)
- Multi-database mapping (ChEBI, PubChem, CAS-RN)
- Molecular weight calculation (anhydrous and hydrated forms)
- DSMZ solution expansion into individual chemicals
- Fuzzy matching for chemical name variations</p>
<p><strong>Output:</strong> <code>composition_kg_mapping_final.tsv</code></p>
<p>Key columns:
- <code>medium_id</code>: Unique medium identifier
- <code>original</code>: Original compound name
- <code>mapped</code>: Database identifier (ChEBI/PubChem/CAS-RN)
- <code>base_compound</code>: Chemical without hydration
- <code>water_molecules</code>: Hydration number
- <code>base_chebi_id</code>: ChEBI identifier
- <code>base_chebi_label</code>: Human-readable name
- <code>base_chebi_formula</code>: Standardized formula
- <code>base_molecular_weight</code>: Anhydrous MW
- <code>hydrated_molecular_weight</code>: Total MW with water</p>
<hr />
<h3 id="2-kg-ai-prep-private">2. kg-ai-prep (PRIVATE)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/kg-ai-prep</p>
<p><strong>Purpose:</strong> Prepare knowledge graphs for AI model training with research-backed filtering strategies.</p>
<p><strong>Problems Addressed:</strong>
- Embedding collapse (trivial values)
- Low-confidence predictions
- Training instability
- Structural noise in raw KGs</p>
<p><strong>Filtering Strategies:</strong></p>
<table>
<thead>
<tr>
<th>Filter Type</th>
<th>Options</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Structural</strong></td>
<td>Singleton removal, leaf pruning, degree thresholds, k-core extraction, hub removal</td>
</tr>
<tr>
<td><strong>Component</strong></td>
<td>Giant component extraction, small component removal, bridge detection</td>
</tr>
<tr>
<td><strong>Relational</strong></td>
<td>Task-specific filtering, redundant relation removal, confidence thresholds</td>
</tr>
<tr>
<td><strong>Metapath</strong></td>
<td>Domain-guided subgraph extraction, multi-hop preservation</td>
</tr>
</tbody>
</table>
<p><strong>Model-Specific Optimizations:</strong></p>
<table>
<thead>
<tr>
<th>Model</th>
<th>Strategy</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>RotatE/TransE</strong></td>
<td>Remove singletons/leaves, dense connectivity</td>
</tr>
<tr>
<td><strong>RGT/GNNs</strong></td>
<td>Preserve attribute nodes, k-core filtering</td>
</tr>
<tr>
<td><strong>A*Net</strong></td>
<td>Ensure path connectivity, minimum degree</td>
</tr>
</tbody>
</table>
<p><strong>CLI Usage:</strong>
<div class="highlight"><pre><span></span><code><span class="c1"># Analyze graph</span>
kg-ai-prep<span class="w"> </span>analyze<span class="w"> </span>nodes.tsv<span class="w"> </span>edges.tsv<span class="w"> </span>--target-relation<span class="w"> </span><span class="s2">&quot;microbe_grows_in_medium&quot;</span>

<span class="c1"># Full preprocessing</span>
kg-ai-prep<span class="w"> </span>preprocess<span class="w"> </span>nodes.tsv<span class="w"> </span>edges.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output-dir<span class="w"> </span>processed_kg/<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--remove-singletons<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--min-degree<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--giant-component-only<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--target-relation<span class="w"> </span><span class="s2">&quot;microbe_grows_in_medium&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--export-rotate
</code></pre></div></p>
<p><strong>Python API:</strong>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">kg_ai_prep</span><span class="w"> </span><span class="kn">import</span> <span class="n">KGPreprocessor</span><span class="p">,</span> <span class="n">FilterConfig</span><span class="p">,</span> <span class="n">SplitConfig</span>

<span class="n">prep</span> <span class="o">=</span> <span class="n">KGPreprocessor</span><span class="p">()</span>
<span class="n">graph</span> <span class="o">=</span> <span class="n">prep</span><span class="o">.</span><span class="n">load_kgx</span><span class="p">(</span><span class="s2">&quot;nodes.tsv&quot;</span><span class="p">,</span> <span class="s2">&quot;edges.tsv&quot;</span><span class="p">)</span>

<span class="n">config</span> <span class="o">=</span> <span class="n">FilterConfig</span><span class="p">(</span>
    <span class="n">target_relation</span><span class="o">=</span><span class="s2">&quot;microbe_grows_in_medium&quot;</span><span class="p">,</span>
    <span class="n">remove_singletons</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">min_degree</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">keep_giant_component_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">model_type</span><span class="o">=</span><span class="s2">&quot;rotate&quot;</span>
<span class="p">)</span>

<span class="n">filtered_graph</span> <span class="o">=</span> <span class="n">prep</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="n">config</span><span class="p">)</span>

<span class="n">split_config</span> <span class="o">=</span> <span class="n">SplitConfig</span><span class="p">(</span>
    <span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;transductive_edge&quot;</span><span class="p">,</span>
    <span class="n">ratios</span><span class="o">=</span><span class="p">(</span><span class="mf">0.8</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">),</span>
    <span class="n">negative_sampling_ratio</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
    <span class="n">hard_negatives</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">splits</span> <span class="o">=</span> <span class="n">prep</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">filtered_graph</span><span class="p">,</span> <span class="n">split_config</span><span class="p">)</span>
<span class="n">prep</span><span class="o">.</span><span class="n">export_rotate</span><span class="p">(</span><span class="n">splits</span><span class="p">,</span> <span class="s2">&quot;rotate_output/&quot;</span><span class="p">)</span>
</code></pre></div></p>
<p><strong>Performance Benefits</strong> (from Ratajczak et al., 2022):
- 20-40% improvement on target prediction tasks
- 26-60% reduction in irrelevant nodes
- Increased prediction confidence
- More stable hyperparameter tuning</p>
<hr />
<h3 id="3-microgrowlink-private">3. MicroGrowLink (PRIVATE)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/MicroGrowLink</p>
<p><strong>Purpose:</strong> Train link prediction models on the microbial growth knowledge graph.</p>
<p><strong>Supported Models:</strong></p>
<table>
<thead>
<tr>
<th>Model</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>RGT</strong></td>
<td>Relational Graph Transformer</td>
<td>Transformer architecture for heterogeneous graphs</td>
</tr>
<tr>
<td><strong>HGT</strong></td>
<td>Heterogeneous Graph Transformer</td>
<td>Attention-based heterogeneous graph learning</td>
</tr>
<tr>
<td><strong>NBFNet</strong></td>
<td>Neural Bellman-Ford Network</td>
<td>Path-based reasoning for link prediction</td>
</tr>
</tbody>
</table>
<p><em>Note: Relational GCNs (RGCNs) have been deprecated.</em></p>
<p><strong>Input Requirements:</strong>
- <code>merged-kg_nodes.tsv</code> (KGX format)
- <code>merged-kg_edges.tsv</code> (KGX format)</p>
<p><strong>Directory Structure:</strong>
<div class="highlight"><pre><span></span><code>MicroGrowLink/
├── data/           # KGX data and taxa lists
├── hpc/            # SLURM scripts for multi-GPU
├── scripts/        # Utility scripts
├── src/
│   ├── learn/      # Model implementations
│   └── eval/       # Evaluation scripts
├── docs/           # Model documentation (RGT.md, HGT.md, NBFNet.md)
└── predictions/    # Output predictions
</code></pre></div></p>
<p><strong>Task:</strong> Predict <code>microbe_grows_in_medium</code> edges (which microbes can grow in which media).</p>
<hr />
<h3 id="4-microbe-rules-public">4. microbe-rules (PUBLIC)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/microbe-rules</p>
<p><strong>Purpose:</strong> Interpretable ML using CatBoost for feature importance analysis.</p>
<p><strong>Models:</strong>
- Binary classification models (CatBoost)
- Two configurations: 514 features and 65 features
- Multiple model variants per configuration</p>
<p><strong>Pipeline:</strong>
<div class="highlight"><pre><span></span><code><span class="c1"># Prepare data</span>
python<span class="w"> </span>01_prepare_data_binary.py

<span class="c1"># Train models</span>
python<span class="w"> </span>02_compute_compare_models.py<span class="w"> </span><span class="m">514</span><span class="w"> </span>--model_id<span class="w"> </span><span class="m">1</span>
python<span class="w"> </span>02_compute_compare_models.py<span class="w"> </span><span class="m">65</span><span class="w"> </span>--model_id<span class="w"> </span><span class="m">0</span>

<span class="c1"># Feature importance</span>
python<span class="w"> </span>03_compute_feature_importance_agreement.py<span class="w"> </span><span class="m">65</span>
python<span class="w"> </span>03_compute_feature_importance_agreement.py<span class="w"> </span><span class="m">514</span>

<span class="c1"># ARA analysis</span>
python<span class="w"> </span>04_compute_ara.py<span class="w"> </span><span class="m">65</span><span class="w"> </span>--trte_set<span class="w"> </span>train
python<span class="w"> </span>04_compute_ara.py<span class="w"> </span><span class="m">514</span><span class="w"> </span>--trte_set<span class="w"> </span><span class="nb">test</span>
</code></pre></div></p>
<p><strong>Output:</strong> Feature importance rankings showing which chemicals/properties predict microbial growth.</p>
<p><strong>Requirements:</strong> Python 3.12 (CatBoost incompatible with 3.13)</p>
<hr />
<h3 id="5-microgrowlinkservice-public">5. MicroGrowLinkService (PUBLIC)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/MicroGrowLinkService</p>
<p><strong>Purpose:</strong> Gradio web application serving growth media predictions.</p>
<p><strong>Features:</strong>
- User-friendly interface for querying predictions
- Takes taxon as input, returns predicted growth media
- Serves trained MicroGrowLink models</p>
<hr />
<h3 id="6-cmm-ai-private">6. CMM-AI (PRIVATE)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/CMM-AI</p>
<p><strong>Purpose:</strong> Lanthanide bioprocessing data pipeline for rare earth element research.</p>
<p><strong>Scientific Focus:</strong>
- XoxF methanol dehydrogenase (lanthanide-dependent enzymes)
- Methylotrophic bacteria
- Environmental metal cycling
- Siderophore/lanthanophore transport</p>
<p><strong>Data Tables (16 total):</strong></p>
<table>
<thead>
<tr>
<th>Table</th>
<th>Rows</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Taxa &amp; Genomes</td>
<td>211</td>
<td>Bacteria/archaea with NCBI Assembly URLs</td>
</tr>
<tr>
<td>Genes &amp; Proteins</td>
<td>226</td>
<td>UniProt/KEGG with annotations</td>
</tr>
<tr>
<td>Biosamples</td>
<td>132</td>
<td>Environmental samples</td>
</tr>
<tr>
<td>Publications</td>
<td>99</td>
<td>PubMed/bioRxiv literature</td>
</tr>
<tr>
<td>Pathways</td>
<td>77</td>
<td>KEGG/MetaCyc pathways</td>
</tr>
<tr>
<td>Chemicals</td>
<td>66</td>
<td>PubChem/ChEBI compounds</td>
</tr>
<tr>
<td>Media Ingredients</td>
<td>63</td>
<td>Growth media components</td>
</tr>
<tr>
<td>Transcriptomics</td>
<td>29</td>
<td>RNA-seq from SRA/GEO</td>
</tr>
<tr>
<td>Strains</td>
<td>24</td>
<td>Culture collection strains</td>
</tr>
<tr>
<td>Structures</td>
<td>18</td>
<td>PDB + AlphaFold predictions</td>
</tr>
<tr>
<td>Protocols</td>
<td>17</td>
<td>Experimental SOPs</td>
</tr>
<tr>
<td>Datasets</td>
<td>15</td>
<td>Research datasets</td>
</tr>
<tr>
<td>Assays</td>
<td>10</td>
<td>Analytical methods</td>
</tr>
<tr>
<td>Growth Media</td>
<td>8</td>
<td>Complete formulations</td>
</tr>
<tr>
<td>Bioprocesses</td>
<td>5</td>
<td>Experimental conditions</td>
</tr>
<tr>
<td>Screening Results</td>
<td>4</td>
<td>HTS data</td>
</tr>
</tbody>
</table>
<p><strong>Total:</strong> ~1,004 data rows across 16 tables</p>
<p><strong>Makefile Commands:</strong>
<div class="highlight"><pre><span></span><code>make<span class="w"> </span>install<span class="w">          </span><span class="c1"># Install dependencies</span>
make<span class="w"> </span>convert-excel<span class="w">    </span><span class="c1"># Excel → TSV conversion</span>
make<span class="w"> </span>update-all<span class="w">       </span><span class="c1"># Run full pipeline</span>
make<span class="w"> </span>update-genomes<span class="w">   </span><span class="c1"># Extend with NCBI data</span>
make<span class="w"> </span>status<span class="w">          </span><span class="c1"># View pipeline status</span>
</code></pre></div></p>
<hr />
<h3 id="7-mate-llm-private">7. MATE-LLM (PRIVATE)</h3>
<p><strong>Repository:</strong> https://github.com/CultureBotAI/MATE-LLM</p>
<p><strong>Purpose:</strong> Metadata And Table Extraction with LLMs</p>
<p><strong>Description:</strong> Uses large language models to extract structured metadata and tabular data from scientific literature.</p>
<hr />
<h2 id="kgx-data-format">KGX Data Format</h2>
<p>The pipeline uses <strong>KGX (Knowledge Graph Exchange)</strong> TSV format throughout.</p>
<h3 id="nodes-file-nodestsv">Nodes File (<code>nodes.tsv</code>)</h3>
<div class="highlight"><pre><span></span><code>id  category    name    description
CHEBI:16828 biolink:ChemicalEntity  L-tryptophan    Amino acid
EC:4.1.99.1 biolink:Enzyme  tryptophan deaminase    Enzyme
NCBITaxon:123456    biolink:OrganismTaxon   Methylobacterium sp.    Bacterium
DSMZ:1234   biolink:EnvironmentalFeature    Medium 1234 Growth medium
</code></pre></div>
<h3 id="edges-file-edgestsv">Edges File (<code>edges.tsv</code>)</h3>
<div class="highlight"><pre><span></span><code>subject predicate   object  confidence
NCBITaxon:123456    microbe_grows_in_medium DSMZ:1234   0.95
CHEBI:16828 biolink:part_of DSMZ:1234   1.0
</code></pre></div>
<h3 id="key-predicates">Key Predicates</h3>
<table>
<thead>
<tr>
<th>Predicate</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>microbe_grows_in_medium</code></td>
<td><strong>Target relation</strong> - taxon can grow in medium</td>
</tr>
<tr>
<td><code>biolink:part_of</code></td>
<td>Chemical is component of medium</td>
</tr>
<tr>
<td><code>biolink:participates_in</code></td>
<td>Chemical participates in reaction</td>
</tr>
<tr>
<td><code>biolink:capable_of</code></td>
<td>Organism has capability</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="machine-learning-models">Machine Learning Models</h2>
<h3 id="graph-embedding-models-microgrowlink">Graph Embedding Models (MicroGrowLink)</h3>
<table>
<thead>
<tr>
<th>Model</th>
<th>Architecture</th>
<th>Strengths</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>RGT</strong></td>
<td>Relational Graph Transformer</td>
<td>Handles relation types explicitly</td>
</tr>
<tr>
<td><strong>HGT</strong></td>
<td>Heterogeneous Graph Transformer</td>
<td>Attention across node/edge types</td>
</tr>
<tr>
<td><strong>NBFNet</strong></td>
<td>Neural Bellman-Ford</td>
<td>Path-based reasoning</td>
</tr>
</tbody>
</table>
<h3 id="interpretable-models-microbe-rules">Interpretable Models (microbe-rules)</h3>
<table>
<thead>
<tr>
<th>Model</th>
<th>Type</th>
<th>Use Case</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>CatBoost</strong></td>
<td>Gradient Boosting</td>
<td>Feature importance, explainability</td>
</tr>
</tbody>
</table>
<h3 id="export-formats">Export Formats</h3>
<p><strong>RotateE Format:</strong>
- <code>train.txt</code>, <code>val.txt</code>, <code>test.txt</code> (tab-separated triples)
- <code>entity2id.txt</code>, <code>relation2id.txt</code> (vocabularies)</p>
<p><strong>RGT Format:</strong>
- <code>train_graph.json</code>, <code>val_graph.json</code>, <code>test_graph.json</code>
- <code>node_features.json</code>, <code>vocabularies.json</code></p>
<p><strong>A*Net Format:</strong>
- <code>graph.json</code> (with path costs)
- <code>val_queries.json</code>, <code>test_queries.json</code></p>
<hr />
<h2 id="your-role-knowledge-graph-construction">Your Role: Knowledge Graph Construction</h2>
<h3 id="what-you-need-to-build">What You Need to Build</h3>
<p><strong>Primary task:</strong> Create KGX-format edges representing <code>microbe_grows_in_medium</code> relationships.</p>
<h3 id="input-sources">Input Sources</h3>
<ol>
<li><strong>BacDive/MediaDive API</strong> - strain-media associations</li>
<li><strong>MicroMediaParam output</strong> - chemical compositions already mapped to ChEBI</li>
<li><strong>DSMZ strain collections</strong> - culture metadata</li>
</ol>
<h3 id="output-requirements">Output Requirements</h3>
<ol>
<li><strong>nodes.tsv</strong> with:</li>
<li>Taxon nodes (<code>biolink:OrganismTaxon</code>)</li>
<li>Medium nodes (<code>biolink:EnvironmentalFeature</code>)</li>
<li>
<p>Chemical nodes (<code>biolink:ChemicalEntity</code>) - from MicroMediaParam</p>
</li>
<li>
<p><strong>edges.tsv</strong> with:</p>
</li>
<li><code>microbe_grows_in_medium</code> edges (strain → medium)</li>
<li><code>biolink:part_of</code> edges (chemical → medium)</li>
<li>Confidence scores where available</li>
</ol>
<h3 id="tools-available">Tools Available</h3>
<ul>
<li><strong>linkml-store</strong> - mentioned in Slack discussions for KG construction</li>
<li><strong>kg-ai-prep</strong> - for preprocessing your KG before ML training</li>
<li><strong>METPO</strong> (Microbial Phenotype Ontology) - for phenotype annotations</li>
</ul>
<h3 id="workflow">Workflow</h3>
<div class="highlight"><pre><span></span><code>1. Extract strain-media pairs from BacDive/MediaDive
          ↓
2. Map strains to NCBITaxon IDs
          ↓
3. Map media to DSMZ IDs (already in MicroMediaParam)
          ↓
4. Create KGX nodes.tsv and edges.tsv
          ↓
5. Run kg-ai-prep for filtering/splitting
          ↓
6. Hand off to Marcin for MicroGrowLink training
</code></pre></div>
<hr />
<h2 id="anticipated-payoffs">Anticipated Payoffs</h2>
<h3 id="scientific-goals">Scientific Goals</h3>
<ol>
<li><strong>Predict growth media for unstudied microbes</strong></li>
<li>Given a new taxon, suggest media it might grow in</li>
<li>
<p>Reduce trial-and-error in cultivation</p>
</li>
<li>
<p><strong>Explain predictions</strong></p>
</li>
<li>Identify which chemicals drive growth predictions</li>
<li>
<p>Understand nutritional requirements</p>
</li>
<li>
<p><strong>Design novel media</strong></p>
</li>
<li>Potentially design new media for difficult-to-culture organisms</li>
<li>Optimize existing media formulations</li>
</ol>
<h3 id="technical-benefits">Technical Benefits</h3>
<ul>
<li><strong>20-40% improvement</strong> in link prediction accuracy (with proper KG filtering)</li>
<li><strong>Reduced training time</strong> by removing irrelevant graph structure</li>
<li><strong>Interpretable results</strong> via CatBoost feature importance</li>
</ul>
<h3 id="broader-impact">Broader Impact</h3>
<ul>
<li>Advance cultivation of rare earth element-processing microbes</li>
<li>Support lanthanide bioprocessing research</li>
<li>Enable culture-independent predictions for environmental samples</li>
</ul>
<hr />
<h2 id="repository-access">Repository Access</h2>
<h3 id="public-repositories-direct-access">Public Repositories (Direct Access)</h3>
<table>
<thead>
<tr>
<th>Repository</th>
<th>URL</th>
</tr>
</thead>
<tbody>
<tr>
<td>MicroMediaParam</td>
<td>https://github.com/CultureBotAI/MicroMediaParam</td>
</tr>
<tr>
<td>microbe-rules</td>
<td>https://github.com/CultureBotAI/microbe-rules</td>
</tr>
<tr>
<td>MicroGrowLinkService</td>
<td>https://github.com/CultureBotAI/MicroGrowLinkService</td>
</tr>
<tr>
<td>PFAS-AI</td>
<td>https://github.com/CultureBotAI/PFAS-AI</td>
</tr>
<tr>
<td>auto-term-catalog</td>
<td>https://github.com/CultureBotAI/auto-term-catalog</td>
</tr>
<tr>
<td>assay-metadata</td>
<td>https://github.com/CultureBotAI/assay-metadata</td>
</tr>
<tr>
<td>eggnog_runner</td>
<td>https://github.com/CultureBotAI/eggnog_runner</td>
</tr>
</tbody>
</table>
<h3 id="private-repositories-gh-cli-access">Private Repositories (gh CLI Access)</h3>
<table>
<thead>
<tr>
<th>Repository</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>kg-ai-prep</td>
<td>KG preprocessing for ML</td>
</tr>
<tr>
<td>MicroGrowLink</td>
<td>Core ML training code</td>
</tr>
<tr>
<td>CMM-AI</td>
<td>Lanthanide data pipeline</td>
</tr>
<tr>
<td>MATE-LLM</td>
<td>LLM-based extraction</td>
</tr>
<tr>
<td>KG-Microbe-search</td>
<td>KG search functionality</td>
</tr>
<tr>
<td>kg-microbe-projects</td>
<td>Project configurations</td>
</tr>
<tr>
<td>eggnogtable</td>
<td>EggNOG data tables</td>
</tr>
</tbody>
</table>
<p><strong>Access via gh CLI:</strong>
<div class="highlight"><pre><span></span><code><span class="c1"># View README</span>
gh<span class="w"> </span>api<span class="w"> </span>repos/CultureBotAI/kg-ai-prep/readme<span class="w"> </span>--jq<span class="w"> </span><span class="s1">&#39;.content&#39;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>base64<span class="w"> </span>-d

<span class="c1"># List files</span>
gh<span class="w"> </span>api<span class="w"> </span>repos/CultureBotAI/kg-ai-prep/contents<span class="w"> </span>--jq<span class="w"> </span><span class="s1">&#39;.[].name&#39;</span>

<span class="c1"># Clone (if needed)</span>
gh<span class="w"> </span>repo<span class="w"> </span>clone<span class="w"> </span>CultureBotAI/kg-ai-prep
</code></pre></div></p>
<hr />
<h2 id="quick-reference">Quick Reference</h2>
<h3 id="key-target-relation">Key Target Relation</h3>
<div class="highlight"><pre><span></span><code>microbe_grows_in_medium
</code></pre></div>
<h3 id="kgx-node-categories">KGX Node Categories</h3>
<div class="highlight"><pre><span></span><code>biolink:OrganismTaxon      # Microbial taxa
biolink:EnvironmentalFeature  # Growth media
biolink:ChemicalEntity     # Chemical compounds
biolink:Enzyme             # Enzymes
biolink:MolecularActivity  # Activities
</code></pre></div>
<h3 id="filtering-commands-kg-ai-prep">Filtering Commands (kg-ai-prep)</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Basic preprocessing</span>
kg-ai-prep<span class="w"> </span>preprocess<span class="w"> </span>nodes.tsv<span class="w"> </span>edges.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--remove-singletons<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--min-degree<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--giant-component-only<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--target-relation<span class="w"> </span><span class="s2">&quot;microbe_grows_in_medium&quot;</span>

<span class="c1"># Model-specific export</span>
--export-rotate<span class="w">   </span><span class="c1"># For RotatE/TransE</span>
--export-rgt<span class="w">      </span><span class="c1"># For RGT</span>
--export-astar<span class="w">    </span><span class="c1"># For A*Net/NBFNet</span>
</code></pre></div>
<h3 id="related-slack-workspaces">Related Slack Workspaces</h3>
<ul>
<li><strong>NMDC</strong> (<code>slack-nmdc</code>)</li>
<li><strong>BerkeleyBOP</strong> (<code>slack-berkeleybop</code>)</li>
<li><strong>BER-CMM</strong> (<code>slack-ber-cmm</code>) - #cmm-ai, #all-ber-cmm-pilot-project</li>
</ul>
<h3 id="related-github-repos">Related GitHub Repos</h3>
<ul>
<li><code>turbomam/issues</code> - Issue #3 (cmm-ai project) and #16 (Slack MCP guide)</li>
<li><code>CultureBotAI/*</code> - All ML and KG repos</li>
</ul>
<hr />
<p><em>Document generated from CultureBotAI repository analysis via gh CLI.</em></p>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>