
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/CMM-AI_Analysis/02_data_model_analysis/">
      
      
      
      
        
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Chapter 2: Data Model Analysis - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#chapter-2-data-model-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Chapter 2: Data Model Analysis
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-two-data-systems" class="md-nav__link">
    <span class="md-ellipsis">
      
        The Two Data Systems
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="The Two Data Systems">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#system-a-google-sheets-xlsx-tsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        System A: Google Sheets -&gt; XLSX -&gt; TSV
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#system-b-python-extend-scripts" class="md-nav__link">
    <span class="md-ellipsis">
      
        System B: Python Extend Scripts
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#case-study-media_ingredientstsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        Case Study: media_ingredients.tsv
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Case Study: media_ingredients.tsv">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-two-versions" class="md-nav__link">
    <span class="md-ellipsis">
      
        The Two Versions
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#where-did-mp-medium-come-from" class="md-nav__link">
    <span class="md-ellipsis">
      
        Where Did MP Medium Come From?
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#column-analysis-media_ingredientstsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        Column Analysis: media_ingredients.tsv
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#foreign-key-analysis-media_ingredientstsv" class="md-nav__link">
    <span class="md-ellipsis">
      
        Foreign Key Analysis: media_ingredients.tsv
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Foreign Key Analysis: media_ingredients.tsv">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#outbound-media_ingredients-references-other-tables" class="md-nav__link">
    <span class="md-ellipsis">
      
        OUTBOUND (media_ingredients references other tables)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#inbound-other-tables-reference-media_ingredients" class="md-nav__link">
    <span class="md-ellipsis">
      
        INBOUND (other tables reference media_ingredients)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#missing-relationships-should-exist-based-on-proposal" class="md-nav__link">
    <span class="md-ellipsis">
      
        Missing Relationships (Should Exist Based on Proposal)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#denormalization-issue" class="md-nav__link">
    <span class="md-ellipsis">
      
        Denormalization Issue
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#mediaingredients-data-model-design-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Media/Ingredients Data Model Design Analysis
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Media/Ingredients Data Model Design Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#current-state-assessment" class="md-nav__link">
    <span class="md-ellipsis">
      
        Current State Assessment
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Current State Assessment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-normalization-violations" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Normalization Violations
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-ad-hoc-ingredient_id-namespace" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Ad-hoc ingredient_id Namespace
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-context-dependent-roles" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. Context-Dependent Roles
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#microbe-media-relationship-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Microbe-Media Relationship Analysis
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Microbe-Media Relationship Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#question-do-any-data-tables-assert-relationships-between-microbes-and-media" class="md-nav__link">
    <span class="md-ellipsis">
      
        Question: Do any data tables assert relationships between microbes and media?
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#what-currently-exists" class="md-nav__link">
    <span class="md-ellipsis">
      
        What Currently EXISTS
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#recommended-fix-new-junction-table" class="md-nav__link">
    <span class="md-ellipsis">
      
        Recommended Fix: New Junction Table
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#linkml-schema-and-data-table-relationships" class="md-nav__link">
    <span class="md-ellipsis">
      
        LinkML Schema and Data Table Relationships
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="LinkML Schema and Data Table Relationships">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#data-flow-architecture" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Flow Architecture
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#schema-class-to-tsv-file-mapping" class="md-nav__link">
    <span class="md-ellipsis">
      
        Schema Class to TSV File Mapping
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bioregistry-vs-data-repository-clarification" class="md-nav__link">
    <span class="md-ellipsis">
      
        Bioregistry vs Data Repository Clarification
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Bioregistry vs Data Repository Clarification">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#common-confusion" class="md-nav__link">
    <span class="md-ellipsis">
      
        Common Confusion
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mediadive-specifically" class="md-nav__link">
    <span class="md-ellipsis">
      
        MediaDive Specifically
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#best-practices-mediadive-reference" class="md-nav__link">
    <span class="md-ellipsis">
      
        Best Practices: MediaDive Reference
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="chapter-2-data-model-analysis">Chapter 2: Data Model Analysis</h1>
<p><a href="../00_index/">&lt;- Back to Index</a> | <a href="../01_project_context/">Previous: Project Context</a> | <a href="../03_data_quality_issues/">Next: Data Quality Issues -&gt;</a></p>
<hr />
<h2 id="the-two-data-systems">The Two Data Systems</h2>
<h3 id="system-a-google-sheets-xlsx-tsv">System A: Google Sheets -&gt; XLSX -&gt; TSV</h3>
<div class="highlight"><pre><span></span><code>Collaborators type/paste into Google Sheets
    | (manual process, undocumented)
    v
Marcin downloads as XLSX
    |
    v
data/sheet/BER CMM Data for AI.xlsx (254 KB, Dec 4 2025)
    |
    v
make convert-excel (runs src/convert_sheets.py)
    |
    v
data/txt/sheet/BER_CMM_Data_for_AI_*.tsv (16 files)
</code></pre></div>
<p><strong>Characteristics:</strong>
- Contains manual entries, possibly LLM-generated content
- Has UTF-8 encoding issues (mojibake: see <a href="../03_data_quality_issues/">Chapter 3</a>)
- Overwrites previous TSV files on each conversion
- No version control on Google Sheet edits
- No record of who entered what or when</p>
<h3 id="system-b-python-extend-scripts">System B: Python Extend Scripts</h3>
<div class="highlight"><pre><span></span><code>Seed data (hardcoded in Python or from TSV)
    |
    v
API queries (NCBI, KEGG, UniProt, PDB, KG-Microbe)
    |
    v
src/extend_*.py scripts
    |
    v
data/txt/sheet/extended/BER_CMM_Data_for_AI_*_extended.tsv (16 files)
</code></pre></div>
<p><strong>Characteristics:</strong>
- Reproducible (run <code>make update-*</code> commands)
- Clean UTF-8 encoding
- Outputs to separate <code>extended/</code> subdirectory
- Source provenance tracked (mostly)
- Some data hardcoded in Python (e.g., <code>CURATED_MEDIA</code> dict in media_search.py)</p>
<hr />
<h2 id="case-study-media_ingredientstsv">Case Study: media_ingredients.tsv</h2>
<h3 id="the-two-versions">The Two Versions</h3>
<table>
<thead>
<tr>
<th>Aspect</th>
<th>Google Sheet Version</th>
<th>Extended Version</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Location</strong></td>
<td><code>data/txt/sheet/BER_CMM_Data_for_AI_media_ingredients.tsv</code></td>
<td><code>data/txt/sheet/extended/BER_CMM_Data_for_AI_media_ingredients_extended.tsv</code></td>
</tr>
<tr>
<td><strong>Date</strong></td>
<td>Dec 4, 2025</td>
<td>Nov 13, 2025</td>
</tr>
<tr>
<td><strong>Size</strong></td>
<td>12 KB (64 rows)</td>
<td>9 KB</td>
</tr>
<tr>
<td><strong>Media count</strong></td>
<td>8 media</td>
<td>7 media</td>
</tr>
<tr>
<td><strong>Has MP medium?</strong></td>
<td>Yes (with detailed description)</td>
<td>No</td>
</tr>
<tr>
<td><strong>UTF-8 clean?</strong></td>
<td>No (mojibake present)</td>
<td>Yes</td>
</tr>
<tr>
<td><strong>kg_microbe_nodes</strong></td>
<td>Populated</td>
<td>Populated (different values)</td>
</tr>
</tbody>
</table>
<h3 id="where-did-mp-medium-come-from">Where Did MP Medium Come From?</h3>
<p>The Google Sheet version contains "MP medium (PIPES-buffered methylotroph minimal medium)" with an elaborate description:</p>
<blockquote>
<p>"Defined minimal medium composed of a PIPES buffering system, phosphorus salts, major inorganic salts, and a trace metals mix..."</p>
</blockquote>
<p>But <strong>MP medium does not exist in the Python code</strong> (<code>src/media_search.py</code> -&gt; <code>CURATED_MEDIA</code> dict only has 7 media: ATCC:1306, NMS, AMS, DSMZ:88, LB, R2A, MPYG).</p>
<p><strong>Possible origins (unknown which is true):</strong>
1. Someone manually typed it into Google Sheets
2. An LLM generated it and someone pasted it
3. A different script generated it (not found in codebase)
4. Copied from literature</p>
<p><strong>This cannot be determined from available evidence.</strong></p>
<h3 id="column-analysis-media_ingredientstsv">Column Analysis: media_ingredients.tsv</h3>
<table>
<thead>
<tr>
<th>Column</th>
<th>Purpose</th>
<th>Populated</th>
<th>Issues</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>ingredient_id</code></td>
<td>Primary key</td>
<td>100%</td>
<td>Pattern: <code>{media_id}_ING_{NNN}</code></td>
</tr>
<tr>
<td><code>ingredient_name</code></td>
<td>Chemical name</td>
<td>100%</td>
<td></td>
</tr>
<tr>
<td><code>media_id</code></td>
<td>FK -&gt; growth_media</td>
<td>100%</td>
<td></td>
</tr>
<tr>
<td><code>media_name</code></td>
<td>Denormalized</td>
<td>100%</td>
<td>Violates 3NF (redundant)</td>
</tr>
<tr>
<td><code>ontology_id</code></td>
<td>CHEBI ID</td>
<td>~75%</td>
<td>Missing for complex mixtures</td>
</tr>
<tr>
<td><code>ontology_label</code></td>
<td>CHEBI label</td>
<td>~75%</td>
<td></td>
</tr>
<tr>
<td><code>chemical_formula</code></td>
<td>Molecular formula</td>
<td>~70%</td>
<td>UTF-8 corruption</td>
</tr>
<tr>
<td><code>concentration</code></td>
<td>Numeric value</td>
<td>100%</td>
<td></td>
</tr>
<tr>
<td><code>unit</code></td>
<td>g/L, mL/L, mM, etc.</td>
<td>100%</td>
<td>Inconsistent units</td>
</tr>
<tr>
<td><code>role</code></td>
<td>Functional category</td>
<td>100%</td>
<td>Undocumented vocabulary</td>
</tr>
<tr>
<td><code>kg_microbe_nodes</code></td>
<td>KG linkage</td>
<td>~80%</td>
<td>Mixed ID types</td>
</tr>
<tr>
<td><code>notes</code></td>
<td>Comments</td>
<td>0%</td>
<td>Unused</td>
</tr>
<tr>
<td><code>source</code></td>
<td>Provenance</td>
<td>100%</td>
<td>All rows say "extend"</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="foreign-key-analysis-media_ingredientstsv">Foreign Key Analysis: media_ingredients.tsv</h2>
<h3 id="outbound-media_ingredients-references-other-tables">OUTBOUND (media_ingredients references other tables)</h3>
<table>
<thead>
<tr>
<th>Column</th>
<th>References</th>
<th>Target Table</th>
<th>Target Column</th>
<th>Status</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>media_id</strong></td>
<td>-&gt;</td>
<td><strong>growth_media</strong></td>
<td><code>media_id</code></td>
<td>Valid FK - All 8 media_ids exist</td>
</tr>
<tr>
<td><strong>ontology_id</strong> (CHEBI:*)</td>
<td>-&gt;</td>
<td><strong>chemicals</strong></td>
<td><code>chebi_id</code></td>
<td>Partial overlap</td>
</tr>
</tbody>
</table>
<h3 id="inbound-other-tables-reference-media_ingredients">INBOUND (other tables reference media_ingredients)</h3>
<table>
<thead>
<tr>
<th>Source Table</th>
<th>Source Column</th>
<th>References</th>
<th>Status</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>bioprocesses</strong></td>
<td><code>growth_conditions</code></td>
<td>mentions media names</td>
<td>Soft reference - Text, not FK</td>
</tr>
<tr>
<td><strong>strains</strong></td>
<td><code>growth_requirements</code></td>
<td>None</td>
<td>Missing - Should link</td>
</tr>
</tbody>
</table>
<h3 id="missing-relationships-should-exist-based-on-proposal">Missing Relationships (Should Exist Based on Proposal)</h3>
<table>
<thead>
<tr>
<th>Relationship</th>
<th>Why It's Needed</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>strains</strong> -&gt; <strong>growth_media</strong></td>
<td>Task 2.1: "Strains...will be grown comparing use of REE in soluble vs insoluble form"</td>
</tr>
<tr>
<td><strong>screening_results</strong> -&gt; <strong>growth_media</strong></td>
<td>Task 1.3: HTP screening needs to record which media was used</td>
</tr>
<tr>
<td><strong>bioprocesses</strong> -&gt; <strong>growth_media</strong></td>
<td>Should be a proper FK, not text like "LB medium"</td>
</tr>
</tbody>
</table>
<h3 id="denormalization-issue">Denormalization Issue</h3>
<p>The <code>media_name</code> column in media_ingredients is <strong>redundant</strong> - it duplicates data from growth_media. This violates 3NF:</p>
<div class="highlight"><pre><span></span><code>media_ingredients.media_id -&gt; growth_media.media_id -&gt; growth_media.media_name
</code></pre></div>
<p>If <code>media_name</code> changes in growth_media, it must be updated in 64 rows of media_ingredients.</p>
<hr />
<h2 id="mediaingredients-data-model-design-analysis">Media/Ingredients Data Model Design Analysis</h2>
<h3 id="current-state-assessment">Current State Assessment</h3>
<h4 id="1-normalization-violations">1. Normalization Violations</h4>
<p><strong>Current structure</strong> (denormalized):
<div class="highlight"><pre><span></span><code>media_ingredients.tsv contains:
- ingredient_id (PK)
- ingredient_name
- media_id (FK)
- media_name  &lt;- REDUNDANT (violates 3NF)
- ontology_id
- ontology_label
- chemical_formula
- concentration
- unit
- role
- kg_microbe_nodes
</code></pre></div></p>
<p><strong>Recommended 3-table structure</strong>:</p>
<table>
<thead>
<tr>
<th>Table</th>
<th>Purpose</th>
<th>Columns</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>growth_media</strong></td>
<td>Media characteristics</td>
<td>media_id (PK), media_name, ph, sterilization, target_organisms, references</td>
</tr>
<tr>
<td><strong>ingredients</strong></td>
<td>Canonical ingredient definitions</td>
<td>ingredient_id (PK, use CHEBI where available), name, formula, cas_number, default_roles[]</td>
</tr>
<tr>
<td><strong>media_compositions</strong></td>
<td>Junction with concentrations</td>
<td>media_id (FK), ingredient_id (FK), concentration, unit, role_in_context, preparation_notes</td>
</tr>
</tbody>
</table>
<h4 id="2-ad-hoc-ingredient_id-namespace">2. Ad-hoc ingredient_id Namespace</h4>
<p><strong>Current</strong>: <code>ATCC1306_001</code>, <code>NMS_001</code> - embeds media context into ingredient IDs</p>
<p><strong>Problem</strong>: Same chemical (MgSO4<em>7H2O) gets multiple IDs:
- <code>AMS_ING_001</code> = MgSO4</em>7H2O in AMS medium
- <code>ATCC:1306_ING_003</code> = MgSO4<em>7H2O in ATCC:1306 medium
- <code>NMS_ING_001</code> = MgSO4</em>7H2O in NMS medium</p>
<p><strong>Recommendation</strong>: Use CHEBI as canonical identifier where available. For compounds without CHEBI:
- Create a project namespace: <code>CMM:ING-0001</code>, <code>CMM:ING-0002</code>, etc.
- Document what each local ID represents
- Store the mapping to CHEBI/CAS when added later</p>
<h4 id="3-context-dependent-roles">3. Context-Dependent Roles</h4>
<p><strong>Yes, ingredients can have different roles depending on context:</strong></p>
<table>
<thead>
<tr>
<th>Ingredient</th>
<th>Context 1</th>
<th>Context 2</th>
</tr>
</thead>
<tbody>
<tr>
<td>K2HPO4 (dipotassium phosphate)</td>
<td><strong>buffer</strong> (pH stabilization)</td>
<td><strong>mineral</strong> (phosphorus source)</td>
</tr>
<tr>
<td>(NH4)2SO4 (ammonium sulfate)</td>
<td><strong>nitrogen source</strong></td>
<td><strong>sulfur source</strong></td>
</tr>
<tr>
<td>FeSO4*7H2O (ferrous sulfate)</td>
<td><strong>trace element</strong> (minimal media)</td>
<td><strong>iron source</strong> (iron-limited experiments)</td>
</tr>
<tr>
<td>NaCl (sodium chloride)</td>
<td><strong>salt</strong> (osmotic balance)</td>
<td><strong>mineral</strong> (Na+ source)</td>
</tr>
</tbody>
</table>
<p><strong>Current model assigns ONE role per ingredient per media</strong> - this loses information.</p>
<p><strong>Recommendation</strong>:
- Allow multivalued roles: <code>roles: ["buffer", "phosphorus source"]</code>
- Or create separate columns: <code>primary_role</code>, <code>secondary_roles</code></p>
<hr />
<h2 id="microbe-media-relationship-analysis">Microbe-Media Relationship Analysis</h2>
<h3 id="question-do-any-data-tables-assert-relationships-between-microbes-and-media">Question: Do any data tables assert relationships between microbes and media?</h3>
<p><strong>Short answer: The relationships are WEAK and mostly implicit. No explicit FK relationships exist.</strong></p>
<h3 id="what-currently-exists">What Currently EXISTS</h3>
<table>
<thead>
<tr>
<th>Table</th>
<th>Column</th>
<th>Relationship to Media</th>
<th>Type</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>growth_media</strong></td>
<td><code>target_organisms</code></td>
<td>Lists genera like "Methylobacterium; Methylorubrum; methylotrophs"</td>
<td>Soft/implicit - free text</td>
</tr>
<tr>
<td><strong>bioprocesses</strong></td>
<td><code>growth_conditions</code></td>
<td>Text like "LB medium, aerobic, 200 rpm shaking"</td>
<td>Soft/implicit - text, not FK</td>
</tr>
<tr>
<td><strong>strains</strong></td>
<td><code>growth_requirements</code></td>
<td>Lists phenotypes like "urease; catalase; mesophilic"</td>
<td>NOT media - enzymatic traits</td>
</tr>
<tr>
<td><strong>screening_results</strong></td>
<td>(none)</td>
<td>No media column at all</td>
<td>Missing</td>
</tr>
</tbody>
</table>
<h3 id="recommended-fix-new-junction-table">Recommended Fix: New Junction Table</h3>
<p>A proper design would add a <code>strain_media_growth</code> junction table:</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>strain_id</code></td>
<td>FK -&gt; strains</td>
<td>Which strain</td>
</tr>
<tr>
<td><code>media_id</code></td>
<td>FK -&gt; growth_media</td>
<td>Which medium</td>
</tr>
<tr>
<td><code>growth_quality</code></td>
<td>enum</td>
<td>none, poor, moderate, good, optimal</td>
</tr>
<tr>
<td><code>doubling_time_hours</code></td>
<td>float</td>
<td>Growth rate metric</td>
</tr>
<tr>
<td><code>max_od600</code></td>
<td>float</td>
<td>Maximum optical density achieved</td>
</tr>
<tr>
<td><code>optimal_temperature</code></td>
<td>float</td>
<td>Optimal growth temperature</td>
</tr>
<tr>
<td><code>optimal_ph</code></td>
<td>float</td>
<td>Optimal pH</td>
</tr>
<tr>
<td><code>ree_supplementation</code></td>
<td>string</td>
<td>REE type and concentration if applicable</td>
</tr>
<tr>
<td><code>notes</code></td>
<td>string</td>
<td>Additional observations</td>
</tr>
<tr>
<td><code>source</code></td>
<td>string</td>
<td>Literature, experiment, DSMZ catalog, etc.</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="linkml-schema-and-data-table-relationships">LinkML Schema and Data Table Relationships</h2>
<h3 id="data-flow-architecture">Data Flow Architecture</h3>
<div class="highlight"><pre><span></span><code>+-----------------------------------------------------------------------------+
|                           DATA SOURCES                                       |
+-----------------------------------------------------------------------------+
|  Google Sheets -&gt; XLSX -&gt; TSV (data/txt/sheet/*.tsv)                        |
|  Python extend scripts -&gt; Extended TSV (data/txt/sheet/extended/*_extended.tsv)|
+-----------------------------------------------------------------------------+
                                    |
                                    v
+-----------------------------------------------------------------------------+
|                        src/tsv_to_linkml.py                                  |
|  Converts TSV files -&gt; LinkML YAML format                                    |
+-----------------------------------------------------------------------------+
                                    |
                                    v
+-----------------------------------------------------------------------------+
|                     data/linkml_database.yaml                                |
|  5,098 lines of YAML data                                                    |
+-----------------------------------------------------------------------------+
                                    |
                                    v
+-----------------------------------------------------------------------------+
|                       linkml-validate                                        |
|  Validates YAML against schema/lanthanide_bioprocessing.yaml                |
+-----------------------------------------------------------------------------+
</code></pre></div>
<h3 id="schema-class-to-tsv-file-mapping">Schema Class to TSV File Mapping</h3>
<table>
<thead>
<tr>
<th>Schema Class</th>
<th>TSV File</th>
<th>Key Fields</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>GenomeRecord</code></td>
<td><code>taxa_and_genomes.tsv</code></td>
<td>scientific_name, ncbi_taxon_id, genome_identifier</td>
</tr>
<tr>
<td><code>BiosampleRecord</code></td>
<td><code>biosamples.tsv</code></td>
<td>sample_id, organism, download_url</td>
</tr>
<tr>
<td><code>PathwayRecord</code></td>
<td><code>pathways.tsv</code></td>
<td>pathway_id, pathway_name, genes</td>
</tr>
<tr>
<td><code>GeneProteinRecord</code></td>
<td><code>genes_and_proteins.tsv</code></td>
<td>gene_protein_id, annotation, ec_number</td>
</tr>
<tr>
<td><code>MacromolecularStructureRecord</code></td>
<td><code>macromolecular_structures.tsv</code></td>
<td>pdb_id, method, resolution</td>
</tr>
<tr>
<td><code>PublicationRecord</code></td>
<td><code>publications.tsv</code></td>
<td>url, title, journal, year, pmid</td>
</tr>
<tr>
<td><code>DatasetRecord</code></td>
<td><code>datasets.tsv</code></td>
<td>dataset_name, data_type, url</td>
</tr>
<tr>
<td><code>ChemicalCompoundRecord</code></td>
<td><code>chemicals.tsv</code></td>
<td>chemical_id, chebi_id, compound_type</td>
</tr>
<tr>
<td><code>AssayMeasurementRecord</code></td>
<td><code>assays.tsv</code></td>
<td>assay_id, assay_type, detection_method</td>
</tr>
<tr>
<td><code>BioprocessConditionsRecord</code></td>
<td><code>bioprocesses.tsv</code></td>
<td>process_id, strain_used, pH, temperature</td>
</tr>
<tr>
<td><code>ScreeningResultRecord</code></td>
<td><code>screening_results.tsv</code></td>
<td>experiment_id, strain_barcode, hit_classification</td>
</tr>
<tr>
<td><code>ProtocolRecord</code></td>
<td><code>protocols.tsv</code></td>
<td>protocol_id, protocol_type, protocol_doi</td>
</tr>
<tr>
<td><code>TranscriptomicsRecord</code></td>
<td><code>transcriptomics.tsv</code></td>
<td>accession, experiment_type</td>
</tr>
<tr>
<td><code>StrainRecord</code></td>
<td><code>strains.tsv</code></td>
<td>strain_id, source_collection</td>
</tr>
<tr>
<td><code>GrowthMediaRecord</code></td>
<td><code>growth_media.tsv</code></td>
<td>media_id, media_name, ph</td>
</tr>
<tr>
<td><code>MediaIngredientRecord</code></td>
<td><code>media_ingredients.tsv</code></td>
<td>ingredient_id, media_id, concentration, role</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="bioregistry-vs-data-repository-clarification">Bioregistry vs Data Repository Clarification</h2>
<h3 id="common-confusion">Common Confusion</h3>
<p>Marcin mentioned "mediadive data are in bioregistry" - this is confusing because:</p>
<table>
<thead>
<tr>
<th>Tool</th>
<th>What It Is</th>
<th>What It Contains</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><a href="https://bioregistry.io/">Bioregistry</a></strong></td>
<td>Prefix registry/resolver</td>
<td>Metadata about databases (prefix patterns, URLs, contacts)</td>
</tr>
<tr>
<td><strong><a href="https://github.com/biopragmatics/pyobo">PyOBO</a></strong></td>
<td>Data converter</td>
<td>Code to download and convert databases to OBO/OWL</td>
</tr>
<tr>
<td><strong><a href="https://github.com/biopragmatics/obo-db-ingest">obo-db-ingest</a></strong></td>
<td>Data repository</td>
<td>Actual converted OBO/OWL files from PyOBO</td>
</tr>
</tbody>
</table>
<h3 id="mediadive-specifically">MediaDive Specifically</h3>
<ul>
<li><strong>Bioregistry HAS</strong>: Prefix registration for <code>mediadive.medium</code> and <code>mediadive.solution</code></li>
<li><code>mediadive.medium:77</code> resolves to <code>https://mediadive.dsmz.de/medium/77</code></li>
<li><strong>Bioregistry DOES NOT HAVE</strong>: The actual MediaDive data</li>
<li><strong>obo-db-ingest</strong>: Does NOT currently contain MediaDive</li>
<li><strong>Raw data lives at</strong>: https://mediadive.dsmz.de/ with RDF export available</li>
</ul>
<h3 id="best-practices-mediadive-reference">Best Practices: MediaDive Reference</h3>
<p><strong>MediaDive</strong> (https://mediadive.dsmz.de/) is the gold standard for microbial growth media databases.</p>
<p><strong>Their data model</strong>:</p>
<table>
<thead>
<tr>
<th>Entity</th>
<th>Count</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Media</td>
<td>3,327</td>
<td>Complete formulations with metadata</td>
</tr>
<tr>
<td>Solutions</td>
<td>5,844</td>
<td>Stock solutions, trace element mixes, vitamin mixes</td>
</tr>
<tr>
<td>Ingredients</td>
<td>1,235</td>
<td>Canonical chemical compounds</td>
</tr>
</tbody>
</table>
<p><strong>Key design principles from MediaDive</strong>:</p>
<ol>
<li><strong>Hierarchical ingredients</strong>: Solutions can contain other solutions</li>
<li><strong>Canonical ingredient registry</strong>: Each unique chemical has ONE record</li>
<li><strong>Concentration in junction table</strong>: Amount, unit, preparation notes</li>
<li><strong>SPARQL endpoint</strong>: Machine-readable access for integration</li>
<li><strong>Provenance</strong>: Each medium has source references</li>
</ol>
<hr />
<p><a href="../00_index/">&lt;- Back to Index</a> | <a href="../01_project_context/">Previous: Project Context</a> | <a href="../03_data_quality_issues/">Next: Data Quality Issues -&gt;</a></p>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>