
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/curie_prefix_strategy/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>CURIE and Prefix Strategy - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#curie-and-prefix-strategy" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CURIE and Prefix Strategy
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-problem" class="md-nav__link">
    <span class="md-ellipsis">
      
        The Problem
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#entity-types-inventory" class="md-nav__link">
    <span class="md-ellipsis">
      
        Entity Types Inventory
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Entity Types Inventory">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#from-external-databases-have-their-own-ids" class="md-nav__link">
    <span class="md-ellipsis">
      
        From External Databases (Have Their Own IDs)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#entities-we-may-need-to-mint-ids-for" class="md-nav__link">
    <span class="md-ellipsis">
      
        Entities We May Need to Mint IDs For
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prefix-strategy-options" class="md-nav__link">
    <span class="md-ellipsis">
      
        Prefix Strategy Options
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Prefix Strategy Options">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#option-1-many-type-specific-prefixes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option 1: Many Type-Specific Prefixes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#option-2-single-database-prefix-typed-local-ids" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option 2: Single Database Prefix + Typed Local IDs
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#option-3-single-database-prefix-numeric-ids-let-database-resolve" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option 3: Single Database Prefix + Numeric IDs (Let Database Resolve)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#option-4-project-prefix-for-minted-ids-database-prefixes-for-external" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option 4: Project Prefix for Minted IDs + Database Prefixes for External
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#option-5-hybrid-project-prefix-with-type-codes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Option 5: Hybrid - Project Prefix with Type Codes
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#questions-for-community" class="md-nav__link">
    <span class="md-ellipsis">
      
        Questions for Community
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Questions for Community">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#for-bioregistry" class="md-nav__link">
    <span class="md-ellipsis">
      
        For Bioregistry
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#for-biolinkkgx" class="md-nav__link">
    <span class="md-ellipsis">
      
        For Biolink/KGX
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#for-w3idorg" class="md-nav__link">
    <span class="md-ellipsis">
      
        For w3id.org
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#two-different-problems-terms-vs-instances" class="md-nav__link">
    <span class="md-ellipsis">
      
        Two Different Problems: Terms vs Instances
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#metpo-for-domain-specific-terms" class="md-nav__link">
    <span class="md-ellipsis">
      
        METPO for Domain-Specific Terms
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="METPO for Domain-Specific Terms">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#using-metpo-for-cmm-categories-and-predicates" class="md-nav__link">
    <span class="md-ellipsis">
      
        Using METPO for CMM Categories and Predicates
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-instance-id-problem-still-unsolved" class="md-nav__link">
    <span class="md-ellipsis">
      
        The Instance ID Problem (Still Unsolved)
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#proposed-strategy-draft" class="md-nav__link">
    <span class="md-ellipsis">
      
        Proposed Strategy (Draft)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Proposed Strategy (Draft)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#phase-1-use-whats-registered" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 1: Use What's Registered
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-2-register-database-prefixes" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 2: Register Database Prefixes
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#phase-3-register-project-prefix-for-minted-ids" class="md-nav__link">
    <span class="md-ellipsis">
      
        Phase 3: Register Project Prefix for Minted IDs
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#local-id-format-for-cmm" class="md-nav__link">
    <span class="md-ellipsis">
      
        Local ID Format for cmm:
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#registry-requirements" class="md-nav__link">
    <span class="md-ellipsis">
      
        Registry Requirements
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#research-findings" class="md-nav__link">
    <span class="md-ellipsis">
      
        Research Findings
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Research Findings">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#mediadive-url-structure-verified-2026-01-08" class="md-nav__link">
    <span class="md-ellipsis">
      
        MediaDive URL Structure (Verified 2026-01-08)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bacdive-url-structure-comparison" class="md-nav__link">
    <span class="md-ellipsis">
      
        BacDive URL Structure (Comparison)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#togomedium-url-structure-from-docstogomediummd" class="md-nav__link">
    <span class="md-ellipsis">
      
        TogoMedium URL Structure (from docs/togomedium.md)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#identifier-registry-landscape" class="md-nav__link">
    <span class="md-ellipsis">
      
        Identifier Registry Landscape
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Identifier Registry Landscape">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#registries-to-consider" class="md-nav__link">
    <span class="md-ellipsis">
      
        Registries to Consider
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#project-prefix-status-verified-2026-01-08" class="md-nav__link">
    <span class="md-ellipsis">
      
        Project Prefix Status (Verified 2026-01-08)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prefix-naming-trade-offs" class="md-nav__link">
    <span class="md-ellipsis">
      
        Prefix Naming Trade-offs
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#custom-prefixes-in-use-audit-2026-01-08" class="md-nav__link">
    <span class="md-ellipsis">
      
        Custom Prefixes in Use (Audit 2026-01-08)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Custom Prefixes in Use (Audit 2026-01-08)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-constants-constantspy" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Constants (constants.py)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#registered-prefixes-for-comparison" class="md-nav__link">
    <span class="md-ellipsis">
      
        Registered Prefixes (for comparison)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-ai-data-usage" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM-AI Data Usage
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#observations" class="md-nav__link">
    <span class="md-ellipsis">
      
        Observations
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#action-items" class="md-nav__link">
    <span class="md-ellipsis">
      
        Action Items
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemical-identifier-mapping-resources" class="md-nav__link">
    <span class="md-ellipsis">
      
        Chemical Identifier Mapping Resources
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Chemical Identifier Mapping Resources">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#unichem-recommended" class="md-nav__link">
    <span class="md-ellipsis">
      
        UniChem (Recommended)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bridgedb" class="md-nav__link">
    <span class="md-ellipsis">
      
        BridgeDb
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#direct-cross-references" class="md-nav__link">
    <span class="md-ellipsis">
      
        Direct Cross-References
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#chemicals-without-chebi-ids-gap-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      
        Chemicals Without ChEBI IDs (Gap Analysis)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Chemicals Without ChEBI IDs (Gap Analysis)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#lanthanide-chloride-hexahydrates" class="md-nav__link">
    <span class="md-ellipsis">
      
        Lanthanide Chloride Hexahydrates
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#options-for-kgx-representation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Options for KGX Representation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#submitting-to-chebi" class="md-nav__link">
    <span class="md-ellipsis">
      
        Submitting to ChEBI
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      
        References
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="curie-and-prefix-strategy">CURIE and Prefix Strategy</h1>
<p><strong>Date:</strong> 2026-01-08
<strong>Status:</strong> Planning document - needs community input
<strong>Related Issues:</strong> #53, #98</p>
<hr />
<h2 id="the-problem">The Problem</h2>
<p>We need to create CURIEs for many entity types. Some come from databases with their own IDs, some are entities we identify from literature. Creating and registering many prefixes is burdensome, but too few prefixes creates ambiguity.</p>
<p><strong>Constraints:</strong>
- Marcin (reasonably) doesn't want to register dozens of prefixes
- We want numerical local IDs, not semantic mnemonics
- CURIEs should ideally resolve to something useful
- Different databases have different entity types (media, strains, solutions, ingredients)</p>
<hr />
<h2 id="entity-types-inventory">Entity Types Inventory</h2>
<h3 id="from-external-databases-have-their-own-ids">From External Databases (Have Their Own IDs)</h3>
<table>
<thead>
<tr>
<th>Source</th>
<th>Entity Types</th>
<th>Their ID Format</th>
<th>Current Approach</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>BacDive</strong></td>
<td>strains</td>
<td>Numeric (161512)</td>
<td><code>bacdive:161512</code> works - registered!</td>
</tr>
<tr>
<td><strong>MediaDive</strong></td>
<td>media, solutions, ingredients, strains</td>
<td>Numeric + letters (104c, J562)</td>
<td>Unregistered <code>mediadive.medium:</code> etc.</td>
</tr>
<tr>
<td><strong>TogoMedium</strong></td>
<td>media, components</td>
<td>M443, SY85</td>
<td><code>togomedium:</code> ?</td>
</tr>
<tr>
<td><strong>NCBI</strong></td>
<td>taxa, genomes, biosamples</td>
<td>Numeric</td>
<td><code>NCBITaxon:</code>, <code>insdc.sra:</code> - registered</td>
</tr>
<tr>
<td><strong>ChEBI</strong></td>
<td>chemicals</td>
<td>Numeric</td>
<td><code>CHEBI:</code> - registered</td>
</tr>
<tr>
<td><strong>PubChem</strong></td>
<td>compounds</td>
<td>Numeric (CID)</td>
<td><code>pubchem.compound:</code> - registered</td>
</tr>
</tbody>
</table>
<h3 id="entities-we-may-need-to-mint-ids-for">Entities We May Need to Mint IDs For</h3>
<table>
<thead>
<tr>
<th>Entity Type</th>
<th>Source</th>
<th>Example</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td>Growth conditions</td>
<td>Literature</td>
<td>"37°C, pH 7, aerobic"</td>
<td>Combinations not in any database</td>
</tr>
<tr>
<td>Experimental observations</td>
<td>Our curation</td>
<td>"Strain X grows on medium Y"</td>
<td>The assertion itself</td>
</tr>
<tr>
<td>Literature-described media</td>
<td>Papers</td>
<td>"Modified MP medium from Smith 2019"</td>
<td>Not in MediaDive</td>
</tr>
<tr>
<td>Unmapped strains</td>
<td>Sheets</td>
<td>Lab designations like "AM1"</td>
<td>Need canonical ID</td>
</tr>
<tr>
<td>CMM-specific phenotypes</td>
<td>Our work</td>
<td>"High REE accumulation"</td>
<td>Domain-specific</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="prefix-strategy-options">Prefix Strategy Options</h2>
<h3 id="option-1-many-type-specific-prefixes">Option 1: Many Type-Specific Prefixes</h3>
<div class="highlight"><pre><span></span><code>mediadive.medium:104c
mediadive.solution:S123
mediadive.ingredient:I456
mediadive.strain:7890
</code></pre></div>
<p><strong>Pros:</strong>
- Type is unambiguous from CURIE alone
- Can have different URL patterns per type</p>
<p><strong>Cons:</strong>
- Many prefixes to register (Marcin's concern)
- Registration overhead
- Maintenance burden</p>
<h3 id="option-2-single-database-prefix-typed-local-ids">Option 2: Single Database Prefix + Typed Local IDs</h3>
<div class="highlight"><pre><span></span><code>mediadive:medium_104c
mediadive:solution_S123
mediadive:ingredient_I456
mediadive:strain_7890
</code></pre></div>
<p><strong>Pros:</strong>
- Only one prefix to register per database
- Type encoded in local ID
- URL pattern: <code>https://mediadive.dsmz.de/{local_id}</code> could route internally</p>
<p><strong>Cons:</strong>
- Less "pure" CURIE semantics
- Local ID is no longer just the database's ID
- Relies on database having unified resolver</p>
<h3 id="option-3-single-database-prefix-numeric-ids-let-database-resolve">Option 3: Single Database Prefix + Numeric IDs (Let Database Resolve)</h3>
<div class="highlight"><pre><span></span><code>mediadive:104c     → https://mediadive.dsmz.de/104c (let them figure out type)
mediadive:S123     → https://mediadive.dsmz.de/S123
bacdive:161512     → https://bacdive.dsmz.de/strain/161512 (they know it&#39;s a strain)
</code></pre></div>
<p><strong>Pros:</strong>
- Simplest - one prefix per database
- Database handles type resolution
- Matches how <code>bacdive:</code> already works</p>
<p><strong>Cons:</strong>
- Requires database to have unified resolver
- Type not visible in CURIE
- May not work for all databases</p>
<h3 id="option-4-project-prefix-for-minted-ids-database-prefixes-for-external">Option 4: Project Prefix for Minted IDs + Database Prefixes for External</h3>
<div class="highlight"><pre><span></span><code># External database entities - use their prefix
bacdive:161512
mediadive:104c
NCBITaxon:408

# Entities we mint - use project prefix with numeric IDs
cmm:1         → growth condition
cmm:2         → experimental observation
cmm:3         → literature-described medium
</code></pre></div>
<p><strong>Pros:</strong>
- Clear separation: external vs minted
- Only need to register ONE new prefix (cmm:)
- Numeric IDs are stable, not semantic</p>
<p><strong>Cons:</strong>
- <code>cmm:1</code> doesn't tell you it's a growth condition
- Need metadata/registry to know what each ID refers to
- Resolver needs to handle all types</p>
<h3 id="option-5-hybrid-project-prefix-with-type-codes">Option 5: Hybrid - Project Prefix with Type Codes</h3>
<div class="highlight"><pre><span></span><code># External - use registered database prefixes
bacdive:161512
CHEBI:32599

# Minted - project prefix + type code + number
cmm:GC0001    → Growth Condition #1
cmm:OB0001    → Observation #1
cmm:LM0001    → Literature Medium #1
cmm:ST0001    → Strain (unmapped) #1
</code></pre></div>
<p><strong>Pros:</strong>
- Single prefix to register
- Type visible in local ID
- Numeric portion is stable</p>
<p><strong>Cons:</strong>
- Type codes are semantic (but short)
- Need to define and document codes
- Still need registry of what each ID means</p>
<hr />
<h2 id="questions-for-community">Questions for Community</h2>
<h3 id="for-bioregistry">For Bioregistry</h3>
<ol>
<li>"We're building a microbial knowledge graph. MediaDive has media, solutions, ingredients, and strains - all with different ID formats. Should we register:</li>
<li>One <code>mediadive:</code> prefix that resolves all entity types?</li>
<li>Separate <code>mediadive.medium:</code>, <code>mediadive.solution:</code>, etc.?</li>
<li>
<p>Something else?"</p>
</li>
<li>
<p>"For entities we mint ourselves (growth conditions, observations), what's the recommended pattern? A single project prefix with typed local IDs?"</p>
</li>
<li>
<p>"What's the minimum viable registration? Can we start with a GitHub-hosted prefix map and formalize later?"</p>
</li>
</ol>
<h3 id="for-biolinkkgx">For Biolink/KGX</h3>
<ol>
<li>
<p>"When we have assertions like 'strain X grows on medium Y at 37°C', should the growth condition be a node? An edge property? A reified statement?"</p>
</li>
<li>
<p>"For literature-described entities not in any database, what category should they have?"</p>
</li>
</ol>
<h3 id="for-w3idorg">For w3id.org</h3>
<ol>
<li>"Can we register a <code>w3id.org/cmm/</code> namespace that routes to our GitHub Pages, so our minted IDs are resolvable even before formal Bioregistry registration?"</li>
</ol>
<hr />
<h2 id="two-different-problems-terms-vs-instances">Two Different Problems: Terms vs Instances</h2>
<p><strong>Critical distinction:</strong></p>
<table>
<thead>
<tr>
<th>Type</th>
<th>What It Is</th>
<th>Namespace</th>
<th>Examples</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Ontology terms</strong> (T-box)</td>
<td>Classes, predicates, categories</td>
<td>METPO, Biolink, OBO ontologies</td>
<td><code>METPO:1004005</code> (the class "Growth Medium")</td>
</tr>
<tr>
<td><strong>Instance data</strong> (A-box)</td>
<td>Specific strains, media, experiments</td>
<td>Database prefixes, project prefix</td>
<td><code>bacdive:161512</code> (a specific strain)</td>
</tr>
</tbody>
</table>
<p><strong>METPO is for terms, NOT instances.</strong> You can't use <code>METPO:</code> to identify a specific strain from BacDive or a specific medium from MediaDive.</p>
<hr />
<h2 id="metpo-for-domain-specific-terms">METPO for Domain-Specific Terms</h2>
<p>METPO provides a registered namespace for microbial phenotype ontology terms:</p>
<table>
<thead>
<tr>
<th>Aspect</th>
<th>METPO Status</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td>Prefix registered?</td>
<td>✅ Yes (<code>METPO:</code>)</td>
<td>In OBO/Bioregistry</td>
</tr>
<tr>
<td>URL resolvable?</td>
<td>✅ Yes</td>
<td>Via OBO PURLs</td>
</tr>
<tr>
<td>Formal local ID registry?</td>
<td>❌ No</td>
<td>But has BioPortal dumps, code to analyze</td>
</tr>
<tr>
<td>Governance</td>
<td>Informal</td>
<td>BBOP team, responsive to requests</td>
</tr>
</tbody>
</table>
<p><strong>What METPO gives us:</strong>
- Registered prefix for ontology terms
- Can create new classes/predicates relatively quickly
- Terms become part of a shared vocabulary
- Resolvable URIs via OBO infrastructure</p>
<p><strong>What we do for METPO local IDs:</strong>
- Code analyzes old sheets and BioPortal dumps (location: TBD - @turbomam has this)
- Track what IDs exist, avoid collisions
- Informal but functional
- Can be agile about creating new terms for small categories</p>
<h3 id="using-metpo-for-cmm-categories-and-predicates">Using METPO for CMM Categories and Predicates</h3>
<p>For domain-specific <strong>categories</strong> (node types) and <strong>predicates</strong> (edge types) that don't fit Biolink:</p>
<table>
<thead>
<tr>
<th>CMM Concept</th>
<th>Type</th>
<th>Potential METPO Term</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td>Growth medium (category)</td>
<td>Class</td>
<td><code>METPO:1004005</code></td>
<td>Already exists</td>
</tr>
<tr>
<td>"grows in" (predicate)</td>
<td>Property</td>
<td><code>METPO:2000517</code></td>
<td>Already exists</td>
</tr>
<tr>
<td>Growth condition (category)</td>
<td>Class</td>
<td><code>METPO:XXXXXXX</code></td>
<td>Could propose</td>
</tr>
<tr>
<td>"optimal pH" (predicate)</td>
<td>Property</td>
<td><code>METPO:XXXXXXX</code></td>
<td>Could propose</td>
</tr>
</tbody>
</table>
<p><strong>METPO does NOT solve:</strong> Identifying specific instances (strains, media, experiments) from external databases.</p>
<hr />
<h2 id="the-instance-id-problem-still-unsolved">The Instance ID Problem (Still Unsolved)</h2>
<p>For actual instance data, we still need:</p>
<table>
<thead>
<tr>
<th>Instance Type</th>
<th>Source</th>
<th>Prefix Needed</th>
</tr>
</thead>
<tbody>
<tr>
<td>BacDive strains</td>
<td>BacDive database</td>
<td><code>bacdive:</code> ✅ registered</td>
</tr>
<tr>
<td>MediaDive media</td>
<td>MediaDive database</td>
<td><code>mediadive:</code> ❌ not registered</td>
</tr>
<tr>
<td>TogoMedium media</td>
<td>TogoMedium database</td>
<td><code>togomedium:</code> ❌ not registered</td>
</tr>
<tr>
<td>Literature-described media</td>
<td>Our curation</td>
<td><code>cmm:</code> ❌ not registered</td>
</tr>
<tr>
<td>Growth conditions</td>
<td>Our curation</td>
<td><code>cmm:</code> ❌ not registered</td>
</tr>
<tr>
<td>Experimental observations</td>
<td>Our curation</td>
<td><code>cmm:</code> ❌ not registered</td>
</tr>
</tbody>
</table>
<p><strong>This is the real prefix registration burden</strong> - not ontology terms (we have METPO), but instance identifiers from databases and our own minted entities.</p>
<hr />
<h2 id="proposed-strategy-draft">Proposed Strategy (Draft)</h2>
<h3 id="phase-1-use-whats-registered">Phase 1: Use What's Registered</h3>
<table>
<thead>
<tr>
<th>Entity Type</th>
<th>Prefix</th>
<th>Status</th>
</tr>
</thead>
<tbody>
<tr>
<td>BacDive strains</td>
<td><code>bacdive:</code></td>
<td>✅ Registered</td>
</tr>
<tr>
<td>NCBI taxa</td>
<td><code>NCBITaxon:</code></td>
<td>✅ Registered</td>
</tr>
<tr>
<td>ChEBI chemicals</td>
<td><code>CHEBI:</code></td>
<td>✅ Registered</td>
</tr>
<tr>
<td>PubChem compounds</td>
<td><code>pubchem.compound:</code></td>
<td>✅ Registered</td>
</tr>
</tbody>
</table>
<h3 id="phase-2-register-database-prefixes">Phase 2: Register Database Prefixes</h3>
<table>
<thead>
<tr>
<th>Database</th>
<th>Proposed Prefix</th>
<th>URL Pattern</th>
<th>Action Needed</th>
</tr>
</thead>
<tbody>
<tr>
<td>MediaDive</td>
<td><code>mediadive:</code></td>
<td><code>https://mediadive.dsmz.de/{id}</code></td>
<td>Check if they have unified resolver</td>
</tr>
<tr>
<td>TogoMedium</td>
<td><code>togomedium:</code></td>
<td><code>http://togomedium.org/{id}</code></td>
<td>Research their ID scheme</td>
</tr>
</tbody>
</table>
<h3 id="phase-3-register-project-prefix-for-minted-ids">Phase 3: Register Project Prefix for Minted IDs</h3>
<table>
<thead>
<tr>
<th>Prefix</th>
<th>Purpose</th>
<th>URL Pattern</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>cmm:</code></td>
<td>Entities we mint</td>
<td><code>https://w3id.org/cmm/{id}</code> → GitHub Pages registry</td>
</tr>
</tbody>
</table>
<h3 id="local-id-format-for-cmm">Local ID Format for <code>cmm:</code></h3>
<p><strong>Option A: Pure Numeric</strong>
<div class="highlight"><pre><span></span><code>cmm:1, cmm:2, cmm:3, ...
</code></pre></div>
Requires separate registry to know what each is.</p>
<p><strong>Option B: Type-Prefixed Numeric</strong>
<div class="highlight"><pre><span></span><code>cmm:GC0001  (Growth Condition)
cmm:OB0001  (Observation)
cmm:LM0001  (Literature Medium)
</code></pre></div>
Type visible, still mostly numeric.</p>
<p><strong>Recommendation:</strong> Option B - minimal semantic content, but enough to route/categorize.</p>
<hr />
<h2 id="registry-requirements">Registry Requirements</h2>
<p>If we mint IDs, we need a registry that tracks:</p>
<div class="highlight"><pre><span></span><code><span class="c1"># Example: cmm_id_registry.yaml</span>
<span class="nt">cmm:GC0001</span><span class="p">:</span>
<span class="w">  </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">GrowthCondition</span>
<span class="w">  </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;Standard</span><span class="nv"> </span><span class="s">methylotroph</span><span class="nv"> </span><span class="s">conditions&quot;</span>
<span class="w">  </span><span class="nt">temperature_celsius</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">30</span>
<span class="w">  </span><span class="nt">ph</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">7.0</span>
<span class="w">  </span><span class="nt">oxygen</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">aerobic</span>
<span class="w">  </span><span class="nt">created</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2026-01-08</span>
<span class="w">  </span><span class="nt">created_by</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">turbomam</span>

<span class="nt">cmm:LM0001</span><span class="p">:</span>
<span class="w">  </span><span class="nt">type</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">LiteratureMedium</span>
<span class="w">  </span><span class="nt">label</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;Modified</span><span class="nv"> </span><span class="s">MP</span><span class="nv"> </span><span class="s">medium</span><span class="nv"> </span><span class="s">(Smith</span><span class="nv"> </span><span class="s">2019)&quot;</span>
<span class="w">  </span><span class="nt">source</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">PMID:12345678</span>
<span class="w">  </span><span class="nt">based_on</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">mediadive:104c</span>
<span class="w">  </span><span class="nt">modifications</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;Added</span><span class="nv"> </span><span class="s">0.1%</span><span class="nv"> </span><span class="s">yeast</span><span class="nv"> </span><span class="s">extract&quot;</span>
<span class="w">  </span><span class="nt">created</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2026-01-08</span>
</code></pre></div>
<p>This could be:
- YAML/JSON in the repo (version controlled)
- MongoDB collection (queryable)
- Both (YAML as source of truth, loaded into MongoDB)</p>
<hr />
<h2 id="research-findings">Research Findings</h2>
<h3 id="mediadive-url-structure-verified-2026-01-08">MediaDive URL Structure (Verified 2026-01-08)</h3>
<p>MediaDive does NOT have a unified resolver. URLs are type-specific:</p>
<table>
<thead>
<tr>
<th>Entity Type</th>
<th>URL Pattern</th>
<th>Example</th>
</tr>
</thead>
<tbody>
<tr>
<td>Media</td>
<td><code>https://mediadive.dsmz.de/medium/{id}</code></td>
<td><code>/medium/104c</code></td>
</tr>
<tr>
<td>Solutions</td>
<td><code>https://mediadive.dsmz.de/solutions/{id}</code></td>
<td><code>/solutions/179</code></td>
</tr>
<tr>
<td>Ingredients</td>
<td><code>https://mediadive.dsmz.de/ingredients/{id}</code></td>
<td><code>/ingredients/124</code></td>
</tr>
<tr>
<td>Strains</td>
<td><code>https://mediadive.dsmz.de/strains/medium/{id}</code></td>
<td><code>/strains/medium/104c</code></td>
</tr>
</tbody>
</table>
<p><strong>Implication:</strong> A single <code>mediadive:104c</code> prefix cannot resolve because the URL needs the entity type.</p>
<p><strong>Options:</strong>
1. Register type-specific prefixes: <code>mediadive.medium:104c</code> → <code>https://mediadive.dsmz.de/medium/104c</code>
2. Include type in local ID: <code>mediadive:medium/104c</code> → <code>https://mediadive.dsmz.de/medium/104c</code>
3. Ask MediaDive to add a unified resolver that redirects based on ID format</p>
<p>Option 2 is interesting because then we only need ONE prefix but include the type in the local ID. The CURIE <code>mediadive:medium/104c</code> expands to the correct URL.</p>
<p><strong>Bioregistry status:</strong> <code>mediadive</code> is NOT registered (404 on lookup).</p>
<h3 id="bacdive-url-structure-comparison">BacDive URL Structure (Comparison)</h3>
<p>BacDive DOES have a unified resolver:
- <code>bacdive:161512</code> → <code>https://bacdive.dsmz.de/strain/161512</code></p>
<p>They automatically add <code>/strain/</code> because all BacDive IDs are strains. This is why <code>bacdive:</code> works with a single prefix.</p>
<h3 id="togomedium-url-structure-from-docstogomediummd">TogoMedium URL Structure (from docs/togomedium.md)</h3>
<p>TogoMedium uses a consistent ID scheme for media:
- Media IDs: M443, M2476, M1871 (prefixed with M)
- Component IDs: Different scheme
- Organism IDs: Different scheme</p>
<p><strong>API access:</strong>
<div class="highlight"><pre><span></span><code>https://togomedium.org/sparqlist/api/gmdb_medium_by_gmid?gm_id=M443
</code></pre></div></p>
<p><strong>Bioregistry status:</strong> <code>togomedium</code> is NOT registered (404 on lookup).</p>
<p><strong>Implication:</strong> Could potentially use a single <code>togomedium:</code> prefix since media IDs start with "M", components with different prefix, etc. But need to verify if their ID scheme is truly unique per entity type.</p>
<hr />
<h2 id="identifier-registry-landscape">Identifier Registry Landscape</h2>
<h3 id="registries-to-consider">Registries to Consider</h3>
<table>
<thead>
<tr>
<th>Registry</th>
<th>Purpose</th>
<th>Relationship</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://bioregistry.io">Bioregistry</a></td>
<td>Meta-registry aggregating from many sources</td>
<td><strong>Primary</strong> - imports from OBO, prefix.cc, Identifiers.org</td>
</tr>
<tr>
<td><a href="https://obofoundry.org">OBO Foundry</a></td>
<td>Ontology registry (T-box terms)</td>
<td>Bioregistry imports OBO prefixes</td>
</tr>
<tr>
<td><a href="https://identifiers.org">Identifiers.org</a></td>
<td>Life science identifier resolution (MIRIAM)</td>
<td>Bioregistry imports from it</td>
</tr>
<tr>
<td><a href="https://prefix.cc">prefix.cc</a></td>
<td>Community prefix registry</td>
<td>Bioregistry imports from it</td>
</tr>
<tr>
<td><a href="https://w3id.org">w3id.org</a></td>
<td><strong>Persistent URI redirect service</strong></td>
<td>Different - provides resolution, not cataloging</td>
</tr>
<tr>
<td><a href="https://n2t.net">N2T.net</a></td>
<td>ARK/EZID identifier resolution</td>
<td>Separate infrastructure</td>
</tr>
</tbody>
</table>
<p><strong>Key insight:</strong> Bioregistry is a meta-registry. If a prefix is in OBO Foundry, prefix.cc, or Identifiers.org, it's likely already in Bioregistry.</p>
<p><strong>w3id.org is different:</strong> It's a redirect service (register <code>w3id.org/cmm/</code> → your GitHub Pages), not a semantic registry.</p>
<h3 id="project-prefix-status-verified-2026-01-08">Project Prefix Status (Verified 2026-01-08)</h3>
<table>
<thead>
<tr>
<th>Prefix</th>
<th>Bioregistry</th>
<th>Identifiers.org</th>
<th>OBO Foundry</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>cmm</code></td>
<td>❌ 404</td>
<td>❌ 404</td>
<td>❌ 404</td>
<td>Available</td>
</tr>
<tr>
<td><code>cmm-ai</code></td>
<td>❌ Not checked</td>
<td>-</td>
<td>-</td>
<td>Likely available</td>
</tr>
<tr>
<td><code>ber-cmm-ai</code></td>
<td>❌ Not checked</td>
<td>-</td>
<td>-</td>
<td>Clearly scoped to project</td>
</tr>
</tbody>
</table>
<p><strong>Note:</strong> <code>cmo</code> (Clinical Measurement Ontology) IS registered - similar but different.</p>
<h3 id="prefix-naming-trade-offs">Prefix Naming Trade-offs</h3>
<p><strong>Short prefix (<code>cmm:</code>):</strong>
- Pro: Concise CURIEs, easy to type
- Con: "CMM" is generic (Capability Maturity Model, etc.), harder to claim ownership</p>
<p><strong>Long prefix (<code>ber-cmm-ai:</code> or <code>cmm-ai:</code>):</strong>
- Pro: Clearly scoped, unlikely to conflict, matches repo names
- Con: Verbose CURIEs</p>
<p><strong>Recommendation:</strong> Lean toward <code>cmm-ai:</code> - matches repository naming, distinguishes from generic CMM uses.</p>
<hr />
<h2 id="custom-prefixes-in-use-audit-2026-01-08">Custom Prefixes in Use (Audit 2026-01-08)</h2>
<h3 id="kg-microbe-constants-constantspy">kg-microbe Constants (<code>constants.py</code>)</h3>
<table>
<thead>
<tr>
<th>Prefix</th>
<th>Bioregistry</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>kgmicrobe.strain:</code></td>
<td>❌ Not registered</td>
<td>For strains not in BacDive</td>
</tr>
<tr>
<td><code>mediadive.medium:</code></td>
<td>❌ Not registered</td>
<td>MediaDive media</td>
</tr>
<tr>
<td><code>mediadive.solution:</code></td>
<td>❌ Not registered</td>
<td>MediaDive solutions</td>
</tr>
<tr>
<td><code>mediadive.ingredient:</code></td>
<td>❌ Not registered</td>
<td>MediaDive ingredients</td>
</tr>
<tr>
<td><code>mediadive.medium-type:</code></td>
<td>❌ Not registered</td>
<td>Complex vs defined</td>
</tr>
<tr>
<td><code>bacdive.isolation_source:</code></td>
<td>❌ Not registered</td>
<td>Isolation source categories</td>
</tr>
<tr>
<td><code>cell_shape:</code></td>
<td>❌ Not registered</td>
<td>Cell morphology</td>
</tr>
<tr>
<td><code>pathways:</code></td>
<td>❌ Not registered</td>
<td>Metabolic pathways</td>
</tr>
<tr>
<td><code>carbon_substrates:</code></td>
<td>❌ Not registered</td>
<td>Carbon source utilization</td>
</tr>
<tr>
<td><code>assay:</code></td>
<td>❌ Not registered</td>
<td>Assay types</td>
</tr>
<tr>
<td><code>debio:</code></td>
<td>❌ Not registered</td>
<td>Debio ontology</td>
</tr>
<tr>
<td><code>BSL:</code></td>
<td>❌ Not registered</td>
<td>Biosafety levels</td>
</tr>
<tr>
<td><code>taxonomy_id:</code></td>
<td>❌ Not registered</td>
<td>Taxonomy identifiers</td>
</tr>
</tbody>
</table>
<h3 id="registered-prefixes-for-comparison">Registered Prefixes (for comparison)</h3>
<table>
<thead>
<tr>
<th>Prefix</th>
<th>Bioregistry</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>bacdive:</code></td>
<td>✅ Registered</td>
<td>BacDive strains</td>
</tr>
<tr>
<td><code>NCBITaxon:</code></td>
<td>✅ Registered</td>
<td>NCBI taxonomy</td>
</tr>
<tr>
<td><code>CHEBI:</code></td>
<td>✅ Registered</td>
<td>ChEBI chemicals</td>
</tr>
<tr>
<td><code>GO:</code></td>
<td>✅ Registered</td>
<td>Gene Ontology</td>
</tr>
<tr>
<td><code>KEGG:</code></td>
<td>✅ Registered</td>
<td>KEGG</td>
</tr>
<tr>
<td><code>RHEA:</code></td>
<td>✅ Registered</td>
<td>Rhea reactions</td>
</tr>
<tr>
<td><code>METPO:</code></td>
<td>✅ Registered</td>
<td>Microbial phenotype ontology</td>
</tr>
<tr>
<td><code>EC:</code></td>
<td>✅ Registered</td>
<td>Enzyme Commission</td>
</tr>
<tr>
<td><code>PubChem:</code></td>
<td>✅ Registered</td>
<td>PubChem compounds</td>
</tr>
<tr>
<td><code>CAS-RN:</code></td>
<td>✅ Registered</td>
<td>CAS Registry Numbers</td>
</tr>
</tbody>
</table>
<h3 id="cmm-ai-data-usage">CMM-AI Data Usage</h3>
<p>Found in <code>CMM-AI/data/</code>:
- <code>kgmicrobe.medium:</code> (e.g., <code>kgmicrobe.medium:ATCC_1306</code>)
- <code>mediadive.medium:</code> (e.g., <code>mediadive.medium:632</code>)</p>
<h3 id="observations">Observations</h3>
<ol>
<li><strong>13+ unregistered prefixes</strong> in active use in kg-microbe</li>
<li>Most follow <code>database.entity_type:</code> convention</li>
<li>None resolve to URLs via Bioregistry</li>
<li>Some are very domain-specific (<code>cell_shape:</code>, <code>carbon_substrates:</code>)</li>
<li>The <code>kgmicrobe.*</code> prefixes are project-specific but not registered</li>
</ol>
<hr />
<h2 id="action-items">Action Items</h2>
<ul>
<li>[x] Research MediaDive's resolver - <strong>No unified resolver, URLs are type-specific</strong></li>
<li>[x] Research TogoMedium's ID scheme - <strong>IDs prefixed by type (M for media), may enable single prefix</strong></li>
<li>[x] Check if <code>cmm</code> is registered - <strong>Not registered as of 2026-01-08</strong></li>
<li>[x] Audit custom prefixes in kg-microbe, CMM-AI, cmm-ai-automation</li>
<li>[ ] Decide on project prefix name (<code>cmm-ai:</code> vs <code>ber-cmm-ai:</code> vs other)</li>
<li>[ ] Draft Bioregistry issue asking about prefix strategy for type-specific resolvers</li>
<li>[ ] Decide on local ID format for minted entities</li>
<li>[ ] Design registry schema for minted IDs</li>
<li>[ ] Register w3id.org/cmm-ai/ namespace (low barrier)</li>
<li>[ ] Consider asking MediaDive to add unified resolver</li>
<li>[ ] Verify TogoMedium ID uniqueness across entity types</li>
<li>[ ] Prioritize which unregistered prefixes to formalize</li>
</ul>
<hr />
<h2 id="chemical-identifier-mapping-resources">Chemical Identifier Mapping Resources</h2>
<p>When chemicals lack ChEBI IDs but have PubChem IDs (or vice versa), these systems can help find cross-references:</p>
<h3 id="unichem-recommended">UniChem (Recommended)</h3>
<p>EBI's unified cross-referencing system for chemical structures.</p>
<table>
<thead>
<tr>
<th>Aspect</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td>URL</td>
<td><a href="https://www.ebi.ac.uk/unichem/">https://www.ebi.ac.uk/unichem/</a></td>
</tr>
<tr>
<td>Coverage</td>
<td>ChEBI, PubChem, ChEMBL, DrugBank, KEGG, 40+ sources</td>
</tr>
<tr>
<td>Method</td>
<td>InChI key-based structure matching</td>
</tr>
<tr>
<td>API</td>
<td><code>https://www.ebi.ac.uk/unichem/rest/src_compound_id/{id}/{src_id}</code></td>
</tr>
<tr>
<td>Docs</td>
<td><a href="https://chembl.gitbook.io/unichem">UniChem 2.0 Documentation</a></td>
</tr>
</tbody>
</table>
<p><strong>Source IDs for API:</strong>
- ChEBI: <code>src_id=7</code>
- PubChem: <code>src_id=22</code>
- ChEMBL: <code>src_id=1</code></p>
<h3 id="bridgedb">BridgeDb</h3>
<p>Framework for gene, protein, and metabolite ID mapping.</p>
<table>
<thead>
<tr>
<th>Aspect</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td>URL</td>
<td><a href="https://www.bridgedb.org/mapping-databases/metabolite-mappings.html">https://www.bridgedb.org/</a></td>
</tr>
<tr>
<td>Coverage</td>
<td>PubChem ↔ ChEBI ↔ HMDB ↔ KEGG ↔ Wikidata</td>
</tr>
<tr>
<td>Format</td>
<td>Pre-built Derby files for local use</td>
</tr>
<tr>
<td>Download</td>
<td><a href="https://figshare.com/projects/BridgeDb_metabolites/28500">Figshare metabolite mappings</a></td>
</tr>
</tbody>
</table>
<h3 id="direct-cross-references">Direct Cross-References</h3>
<ul>
<li><strong>ChEBI entries</strong> include PubChem links in "Database Links" section</li>
<li><strong>PubChem entries</strong> show ChEBI in "Depositor-Supplied Identifiers"</li>
<li><strong>Wikidata</strong> often has both PubChem and ChEBI properties</li>
</ul>
<hr />
<h2 id="chemicals-without-chebi-ids-gap-analysis">Chemicals Without ChEBI IDs (Gap Analysis)</h2>
<p><strong>Date:</strong> 2026-01-21</p>
<p>Some chemicals in our media formulations lack ChEBI IDs. This section documents known gaps.</p>
<h3 id="lanthanide-chloride-hexahydrates">Lanthanide Chloride Hexahydrates</h3>
<p>These rare earth salts are used in methylotroph growth media (MP medium) but have no ChEBI entries:</p>
<table>
<thead>
<tr>
<th>Compound</th>
<th>ChEBI</th>
<th>PubChem CID</th>
<th>CAS</th>
</tr>
</thead>
<tbody>
<tr>
<td>Dysprosium(III) chloride hexahydrate</td>
<td>❌ None</td>
<td><a href="https://pubchem.ncbi.nlm.nih.gov/compound/71311204">71311204</a></td>
<td>15059-52-6</td>
</tr>
<tr>
<td>Neodymium(III) chloride hexahydrate</td>
<td>❌ None</td>
<td><a href="https://pubchem.ncbi.nlm.nih.gov/compound/18355679">18355679</a></td>
<td>13477-89-9</td>
</tr>
<tr>
<td>Praseodymium(III) chloride hexahydrate</td>
<td>❌ None</td>
<td><a href="https://pubchem.ncbi.nlm.nih.gov/compound/91886610">91886610</a></td>
<td>17272-46-7</td>
</tr>
</tbody>
</table>
<p><strong>ChEBI has entries for the elements only:</strong>
- Dysprosium atom: <a href="https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:33377">CHEBI:33377</a>
- Praseodymium atom: <a href="https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:49828">CHEBI:49828</a>
- Praseodymium acetate: <a href="https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:63078">CHEBI:63078</a> (only Pr salt)</p>
<p><strong>For comparison</strong>, lanthanum trichloride (anhydrous) has <a href="https://www.ebi.ac.uk/chebi/searchId.do?chebiId=231515">CHEBI:231515</a>.</p>
<h3 id="options-for-kgx-representation">Options for KGX Representation</h3>
<ol>
<li><strong>Use PubChem IDs</strong>: <code>PUBCHEM.COMPOUND:71311204</code> (registered prefix)</li>
<li><strong>Submit to ChEBI</strong>: Request new entries for these compounds</li>
<li><strong>Use UUIDs</strong>: Mint local IDs with chemical names as labels</li>
</ol>
<p><strong>Recommendation:</strong> Use <code>PUBCHEM.COMPOUND:</code> prefix since it's registered in Bioregistry and PubChem has comprehensive coverage of these compounds.</p>
<h3 id="submitting-to-chebi">Submitting to ChEBI</h3>
<p>ChEBI accepts submissions for new compounds:
- <a href="https://www.ebi.ac.uk/chebi/submissions/login.do">ChEBI submission form</a>
- Requires: structure, name, synonyms, literature references
- Turnaround: typically 2-4 weeks</p>
<hr />
<h2 id="references">References</h2>
<ul>
<li><a href="https://bioregistry.io/">Bioregistry</a> - Prefix registration</li>
<li><a href="https://w3id.org/">w3id.org</a> - Persistent URI service</li>
<li><a href="https://github.com/biolink/information-resource-registry">Biolink information-resource-registry</a> - infores: registration</li>
<li><a href="https://www.ebi.ac.uk/unichem/">UniChem</a> - Chemical cross-referencing</li>
<li><a href="https://www.bridgedb.org/">BridgeDb</a> - Identifier mapping framework</li>
<li>Issue #53 - Standardize CURIE prefixes</li>
<li>Issue #98 - Canonical namespace policy</li>
</ul>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>