
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://turbomam.github.io/cmm-ai-automation/cmm_vs_kg_microbe/">
      
      
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>CMM-AI vs kg-microbe: Methods and Outcomes - cmm-ai-automation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
      


    
    

      
    
<script src="../assets/pymdownx-extras/material-extra-theme-2ek1P7jT.js" type="text/javascript"></script>
<script src="../assets/pymdownx-extras/material-extra-3rdparty-BUCF4rjN.js" type="text/javascript"></script>

    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../css/ansi-colours.css">
    
      <link rel="stylesheet" href="../css/jupyter-cells.css">
    
      <link rel="stylesheet" href="../css/pandas-dataframe.css">
    
      <link rel="stylesheet" href="../assets/pymdownx-extras/extra-c6ab11540a.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cmm-ai-vs-kg-microbe-methods-and-outcomes" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      


<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="header.title"
  >
    <a
      href="https://turbomam.github.io/cmm-ai-automation/"
      title="cmm-ai-automation"
      class="md-header__button md-logo"
      aria-label="cmm-ai-automation"
    >
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cmm-ai-automation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CMM-AI vs kg-microbe: Methods and Outcomes
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m15.07 11.25-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="cmm-ai-automation" class="md-nav__button md-logo" aria-label="cmm-ai-automation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m19 2-5 4.5v11l5-4.5zM6.5 5C4.55 5 2.45 5.4 1 6.5v14.66c0 .25.25.5.5.5.1 0 .15-.07.25-.07 1.35-.65 3.3-1.09 4.75-1.09 1.95 0 4.05.4 5.5 1.5 1.35-.85 3.8-1.5 5.5-1.5 1.65 0 3.35.31 4.75 1.06.1.05.15.03.25.03.25 0 .5-.25.5-.5V6.5c-.6-.45-1.25-.75-2-1V19c-1.1-.35-2.3-.5-3.5-.5-1.7 0-4.15.65-5.5 1.5V6.5C10.55 5.4 8.45 5 6.5 5"/></svg>

    </a>
    cmm-ai-automation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/turbomam/cmm-ai-automation" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../best-practices/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Best Practices
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../elements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Schema
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../architecture/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Architecture
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../togomedium/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    TogoMedium
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../about/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    About
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#executive-summary" class="md-nav__link">
    <span class="md-ellipsis">
      
        Executive Summary
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#1-prefix-strategy-and-curie-validity" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Prefix Strategy and CURIE Validity
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Prefix Strategy and CURIE Validity">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-approach" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Approach
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#2-schema-driven-vs-ad-hoc-generation" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Schema-Driven vs Ad-hoc Generation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="2. Schema-Driven vs Ad-hoc Generation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-approach_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Approach
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach_1" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#3-biolink-model-compliance" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. Biolink Model Compliance
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="3. Biolink Model Compliance">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-issues" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Issues
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#4-entity-resolution-and-provenance" class="md-nav__link">
    <span class="md-ellipsis">
      
        4. Entity Resolution and Provenance
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="4. Entity Resolution and Provenance">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-approach_2" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Approach
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach_3" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#5-media-grounding-quality" class="md-nav__link">
    <span class="md-ellipsis">
      
        5. Media Grounding Quality
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="5. Media Grounding Quality">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-cmm-ai-approach-from-related-project" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe / CMM-AI Approach (from related project)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach-planned" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach (Planned)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#6-testing-and-validation" class="md-nav__link">
    <span class="md-ellipsis">
      
        6. Testing and Validation
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="6. Testing and Validation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#7-kgx-field-usage-and-provenance" class="md-nav__link">
    <span class="md-ellipsis">
      
        7. KGX Field Usage and Provenance
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="7. KGX Field Usage and Provenance">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-field-issues" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Field Issues
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach-strict-kgx-compliance" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach: Strict KGX Compliance
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="CMM Approach: Strict KGX Compliance">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cmm-rules" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Rules
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#8-custom-vs-standard-kgx-fields" class="md-nav__link">
    <span class="md-ellipsis">
      
        8. Custom vs Standard KGX Fields
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="8. Custom vs Standard KGX Fields">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#kg-microbe-custom-fields" class="md-nav__link">
    <span class="md-ellipsis">
      
        kg-microbe Custom Fields
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-approach-lenient-but-documented" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Approach: Lenient But Documented
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#9-documented-gray-zones-incompatibilities" class="md-nav__link">
    <span class="md-ellipsis">
      
        9. Documented Gray Zones / Incompatibilities
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="9. Documented Gray Zones / Incompatibilities">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#known-kg-microbe-issues-referenced-in-cmm-docs" class="md-nav__link">
    <span class="md-ellipsis">
      
        Known kg-microbe Issues Referenced in CMM Docs
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-intentional-divergences" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Intentional Divergences
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#10-areas-of-potential-coordination" class="md-nav__link">
    <span class="md-ellipsis">
      
        10. Areas of Potential Coordination
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#11-when-to-use-each-approach" class="md-nav__link">
    <span class="md-ellipsis">
      
        11. When to Use Each Approach
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="11. When to Use Each Approach">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#use-kg-microbe-when" class="md-nav__link">
    <span class="md-ellipsis">
      
        Use kg-microbe When
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#use-cmm-when" class="md-nav__link">
    <span class="md-ellipsis">
      
        Use CMM When
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#12-references" class="md-nav__link">
    <span class="md-ellipsis">
      
        12. References
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="12. References">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cmm-documentation" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Documentation
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cmm-issues" class="md-nav__link">
    <span class="md-ellipsis">
      
        CMM Issues
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
  
                  


  
  


<h1 id="cmm-ai-vs-kg-microbe-methods-and-outcomes">CMM-AI vs kg-microbe: Methods and Outcomes</h1>
<p><strong>Purpose:</strong> Document key differences in approach and resulting data quality between this repository and kg-microbe.</p>
<p><strong>Date:</strong> 2026-01-07<br />
<strong>Related Issue:</strong> #123</p>
<hr />
<h2 id="executive-summary">Executive Summary</h2>
<p>While both projects create knowledge graphs for microbial data, cmm-ai-automation prioritizes strict KGX/Biolink compliance, explicit provenance tracking, and schema-driven validation. This leads to higher data quality but requires more upfront modeling work.</p>
<hr />
<h2 id="1-prefix-strategy-and-curie-validity">1. Prefix Strategy and CURIE Validity</h2>
<h3 id="kg-microbe-approach">kg-microbe Approach</h3>
<ul>
<li>Uses <strong>generic prefixes</strong> with source embedded in local ID:</li>
<li><code>strain:bacdive_161512</code></li>
<li><code>medium:104c</code> (ambiguous - MediaDive? TogoMedium?)</li>
<li><code>assay:API_zym_Cystine arylamidase</code> ⚠️ <strong>spaces in CURIE</strong></li>
</ul>
<p><strong>Issues</strong> (see kg-microbe#430):
- 100 invalid CURIEs (0.05% of nodes) with spaces or extra colons
- Examples: <code>assay:API_zym_Acid phosphatase</code>, <code>strain:NRRL-:-NRS-341</code>
- Breaks RDF export and CURIE expansion</p>
<h3 id="cmm-approach">CMM Approach</h3>
<ul>
<li>Uses <strong>scoped bioregistry-style prefixes</strong>:</li>
<li><code>bacdive.strain:161512</code></li>
<li><code>mediadive.medium:104c</code></li>
<li><code>NCBITaxon:408</code></li>
</ul>
<p><strong>Advantages:</strong>
- Unambiguous source attribution
- Valid CURIE format (no spaces or invalid chars)
- URL-expandable via bioregistry</p>
<p><strong>Documentation:</strong> See <a href="../best-practices/">best-practices.md</a> and <a href="https://github.com/turbomam/cmm-ai-automation/issues/53">#53</a></p>
<hr />
<h2 id="2-schema-driven-vs-ad-hoc-generation">2. Schema-Driven vs Ad-hoc Generation</h2>
<h3 id="kg-microbe-approach_1">kg-microbe Approach</h3>
<ul>
<li>TSV generation via custom Transform classes</li>
<li>No formal schema validation before export</li>
<li>Flexibility for rapid iteration</li>
</ul>
<h3 id="cmm-approach_1">CMM Approach</h3>
<ul>
<li><strong>LinkML schema first</strong>: <code>src/cmm_ai_automation/schema/cmm_ai_automation.yaml</code></li>
<li>Classes explicitly map to Biolink categories</li>
<li>Enums with <code>meaning:</code> slots for ontology grounding</li>
<li>Generated Python dataclasses via <code>gen-pydantic</code></li>
</ul>
<p><strong>Advantages:</strong>
- Type safety and validation
- Self-documenting data structures
- Easier to detect incompatibilities early</p>
<p><strong>Documentation:</strong> See <a href="../architecture/">architecture.md</a></p>
<hr />
<h2 id="3-biolink-model-compliance">3. Biolink Model Compliance</h2>
<h3 id="kg-microbe-issues">kg-microbe Issues</h3>
<p>From kg-microbe#438:
- <strong>186,197 edges</strong> use <code>biolink:capable_of</code> with wrong object type
- Objects categorized as <code>biolink:PhenotypicQuality</code> (EC enzyme codes)
- Biolink Model requires <code>Occurrent</code> (processes/activities) as range</p>
<p>From kg-microbe#436:
- Empty or inconsistent <code>relation</code> values in edges
- Non-CURIE <code>primary_knowledge_source</code> values
- Subject/object with spaces (see #430)</p>
<h3 id="cmm-approach_2">CMM Approach</h3>
<ul>
<li>Explicit Biolink category mappings in schema:
  <div class="highlight"><pre><span></span><code><span class="nt">classes</span><span class="p">:</span>
<span class="w">  </span><span class="nt">Strain</span><span class="p">:</span>
<span class="w">    </span><span class="nt">class_uri</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">biolink:OrganismalEntity</span>
<span class="w">  </span><span class="nt">Taxon</span><span class="p">:</span>
<span class="w">    </span><span class="nt">class_uri</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">biolink:OrganismTaxon</span>
<span class="w">  </span><span class="nt">GrowthMedium</span><span class="p">:</span>
<span class="w">    </span><span class="c1"># Evaluating: ChemicalMixture vs custom METPO class</span>
</code></pre></div></li>
<li>Edge predicates validated against Biolink Model or METPO</li>
<li><code>kgx validate</code> target (currently non-blocking during METPO registration - see #91)</li>
</ul>
<p><strong>Documentation:</strong> See <a href="https://github.com/turbomam/cmm-ai-automation/issues/59">#59</a> for alignment strategy</p>
<hr />
<h2 id="4-entity-resolution-and-provenance">4. Entity Resolution and Provenance</h2>
<h3 id="kg-microbe-approach_2">kg-microbe Approach</h3>
<ul>
<li>Generates nodes/edges per source</li>
<li>Post-merge deduplication via koza</li>
<li>Limited provenance metadata</li>
</ul>
<h3 id="cmm-approach_3">CMM Approach</h3>
<ul>
<li>
<p><strong>SourceRecord</strong> class tracks which API provided which fields:
  <div class="highlight"><pre><span></span><code><span class="nt">SourceRecord</span><span class="p">:</span>
<span class="w">  </span><span class="nt">slots</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">source_name</span><span class="w">      </span><span class="c1"># &quot;pubchem&quot;, &quot;chebi&quot;, &quot;node_normalizer&quot;</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">source_query</span><span class="w">     </span><span class="c1"># Original query used</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">source_timestamp</span><span class="w"> </span><span class="c1"># When fetched</span>
</code></pre></div></p>
</li>
<li>
<p><strong>DataConflict</strong> class records disagreements:
  <div class="highlight"><pre><span></span><code><span class="nt">DataConflict</span><span class="p">:</span>
<span class="w">  </span><span class="nt">slots</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">field_name</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">primary_source</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">primary_value</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conflicting_source</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">conflicting_value</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">resolution</span><span class="w">  </span><span class="c1"># enum: KEPT_PRIMARY, USED_CONFLICTING, MANUAL</span>
</code></pre></div></p>
</li>
</ul>
<p><strong>Advantages:</strong>
- Transparent data lineage
- Reproducible enrichment
- Explicit conflict resolution</p>
<p><strong>Documentation:</strong> See <code>src/cmm_ai_automation/store/enrichment_store.py</code></p>
<hr />
<h2 id="5-media-grounding-quality">5. Media Grounding Quality</h2>
<h3 id="kg-microbe-cmm-ai-approach-from-related-project">kg-microbe / CMM-AI Approach (from related project)</h3>
<p>Analysis of mappings from CultureBotAI/CMM-AI project (which uses kg-microbe as target):</p>
<ul>
<li><strong>Method:</strong> SQL <code>LIKE '%keyword%'</code> matching on media names</li>
<li><strong>Quality:</strong> 25% of mappings are WRONG due to substring matches</li>
<li>"MP" → "LAMPROBACTER" (contains "MP")</li>
<li>"MP" → "AMPHIBACILLUS" (contains "MP")</li>
</ul>
<p><strong>Documentation:</strong> See <a href="../kg_microbe_nodes_analysis/">kg_microbe_nodes_analysis.md</a></p>
<h3 id="cmm-approach-planned">CMM Approach (Planned)</h3>
<ul>
<li>
<p>Structured mapping table with quality metadata:
  <div class="highlight"><pre><span></span><code><span class="nt">MediaMapping</span><span class="p">:</span>
<span class="w">  </span><span class="nt">slots</span><span class="p">:</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">sheet_media_id</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">target_id</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">match_type</span><span class="w">  </span><span class="c1"># enum: EXACT, VARIANT, CLOSE, WRONG</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">confidence</span><span class="w">  </span><span class="c1"># 0.0-1.0</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">method</span><span class="w">      </span><span class="c1"># &quot;manual&quot;, &quot;semantic_search&quot;, &quot;sql_like&quot;</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">curator</span><span class="w">     </span><span class="c1"># Who validated it</span>
</code></pre></div></p>
</li>
<li>
<p>No semicolon-delimited multi-value fields (see #85)</p>
</li>
<li>Explicit tracking of alignment quality (see #84)</li>
</ul>
<hr />
<h2 id="6-testing-and-validation">6. Testing and Validation</h2>
<h3 id="kg-microbe">kg-microbe</h3>
<ul>
<li>Has test suite (tox.ini)</li>
<li>Does NOT run <code>kgx validate</code> in CI (see #81)</li>
<li>Manual validation findings documented in issues</li>
</ul>
<h3 id="cmm">CMM</h3>
<ul>
<li>Pytest suite: 426 tests passing</li>
<li>Pre-commit hooks: mypy, ruff, bandit, linkml-lint, pip-audit</li>
<li><code>kgx validate</code> target (non-blocking until METPO prefix registered)</li>
<li>Integration tests marked separately</li>
</ul>
<p><strong>Documentation:</strong> See <code>pyproject.toml</code> and <code>.pre-commit-config.yaml</code></p>
<hr />
<h2 id="7-kgx-field-usage-and-provenance">7. KGX Field Usage and Provenance</h2>
<h3 id="kg-microbe-field-issues">kg-microbe Field Issues</h3>
<p>From kg-microbe#436 (BacDive edges.tsv violations):
- <strong>Empty or inconsistent <code>relation</code> values</strong> - unclear when to populate
- <strong>Non-CURIE <code>primary_knowledge_source</code></strong> - violates infores: requirement
- <strong>Missing required provenance</strong> - knowledge_level, agent_type often absent</p>
<h3 id="cmm-approach-strict-kgx-compliance">CMM Approach: Strict KGX Compliance</h3>
<p><strong>Edge Model</strong> (<code>src/cmm_ai_automation/transform/kgx.py</code>):
<div class="highlight"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">KGXEdge</span><span class="p">(</span><span class="n">BaseModel</span><span class="p">):</span>
    <span class="c1"># Required fields</span>
    <span class="n">subject</span><span class="p">:</span> <span class="nb">str</span>                              <span class="c1"># CURIE</span>
    <span class="n">predicate</span><span class="p">:</span> <span class="nb">str</span>                            <span class="c1"># CURIE (biolink: or RO:)</span>
    <span class="nb">object</span><span class="p">:</span> <span class="nb">str</span>                               <span class="c1"># CURIE</span>
    <span class="n">knowledge_level</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span>                 <span class="c1"># Biolink enum</span>
        <span class="s2">&quot;knowledge_assertion&quot;</span><span class="p">,</span>
        <span class="s2">&quot;logical_entailment&quot;</span><span class="p">,</span>
        <span class="s2">&quot;prediction&quot;</span><span class="p">,</span>
        <span class="s2">&quot;statistical_association&quot;</span>
    <span class="p">]</span>
    <span class="n">agent_type</span><span class="p">:</span> <span class="n">Literal</span><span class="p">[</span>                      <span class="c1"># Biolink enum</span>
        <span class="s2">&quot;manual_agent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;automated_agent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;computational_model&quot;</span><span class="p">,</span>
        <span class="s2">&quot;not_provided&quot;</span>
    <span class="p">]</span>

    <span class="c1"># Provenance (recommended)</span>
    <span class="n">primary_knowledge_source</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span>  <span class="c1"># infores: CURIEs</span>
    <span class="n">aggregator_knowledge_source</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span>

    <span class="c1"># Optional but useful</span>
    <span class="nb">id</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span>                            <span class="c1"># Deterministic SHA256 hash</span>
    <span class="n">publications</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span>            <span class="c1"># PMID: CURIEs</span>
</code></pre></div></p>
<h4 id="cmm-rules">CMM Rules</h4>
<p><strong>1. Predicate Selection</strong></p>
<ul>
<li>Use <strong>Biolink predicates</strong> when available:</li>
<li><code>biolink:in_taxon</code> for taxonomy relationships</li>
<li><code>biolink:capable_of</code> for enzyme/process assertions</li>
<li>
<p><code>biolink:has_part</code> for composition</p>
</li>
<li>
<p>Use <strong>RO predicates</strong> for fine-grained semantics:</p>
</li>
<li><code>RO:0002162</code> (in taxon) - more specific than biolink</li>
<li>
<p><code>RO:0001019</code> (contains) for ingredient relationships</p>
</li>
<li>
<p><strong>Never use custom predicates</strong> without Biolink/OBO registration</p>
</li>
</ul>
<p><strong>2. Relation vs Predicate</strong></p>
<ul>
<li><code>predicate</code> field: REQUIRED, always a CURIE</li>
<li><code>relation</code> field: DEPRECATED in Biolink 3.0+, do not use</li>
<li>kg-microbe mixes these inconsistently</li>
</ul>
<p><strong>3. Provenance Requirements</strong></p>
<p>All edges MUST have:
- <code>knowledge_level</code>: How was this asserted?
  - <code>"knowledge_assertion"</code> - direct database claim
  - <code>"prediction"</code> - ML/AI inferred
  - <code>"statistical_association"</code> - correlation-based</p>
<ul>
<li><code>agent_type</code>: Who/what asserted it?</li>
<li><code>"manual_agent"</code> - human curator</li>
<li>
<p><code>"automated_agent"</code> - script/pipeline</p>
</li>
<li>
<p><code>primary_knowledge_source</code>: Where did it originate?</p>
</li>
<li>MUST use <code>infores:</code> prefix</li>
<li>Example: <code>["infores:bacdive"]</code>, <code>["infores:mediadive"]</code></li>
<li>Register at https://github.com/biolink/information-resource-registry</li>
</ul>
<p><strong>4. infores: Usage</strong></p>
<p>The <code>infores:</code> namespace identifies <strong>information resources</strong> (databases, APIs):</p>
<p>✅ <strong>CORRECT:</strong>
- <code>infores:bacdive</code> - BacDive database
- <code>infores:ncbi-taxonomy</code> - NCBI Taxonomy
- <code>infores:pubchem</code> - PubChem
- <code>infores:chebi</code> - ChEBI</p>
<p>❌ <strong>WRONG:</strong>
- <code>"BacDive"</code> - not a CURIE
- <code>bacdive:7142</code> - this is a strain ID, not a knowledge source
- <code>http://bacdive.dsmz.de</code> - not registered</p>
<p><strong>Before using a new infores:</strong>, check:
1. Is it registered in the <a href="https://github.com/biolink/information-resource-registry">Biolink information resource registry</a>?
2. If not, submit a registration PR</p>
<p><strong>5. Node Provenance</strong></p>
<p>Nodes use <code>provided_by</code> field (not <code>primary_knowledge_source</code>):
<div class="highlight"><pre><span></span><code><span class="n">KGXNode</span><span class="p">(</span>
    <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;NCBITaxon:408&quot;</span><span class="p">,</span>
    <span class="n">category</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;biolink:OrganismTaxon&quot;</span><span class="p">],</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Methylorubrum extorquens&quot;</span><span class="p">,</span>
    <span class="n">provided_by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;infores:ncbi-taxonomy&quot;</span><span class="p">]</span>
<span class="p">)</span>
</code></pre></div></p>
<p><strong>Deduplication</strong>: When same node from multiple sources, merge <code>provided_by</code>:
<div class="highlight"><pre><span></span><code><span class="c1"># Before merge:</span>
<span class="n">Node</span> <span class="n">A</span><span class="p">:</span> <span class="n">provided_by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;infores:ncbi&quot;</span><span class="p">]</span>
<span class="n">Node</span> <span class="n">B</span><span class="p">:</span> <span class="n">provided_by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;infores:bacdive&quot;</span><span class="p">]</span>

<span class="c1"># After merge:</span>
<span class="n">Node</span><span class="p">:</span> <span class="n">provided_by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;infores:bacdive&quot;</span><span class="p">,</span> <span class="s2">&quot;infores:ncbi&quot;</span><span class="p">]</span>  <span class="c1"># sorted</span>
</code></pre></div></p>
<p><strong>Documentation:</strong>
- <a href="../bacdive_kgx_pipeline/">bacdive_kgx_pipeline.md</a> - Provenance patterns
- <a href="../best_practices_strain_data_curation/">best_practices_strain_data_curation.md</a> - CURIE requirements</p>
<hr />
<h2 id="8-custom-vs-standard-kgx-fields">8. Custom vs Standard KGX Fields</h2>
<h3 id="kg-microbe-custom-fields">kg-microbe Custom Fields</h3>
<p>kg-microbe nodes/edges often include non-standard fields without documentation:
- Mixed use of Biolink standard vs custom properties
- Unclear which fields are queryable/meaningful
- No schema definition for validation</p>
<h3 id="cmm-approach-lenient-but-documented">CMM Approach: Lenient But Documented</h3>
<p><strong>Node/Edge models use Pydantic with <code>extra="allow"</code>:</strong>
<div class="highlight"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">KGXNode</span><span class="p">(</span><span class="n">BaseModel</span><span class="p">):</span>
    <span class="c1"># Required Biolink fields</span>
    <span class="nb">id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="n">category</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span>

    <span class="c1"># Optional Biolink fields  </span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">description</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">provided_by</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">xref</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">synonym</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">in_taxon</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="n">model_config</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;extra&quot;</span><span class="p">:</span> <span class="s2">&quot;allow&quot;</span><span class="p">}</span>  <span class="c1"># Permits custom fields</span>
</code></pre></div></p>
<p><strong>Custom fields we add:</strong>
- <code>type_strain: bool</code> - Type strain status (microbiology-specific)
- <code>biosafety_level: int</code> - Risk group (microbiology-specific)
- <code>strain_designation: str</code> - Lab designation (domain-specific)
- <code>has_genome: list[str]</code> - Genome assembly CURIEs</p>
<p><strong>Key difference</strong>: Custom fields are:
1. <strong>Documented</strong> in LinkML schema with descriptions
2. <strong>Typed</strong> with explicit Python type hints
3. <strong>Justified</strong> as domain-specific extensions
4. <strong>Proposed</strong> to Biolink Model when broadly applicable</p>
<p><strong>Example from schema:</strong>
<div class="highlight"><pre><span></span><code><span class="nt">slots</span><span class="p">:</span>
<span class="w">  </span><span class="nt">type_strain</span><span class="p">:</span>
<span class="w">    </span><span class="nt">description</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;Whether</span><span class="nv"> </span><span class="s">this</span><span class="nv"> </span><span class="s">strain</span><span class="nv"> </span><span class="s">is</span><span class="nv"> </span><span class="s">the</span><span class="nv"> </span><span class="s">nomenclatural</span><span class="nv"> </span><span class="s">type</span><span class="nv"> </span><span class="s">for</span><span class="nv"> </span><span class="s">the</span><span class="nv"> </span><span class="s">species&quot;</span>
<span class="w">    </span><span class="nt">range</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">boolean</span>
<span class="w">    </span><span class="nt">comments</span><span class="p">:</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&quot;Per</span><span class="nv"> </span><span class="s">ICNP</span><span class="nv"> </span><span class="s">(International</span><span class="nv"> </span><span class="s">Code</span><span class="nv"> </span><span class="s">of</span><span class="nv"> </span><span class="s">Nomenclature</span><span class="nv"> </span><span class="s">of</span><span class="nv"> </span><span class="s">Prokaryotes)&quot;</span>
<span class="w">      </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="s">&quot;Specific</span><span class="nv"> </span><span class="s">to</span><span class="nv"> </span><span class="s">microbiology,</span><span class="nv"> </span><span class="s">not</span><span class="nv"> </span><span class="s">general</span><span class="nv"> </span><span class="s">Biolink&quot;</span>
</code></pre></div></p>
<hr />
<h2 id="9-documented-gray-zones-incompatibilities">9. Documented Gray Zones / Incompatibilities</h2>
<h3 id="known-kg-microbe-issues-referenced-in-cmm-docs">Known kg-microbe Issues Referenced in CMM Docs</h3>
<ol>
<li><strong>Invalid CURIEs</strong> (kg-microbe#430)</li>
<li>100 nodes with spaces or extra colons</li>
<li>
<p>Documented in <a href="../best-practices/">best-practices.md</a></p>
</li>
<li>
<p><strong>Biolink Model violations</strong> (kg-microbe#438)</p>
</li>
<li>Wrong object types for <code>capable_of</code> predicate</li>
<li>
<p>186K+ edges affected</p>
</li>
<li>
<p><strong>Prefix ambiguity</strong></p>
</li>
<li><code>medium:</code> could be MediaDive, TogoMedium, or local</li>
<li>
<p>CMM uses scoped prefixes: <code>mediadive.medium:</code>, <code>togomedium:M443</code></p>
</li>
<li>
<p><strong>Media grounding quality</strong></p>
</li>
<li>Substring matching produces false positives</li>
<li>CMM plans explicit quality tracking (#84)</li>
</ol>
<h3 id="cmm-intentional-divergences">CMM Intentional Divergences</h3>
<table>
<thead>
<tr>
<th>Design Choice</th>
<th>CMM</th>
<th>kg-microbe</th>
<th>Rationale</th>
</tr>
</thead>
<tbody>
<tr>
<td>Prefix style</td>
<td><code>bacdive.strain:12345</code></td>
<td><code>strain:bacdive_12345</code></td>
<td>URL-expandable, unambiguous source</td>
</tr>
<tr>
<td>Schema</td>
<td>LinkML with Biolink mappings</td>
<td>Ad-hoc Python classes</td>
<td>Type safety, validation, documentation</td>
</tr>
<tr>
<td>Provenance</td>
<td><code>SourceRecord</code> and <code>DataConflict</code> classes</td>
<td>Implicit in file structure</td>
<td>Transparent lineage, conflict resolution</td>
</tr>
<tr>
<td>Validation</td>
<td>CI + kgx validate</td>
<td>Manual + tox</td>
<td>Catch issues early</td>
</tr>
<tr>
<td>Entity IDs</td>
<td>Canonical namespace policy (#98)</td>
<td>Post-merge dedup</td>
<td>One ID per entity, explicit policy</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="10-areas-of-potential-coordination">10. Areas of Potential Coordination</h2>
<p>Despite differences, there are collaboration opportunities:</p>
<ol>
<li><strong>MediaDive transform</strong> (#114)</li>
<li>Both projects transform MediaDive</li>
<li>CMM documents composition model gaps (#111)</li>
<li>
<p>Could share findings on solution nesting</p>
</li>
<li>
<p><strong>BacDive data quality</strong></p>
</li>
<li>CMM has detailed BacDive JSON schema analysis</li>
<li>
<p>Could contribute irregular path handling patterns (#69)</p>
</li>
<li>
<p><strong>METPO term proposals</strong> (#9, #106)</p>
</li>
<li>Both need cultivation-specific predicates</li>
<li>
<p>Could jointly propose terms to berkeleybop/metpo</p>
</li>
<li>
<p><strong>TogoMedium integration</strong> (#75)</p>
</li>
<li>Neither project fully integrates TogoMedium yet</li>
<li>
<p>Could share SPARQL patterns</p>
</li>
<li>
<p><strong>MicroMediaParam mappings</strong> (#19, #22)</p>
</li>
<li>kg-microbe references MicroMediaParam compound mappings</li>
<li>CMM ingredient enrichment (#23) could leverage same data</li>
</ol>
<hr />
<h2 id="11-when-to-use-each-approach">11. When to Use Each Approach</h2>
<h3 id="use-kg-microbe-when">Use kg-microbe When</h3>
<ul>
<li>Need comprehensive microbial KG quickly</li>
<li>Willing to accept some data quality issues</li>
<li>Post-processing pipeline can handle invalid CURIEs</li>
<li>Flexibility more important than strict validation</li>
</ul>
<h3 id="use-cmm-when">Use CMM When</h3>
<ul>
<li>Strict KGX/Biolink compliance required</li>
<li>Need explicit provenance tracking</li>
<li>Want schema-driven type safety</li>
<li>Working on CMM-specific (lanthanide/rare earth) research</li>
<li>Contributing data to Biolink ecosystem</li>
</ul>
<hr />
<h2 id="12-references">12. References</h2>
<h3 id="cmm-documentation">CMM Documentation</h3>
<ul>
<li><a href="../architecture/">Architecture</a></li>
<li><a href="../best-practices/">Best Practices</a></li>
<li><a href="../kg_microbe_nodes_analysis/">kg-microbe Nodes Analysis</a></li>
<li><a href="../session-notes-2025-12-16/">Session Notes 2025-12-16</a></li>
</ul>
<h3 id="cmm-issues">CMM Issues</h3>
<ul>
<li>
<h1 id="53-standardize-curie-prefixes">53 - Standardize CURIE prefixes</h1>
</li>
<li>
<h1 id="58-document-modeling-differences-from-kg-microbe">58 - Document modeling differences from kg-microbe</h1>
</li>
<li>
<h1 id="59-align-linkml-schema-with-kg-microbe-patterns">59 - Align LinkML schema with kg-microbe patterns</h1>
</li>
<li>
<h1 id="67-adopt-kg-microbe-cli-pattern-downloadtransformmerge">67 - Adopt kg-microbe CLI pattern (download/transform/merge)</h1>
</li>
<li>
<h1 id="81-add-kgx-validate-to-ci">81 - Add KGX validate to CI</h1>
</li>
<li>
<h1 id="84-track-alignment-quality">84 - Track alignment quality</h1>
</li>
<li>
<h1 id="85-avoid-semicolon-delimited-multi-value-fields">85 - Avoid semicolon-delimited multi-value fields</h1>
</li>
<li>
<h1 id="91-metpo-prefix-registration">91 - METPO prefix registration</h1>
</li>
<li>
<h1 id="98-canonical-namespace-policy">98 - Canonical namespace policy</h1>
</li>
<li>
<h1 id="114-coordinate-with-kg-microbe-mediadive-transform">114 - Coordinate with kg-microbe MediaDive transform</h1>
</li>
</ul>
<h3 id="kg-microbe-issues_1">kg-microbe Issues</h3>
<ul>
<li>kg-microbe#430 - Invalid CURIEs (spaces, colons)</li>
<li>kg-microbe#436 - KGX format violations</li>
<li>kg-microbe#438 - Biolink Model predicate range violations</li>
</ul>
<h3 id="external-resources">External Resources</h3>
<ul>
<li><a href="https://biolink.github.io/biolink-model/">Biolink Model</a></li>
<li><a href="https://bioregistry.io/">Bioregistry</a></li>
<li><a href="https://github.com/berkeleybop/metpo">METPO Ontology</a></li>
<li><a href="https://github.com/biolink/kgx">KGX Documentation</a></li>
</ul>
<hr />
<p><strong>Last Updated:</strong> 2026-01-07<br />
<strong>Maintainer:</strong> @turbomam</p>












                
  

              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
<div class="md-social">
  
</div>
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
      <div class="md-progress" data-md-component="progress" role="progressbar"></div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.tabs.link", "header.autohide", "navigation.expand", "navigation.instant", "navigation.instant.progress", "navigation.tracking", "search.suggest", "content.code.copy"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../assets/pymdownx-extras/extra-loader-B5CKpNQx.js"></script>
      
    
  </body>
</html>